
<!DOCTYPE html>


<html lang="en" data-content_root="./" data-theme="auto">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>PhaseLab API Guide &#8212; PhaseLab 0.6.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "auto";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "auto";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=09465d15" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=2bd037f0"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=30646c52"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'API_GUIDE';</script>
    <link rel="icon" href="_static/favicon.svg"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="auto">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo-lm.svg" class="logo__image only-light" alt=""/>
    <img src="_static/logo.svg" class="logo__image only-dark pst-js-only" alt=""/>
  
  
    <p class="title logo__title">PhaseLab</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/explanations/index.html">
    Explanations
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="reference/index.html">
    API Reference
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="research/index.html">
    Research Applications
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="developer/index.html">
    Developer Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="citing.html">
    Citing PhaseLab
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/explanations/index.html">
    Explanations
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="reference/index.html">
    API Reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="research/index.html">
    Research Applications
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="developer/index.html">
    Developer Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="citing.html">
    Citing PhaseLab
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">PhaseLab API Guide</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="phaselab-api-guide">
<h1>PhaseLab API Guide<a class="headerlink" href="#phaselab-api-guide" title="Link to this heading">#</a></h1>
<p><strong>Perturbation reliability analysis for biological and chemical systems</strong></p>
<blockquote>
<div><p><strong>v1.0.0 Update</strong>: This release introduces spatial coherence as the validated methodology for predicting perturbation outcomes. Guide-sequence coherence from earlier versions has been deprecated based on experimental validation (E200-E216).</p>
</div></blockquote>
<section id="what-is-phaselab">
<h2>What is PhaseLab?<a class="headerlink" href="#what-is-phaselab" title="Link to this heading">#</a></h2>
<p>PhaseLab helps you find <strong>reliable perturbation sites</strong> in biological systems. When you perturb a system (CRISPR edit, drug treatment, mutation), some locations give consistent results while others are unpredictable. PhaseLab identifies which regions will respond reliably.</p>
<p><strong>The key insight</strong>: Regions where nearby perturbations produce similar effects (spatially coherent) tend to give reproducible results. Regions where effects vary wildly position-to-position are unreliable.</p>
</section>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#installation">Installation</a></p></li>
<li><p><a class="reference internal" href="#quick-start">Quick Start</a></p></li>
<li><p><a class="reference internal" href="#core-concepts">Core Concepts</a></p>
<ul class="simple">
<li><p><a class="reference internal" href="#what-is-spatial-coherence">What is Spatial Coherence?</a></p></li>
<li><p><a class="reference internal" href="#stable-vs-amplifying-regions">Stable vs Amplifying Regions</a></p></li>
<li><p><a class="reference internal" href="#when-to-trust-predictions">When to Trust Predictions</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#landscapes-module">Landscapes Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#response-landscapes">Response Landscapes</a></p></li>
<li><p><a class="reference internal" href="#coherence-profiles">Coherence Profiles</a></p></li>
<li><p><a class="reference internal" href="#region-classification">Region Classification</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#spatial-analysis-module">Spatial Analysis Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#tiling-screen-analysis">Tiling Screen Analysis</a></p></li>
<li><p><a class="reference internal" href="#validated-results">Validated Results</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#crispr-module">CRISPR Module</a></p>
<ul class="simple">
<li><p><a class="reference internal" href="#important-deprecated-features">Important: Deprecated Features</a></p></li>
<li><p><a class="reference internal" href="#crispra-guide-design">CRISPRa Guide Design</a></p></li>
<li><p><a class="reference internal" href="#integration-with-spatial-coherence">Integration with Spatial Coherence</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#crispr-surf-integration">CRISPR-SURF Integration</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#what-is-crispr-surf">What is CRISPR-SURF?</a></p></li>
<li><p><a class="reference internal" href="#coherence-on-deconvolved-data">Coherence on Deconvolved Data</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#omics-module">Omics Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#atac-seq-coherence">ATAC-seq Coherence</a></p></li>
<li><p><a class="reference internal" href="#chip-seq-coherence">ChIP-seq Coherence</a></p></li>
<li><p><a class="reference internal" href="#rna-seq-coherence">RNA-seq Coherence</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#microbiology-module">Microbiology Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#tnseq-analysis">TnSeq Analysis</a></p></li>
<li><p><a class="reference internal" href="#bacterial-crispri-screens">Bacterial CRISPRi Screens</a></p></li>
<li><p><a class="reference internal" href="#drug-response-landscapes">Drug Response Landscapes</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#chemistry-module">Chemistry Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#binding-landscapes">Binding Landscapes</a></p></li>
<li><p><a class="reference internal" href="#reaction-optimization">Reaction Optimization</a></p></li>
<li><p><a class="reference internal" href="#hts-screening">HTS Screening</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#protein-module">Protein Module</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#mutational-scanning">Mutational Scanning</a></p></li>
<li><p><a class="reference internal" href="#structure-mapping">Structure Mapping</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#quantum-mode-configuration">Quantum Mode Configuration</a> <em>(v1.0.0+)</em></p>
<ul class="simple">
<li><p><a class="reference internal" href="#mode-options">Mode Options</a></p></li>
<li><p><a class="reference internal" href="#when-to-use-quantum">When to Use Quantum</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#sms-trials-module">SMS Trials Module</a></p>
<ul class="simple">
<li><p><a class="reference internal" href="#trial-runners">Trial Runners</a></p></li>
<li><p><a class="reference internal" href="#sms-pipeline">SMS Pipeline</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#circadian-module">Circadian Module</a></p>
<ul class="simple">
<li><p><a class="reference internal" href="#sms-clock-model">SMS Clock Model</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#complete-examples">Complete Examples</a></p></li>
<li><p><a class="reference internal" href="#api-reference">API Reference</a></p></li>
</ol>
</section>
<hr class="docutils" />
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Link to this heading">#</a></h2>
<section id="basic-installation">
<h3>Basic Installation<a class="headerlink" href="#basic-installation" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>phaselab
</pre></div>
</div>
</section>
<section id="with-quantum-support-ibm-quantum">
<h3>With Quantum Support (IBM Quantum)<a class="headerlink" href="#with-quantum-support-ibm-quantum" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>phaselab<span class="o">[</span>quantum<span class="o">]</span>
</pre></div>
</div>
</section>
<section id="with-plotting-support">
<h3>With Plotting Support<a class="headerlink" href="#with-plotting-support" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>phaselab<span class="o">[</span>plotting<span class="o">]</span>
</pre></div>
</div>
</section>
<section id="full-installation">
<h3>Full Installation<a class="headerlink" href="#full-installation" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>phaselab<span class="o">[</span>all<span class="o">]</span>
</pre></div>
</div>
</section>
<section id="development-installation">
<h3>Development Installation<a class="headerlink" href="#development-installation" title="Link to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/followthesapper/phaselab.git
<span class="nb">cd</span><span class="w"> </span>phaselab
pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span><span class="s2">&quot;.[dev]&quot;</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="quick-start">
<h2>Quick Start<a class="headerlink" href="#quick-start" title="Link to this heading">#</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># v1.0.0: Spatial coherence for perturbation reliability</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes</span><span class="w"> </span><span class="kn">import</span> <span class="n">ResponseLandscape</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.spatial</span><span class="w"> </span><span class="kn">import</span> <span class="n">analyze_tiling_coherence</span>

<span class="c1"># Your tiling screen data: positions and measured effects</span>
<span class="n">positions</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">110</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">130</span><span class="p">,</span> <span class="mi">140</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="mi">160</span><span class="p">,</span> <span class="mi">170</span><span class="p">,</span> <span class="mi">180</span><span class="p">,</span> <span class="mi">190</span><span class="p">]</span>
<span class="n">effects</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.82</span><span class="p">,</span> <span class="mf">0.78</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.12</span><span class="p">,</span> <span class="mf">0.88</span><span class="p">,</span> <span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]</span>

<span class="c1"># Create a response landscape</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">ResponseLandscape</span><span class="p">(</span>
    <span class="n">coords</span><span class="o">=</span><span class="n">positions</span><span class="p">,</span>
    <span class="n">responses</span><span class="o">=</span><span class="n">effects</span><span class="p">,</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;MYC&#39;</span><span class="p">,</span> <span class="s1">&#39;assay&#39;</span><span class="p">:</span> <span class="s1">&#39;CRISPRa&#39;</span><span class="p">}</span>
<span class="p">)</span>

<span class="c1"># Analyze spatial coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_tiling_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find stable regions (reliable for targeting)</span>
<span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">stable_regions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable region: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean effect: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;mean_response&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>What this tells you</strong>: Positions 100-140 and 160-190 have coherent responses (neighboring positions give similar effects), making them reliable targets. Position 150 is in an incoherent region - effects there are unpredictable.</p>
</section>
<hr class="docutils" />
<section id="core-concepts">
<h2>Core Concepts<a class="headerlink" href="#core-concepts" title="Link to this heading">#</a></h2>
<section id="what-is-spatial-coherence">
<h3>What is Spatial Coherence?<a class="headerlink" href="#what-is-spatial-coherence" title="Link to this heading">#</a></h3>
<p>When you perturb a biological system at different positions (CRISPR guides across a promoter, mutations across a protein, drugs at different concentrations), you get a <strong>response landscape</strong> - a map of position → effect.</p>
<p><strong>Spatial coherence</strong> measures how smoothly effects change across positions:</p>
<ul class="simple">
<li><p><strong>High coherence</strong>: Nearby positions give similar effects</p></li>
<li><p><strong>Low coherence</strong>: Effects jump around unpredictably</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>High Coherence (Stable):          Low Coherence (Unreliable):
Effect                            Effect
  │   ●●●●●                         │   ●   ●
  │  ●     ●●                       │ ●   ●
  │ ●        ●                      │    ●  ●
  │●          ●●●                   │  ●     ●
  └────────────── Position          └────────────── Position
</pre></div>
</div>
</section>
<section id="stable-vs-amplifying-regions">
<h3>Stable vs Amplifying Regions<a class="headerlink" href="#stable-vs-amplifying-regions" title="Link to this heading">#</a></h3>
<p>PhaseLab classifies regions into stability categories:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Category</p></th>
<th class="head"><p>Coherence</p></th>
<th class="head"><p>Meaning</p></th>
<th class="head"><p>Recommendation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>STABLE</strong></p></td>
<td><p>&gt; 0.7</p></td>
<td><p>Predictable, reproducible</p></td>
<td><p>Safe to target</p></td>
</tr>
<tr class="row-odd"><td><p><strong>MIXED</strong></p></td>
<td><p>0.4-0.7</p></td>
<td><p>Moderate variability</p></td>
<td><p>Use with caution</p></td>
</tr>
<tr class="row-even"><td><p><strong>AMPLIFYING</strong></p></td>
<td><p>&lt; 0.4</p></td>
<td><p>Highly variable</p></td>
<td><p>Avoid</p></td>
</tr>
<tr class="row-odd"><td><p><strong>IRRELEVANT</strong></p></td>
<td><p>-</p></td>
<td><p>No measurable effect</p></td>
<td><p>Skip</p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Amplifying regions</strong> (like super-enhancers) actually show <em>positive</em> correlation between local variance and global variance - small perturbations can have outsized effects. These are biologically important but therapeutically risky.</p>
</section>
<section id="when-to-trust-predictions">
<h3>When to Trust Predictions<a class="headerlink" href="#when-to-trust-predictions" title="Link to this heading">#</a></h3>
<p>PhaseLab predictions are most reliable when:</p>
<ol class="arabic simple">
<li><p><strong>You have dense tiling data</strong>: At least 50 perturbations across your region</p></li>
<li><p><strong>Coherence is validated</strong>: The negative correlation between local coherence and outcome variance is statistically significant (p &lt; 0.05)</p></li>
<li><p><strong>You’re selecting within stable regions</strong>: Guides/perturbations in high-coherence zones</p></li>
</ol>
<p><strong>Experimental validation</strong> (E213-E216) showed:</p>
<ul class="simple">
<li><p>32-49% variance reduction when selecting guides from stable regions</p></li>
<li><p>Correlation r = -0.24 to -0.50 between coherence and outcome variance</p></li>
<li><p>115,251 sgRNAs tested across 6 genes</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="landscapes-module">
<h2>Landscapes Module<a class="headerlink" href="#landscapes-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>The landscapes module provides the core data structures for perturbation-response analysis.</p>
<section id="response-landscapes">
<h3>Response Landscapes<a class="headerlink" href="#response-landscapes" title="Link to this heading">#</a></h3>
<p>A <code class="docutils literal notranslate"><span class="pre">ResponseLandscape</span></code> represents any perturbation screen data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes</span><span class="w"> </span><span class="kn">import</span> <span class="n">ResponseLandscape</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># From numpy arrays</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">ResponseLandscape</span><span class="p">(</span>
    <span class="n">coords</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">50</span><span class="p">]),</span>  <span class="c1"># Positions</span>
    <span class="n">responses</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">,</span> <span class="mf">0.82</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.78</span><span class="p">,</span> <span class="mf">0.85</span><span class="p">]),</span>  <span class="c1"># Effects</span>
    <span class="n">coord_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;guide_1&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_2&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_3&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_4&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_5&#39;</span><span class="p">,</span> <span class="s1">&#39;guide_6&#39;</span><span class="p">],</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="s1">&#39;RAI1&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_line&#39;</span><span class="p">:</span> <span class="s1">&#39;HEK293T&#39;</span><span class="p">}</span>
<span class="p">)</span>

<span class="c1"># Properties</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Positions: </span><span class="si">{</span><span class="n">landscape</span><span class="o">.</span><span class="n">n_coords</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean response: </span><span class="si">{</span><span class="n">landscape</span><span class="o">.</span><span class="n">mean_response</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Response range: </span><span class="si">{</span><span class="n">landscape</span><span class="o">.</span><span class="n">response_range</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="coherence-profiles">
<h3>Coherence Profiles<a class="headerlink" href="#coherence-profiles" title="Link to this heading">#</a></h3>
<p>A <code class="docutils literal notranslate"><span class="pre">CoherenceProfile</span></code> captures how coherence varies across the landscape:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes</span><span class="w"> </span><span class="kn">import</span> <span class="n">CoherenceProfile</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes.coherence</span><span class="w"> </span><span class="kn">import</span> <span class="n">compute_spatial_coherence</span>

<span class="c1"># Compute coherence profile</span>
<span class="n">profile</span> <span class="o">=</span> <span class="n">compute_spatial_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Access profile data</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean coherence: </span><span class="si">{</span><span class="n">profile</span><span class="o">.</span><span class="n">mean_coherence</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Variance correlation: </span><span class="si">{</span><span class="n">profile</span><span class="o">.</span><span class="n">correlation</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;P-value: </span><span class="si">{</span><span class="n">profile</span><span class="o">.</span><span class="n">p_value</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Validated: </span><span class="si">{</span><span class="n">profile</span><span class="o">.</span><span class="n">is_validated</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Per-position coherence values</span>
<span class="k">for</span> <span class="n">pos</span><span class="p">,</span> <span class="n">coh</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">profile</span><span class="o">.</span><span class="n">coords</span><span class="p">,</span> <span class="n">profile</span><span class="o">.</span><span class="n">coherence</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Position </span><span class="si">{</span><span class="n">pos</span><span class="si">}</span><span class="s2">: coherence = </span><span class="si">{</span><span class="n">coh</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="region-classification">
<h3>Region Classification<a class="headerlink" href="#region-classification" title="Link to this heading">#</a></h3>
<p>Classify your landscape into stability regions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes.classification</span><span class="w"> </span><span class="kn">import</span> <span class="n">classify_regions</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.landscapes</span><span class="w"> </span><span class="kn">import</span> <span class="n">StabilityClass</span>

<span class="c1"># Classify regions</span>
<span class="n">classification</span> <span class="o">=</span> <span class="n">classify_regions</span><span class="p">(</span>
    <span class="n">landscape</span><span class="p">,</span>
    <span class="n">profile</span><span class="o">=</span><span class="n">profile</span><span class="p">,</span>
    <span class="n">stable_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
    <span class="n">amplifying_threshold</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Access classified regions</span>
<span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">stability</span><span class="p">,</span> <span class="n">score</span> <span class="ow">in</span> <span class="n">classification</span><span class="o">.</span><span class="n">regions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">] </span><span class="si">{</span><span class="n">stability</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">: coherence=</span><span class="si">{</span><span class="n">score</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Filter by stability class</span>
<span class="n">stable</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">classification</span><span class="o">.</span><span class="n">regions</span> <span class="k">if</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="n">StabilityClass</span><span class="o">.</span><span class="n">STABLE</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">stable</span><span class="p">)</span><span class="si">}</span><span class="s2"> stable regions&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="spatial-analysis-module">
<h2>Spatial Analysis Module<a class="headerlink" href="#spatial-analysis-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>The spatial module implements the E213-validated methodology for tiling screen analysis.</p>
<section id="tiling-screen-analysis">
<h3>Tiling Screen Analysis<a class="headerlink" href="#tiling-screen-analysis" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.spatial</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">analyze_tiling_coherence</span><span class="p">,</span>
    <span class="n">TilingResult</span><span class="p">,</span>
    <span class="n">load_tiling_screen</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load from file (supports TSV, CSV)</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_tiling_screen</span><span class="p">(</span>
    <span class="s1">&#39;my_screen.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;genomic_position&#39;</span><span class="p">,</span>
    <span class="n">response_col</span><span class="o">=</span><span class="s1">&#39;log2fc&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;MYC&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Full analysis</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_tiling_coherence</span><span class="p">(</span>
    <span class="n">landscape</span><span class="p">,</span>
    <span class="n">window</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>                  <span class="c1"># Positions for local coherence</span>
    <span class="n">stable_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>       <span class="c1"># Coherence threshold for stable</span>
    <span class="n">min_region_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>          <span class="c1"># Minimum positions per region</span>
<span class="p">)</span>

<span class="c1"># Results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
<span class="c1">#</span>
<span class="c1"># TILING COHERENCE ANALYSIS: MYC</span>
<span class="c1"># ============================================================</span>
<span class="c1"># Positions: 1847</span>
<span class="c1"># Coherence-variance correlation: -0.312</span>
<span class="c1"># P-value: 0.0001</span>
<span class="c1"># VALIDATED: YES</span>
<span class="c1">#</span>
<span class="c1"># Stable regions: 12</span>
<span class="c1"># Amplifying regions: 3</span>
<span class="c1"># Mixed regions: 8</span>
<span class="c1">#</span>

<span class="c1"># Access stable regions for targeting</span>
<span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">stable_regions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Region </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Positions: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;n_positions&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean effect: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;mean_response&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="validated-results">
<h3>Validated Results<a class="headerlink" href="#validated-results" title="Link to this heading">#</a></h3>
<p>The E213 methodology has been validated on real tiling screen data:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Gene</p></th>
<th class="head"><p>sgRNAs</p></th>
<th class="head"><p>Correlation</p></th>
<th class="head"><p>Variance Reduction</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>MYC</p></td>
<td><p>1,847</p></td>
<td><p>+0.45*</p></td>
<td><p>N/A (amplifying)</p></td>
</tr>
<tr class="row-odd"><td><p>CD69</p></td>
<td><p>5,012</p></td>
<td><p>-0.31</p></td>
<td><p>38%</p></td>
</tr>
<tr class="row-even"><td><p>IL2RA</p></td>
<td><p>8,234</p></td>
<td><p>-0.28</p></td>
<td><p>32%</p></td>
</tr>
<tr class="row-odd"><td><p>HBG1/2</p></td>
<td><p>12,456</p></td>
<td><p>-0.50</p></td>
<td><p>49%</p></td>
</tr>
<tr class="row-even"><td><p>BCL11A</p></td>
<td><p>15,789</p></td>
<td><p>-0.24</p></td>
<td><p>33%</p></td>
</tr>
<tr class="row-odd"><td><p>GATA1</p></td>
<td><p>18,234</p></td>
<td><p>-0.41</p></td>
<td><p>44%</p></td>
</tr>
</tbody>
</table>
</div>
<p>*MYC shows positive correlation because it contains a super-enhancer (amplifying region). This is biologically correct - PhaseLab correctly identifies it as high-risk for therapeutic targeting.</p>
</section>
</section>
<hr class="docutils" />
<section id="crispr-module">
<h2>CRISPR Module<a class="headerlink" href="#crispr-module" title="Link to this heading">#</a></h2>
<p>The CRISPR module provides tools for CRISPR guide RNA design.</p>
<section id="important-deprecated-features">
<h3>Important: Deprecated Features<a class="headerlink" href="#important-deprecated-features" title="Link to this heading">#</a></h3>
<blockquote>
<div><p><strong>v1.0.0 BREAKING CHANGE</strong>: Guide-sequence coherence has been <strong>deprecated</strong>.</p>
<p>Experiments E200-E211 showed that computing coherence from guide sequences themselves (GC content, thermodynamic properties, etc.) has <strong>no predictive value</strong> for experimental outcomes (r ≈ 0).</p>
<p>Instead, use <strong>spatial coherence</strong> from tiling screen data via the <code class="docutils literal notranslate"><span class="pre">phaselab.spatial</span></code> module to identify reliable targeting regions.</p>
</div></blockquote>
<p><strong>What this means for your code:</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># OLD (deprecated, does nothing in v1.0.0):</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span>
<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">tss_index</span><span class="p">,</span> <span class="n">compute_coherence</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># Warning: compute_coherence is deprecated and has no effect</span>

<span class="c1"># NEW (v1.0.0 approach):</span>
<span class="c1"># 1. Get spatial coherence from tiling screen</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.spatial</span><span class="w"> </span><span class="kn">import</span> <span class="n">analyze_tiling_coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_tiling_coherence</span><span class="p">(</span><span class="n">tiling_landscape</span><span class="p">)</span>

<span class="c1"># 2. Filter guides to stable regions</span>
<span class="n">stable_positions</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">stable_regions</span><span class="p">]</span>

<span class="c1"># 3. Design guides only in stable regions</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_crispra_guides</span>
<span class="n">guides</span> <span class="o">=</span> <span class="n">design_crispra_guides</span><span class="p">(</span>
    <span class="n">promoter_sequence</span><span class="o">=</span><span class="n">sequence</span><span class="p">,</span>
    <span class="n">tss_position</span><span class="o">=</span><span class="n">tss</span><span class="p">,</span>
    <span class="n">restrict_to_positions</span><span class="o">=</span><span class="n">stable_positions</span><span class="p">,</span>  <span class="c1"># Only stable regions</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="crispra-guide-design">
<h3>CRISPRa Guide Design<a class="headerlink" href="#crispra-guide-design" title="Link to this heading">#</a></h3>
<p>The primary API for CRISPRa guide design with binding-aware enumeration:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_crispra_guides</span><span class="p">,</span> <span class="n">Nuclease</span><span class="p">,</span> <span class="n">NucleaseRole</span>

<span class="c1"># Design CRISPRa guides with explicit binding mode</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">design_crispra_guides</span><span class="p">(</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s2">&quot;Rai1&quot;</span><span class="p">,</span>
    <span class="n">promoter_sequence</span><span class="o">=</span><span class="n">promoter_seq</span><span class="p">,</span>
    <span class="n">tss_position</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span>
    <span class="n">nuclease</span><span class="o">=</span><span class="n">Nuclease</span><span class="o">.</span><span class="n">SACAS9</span><span class="p">,</span>
    <span class="n">relaxed_pam</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># BINDING mode (default for CRISPRa)</span>
    <span class="n">guide_length</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>     <span class="c1"># Override default 21bp</span>
<span class="p">)</span>

<span class="c1"># Access results by tier</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total guides: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">total_guides</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tier A (safest): </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">tier_a_guides</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tier B (good): </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">tier_b_guides</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tier C (acceptable): </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">tier_c_guides</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># View top candidates</span>
<span class="k">for</span> <span class="n">guide</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">tier_a_guides</span><span class="p">[:</span><span class="mi">5</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">guide</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> TSS</span><span class="si">{</span><span class="n">guide</span><span class="p">[</span><span class="s1">&#39;tss_relative_position&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">+d</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Key v0.9.3 Features:</strong></p>
<ul class="simple">
<li><p><strong>NucleaseRole</strong>: <code class="docutils literal notranslate"><span class="pre">BINDING</span></code> (CRISPRa/CRISPRi) vs <code class="docutils literal notranslate"><span class="pre">CUTTING</span></code> (knockout)</p></li>
<li><p><strong>Relaxed PAM patterns</strong>: SaCas9 uses NNGRRN for binding vs NNGRRT for cutting</p></li>
<li><p><strong>Sliding binding register</strong>: ±2bp enumeration in BINDING mode</p></li>
<li><p><strong>Configurable guide length</strong>: Override defaults (e.g., 20bp with SaCas9)</p></li>
</ul>
<p><strong>Why this matters</strong>: Standard CRISPR pipelines use rigid PAM-spacer anchoring that works for cutting but systematically misses experimentally validated CRISPRa guides in GC-dense promoters.</p>
</section>
<section id="pam-scanning">
<h3>PAM Scanning<a class="headerlink" href="#pam-scanning" title="Link to this heading">#</a></h3>
<p>Find all PAM sites in a sequence:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">find_pam_sites</span>

<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;ATGCGATCGATCGAGGCGATCGATCGATCGATCGATCGATCG&quot;</span>

<span class="c1"># Find NGG PAM sites (default for SpCas9)</span>
<span class="n">sites</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">)</span>
<span class="c1"># Returns: [(position, strand, guide_sequence), ...]</span>

<span class="c1"># Find multiple PAM types</span>
<span class="n">sites_ngg</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">)</span>   <span class="c1"># SpCas9</span>
<span class="n">sites_nag</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NAG&quot;</span><span class="p">)</span>   <span class="c1"># SpCas9 alternate</span>
<span class="n">sites_tttn</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;TTTN&quot;</span><span class="p">)</span> <span class="c1"># Cas12a</span>
</pre></div>
</div>
<p><strong>Parameters:</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code>: DNA sequence string</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pam</span></code>: PAM sequence pattern (default: <code class="docutils literal notranslate"><span class="pre">&quot;NGG&quot;</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">guide_length</span></code>: Length of guide sequence (default: <code class="docutils literal notranslate"><span class="pre">20</span></code>)</p></li>
</ul>
</section>
<section id="guide-scoring">
<h3>Guide Scoring<a class="headerlink" href="#guide-scoring" title="Link to this heading">#</a></h3>
<p>Score individual guide sequences:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">score_guide</span><span class="p">,</span> <span class="n">gc_content</span><span class="p">,</span> <span class="n">santalucia_delta_g</span>

<span class="n">guide</span> <span class="o">=</span> <span class="s2">&quot;GAAGGAGAGCAAGAGCGCGA&quot;</span>

<span class="c1"># GC content (optimal: 40-70%)</span>
<span class="n">gc</span> <span class="o">=</span> <span class="n">gc_content</span><span class="p">(</span><span class="n">guide</span><span class="p">)</span>
<span class="c1"># Returns: 0.60</span>

<span class="c1"># Thermodynamic stability (SantaLucia ΔG)</span>
<span class="n">delta_g</span> <span class="o">=</span> <span class="n">santalucia_delta_g</span><span class="p">(</span><span class="n">guide</span><span class="p">)</span>
<span class="c1"># Returns: -45.2 (kcal/mol)</span>

<span class="c1"># Combined scoring</span>
<span class="n">score</span> <span class="o">=</span> <span class="n">score_guide</span><span class="p">(</span>
    <span class="n">guide</span><span class="p">,</span>
    <span class="n">gc_weight</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span>
    <span class="n">thermo_weight</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span>
    <span class="n">position_weight</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
    <span class="n">accessibility_weight</span><span class="o">=</span><span class="mf">0.2</span>
<span class="p">)</span>
<span class="c1"># Returns: 0.78</span>
</pre></div>
</div>
</section>
<section id="design-pipeline">
<h3>Design Pipeline<a class="headerlink" href="#design-pipeline" title="Link to this heading">#</a></h3>
<p>The high-level pipeline for guide RNA design:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span>

<span class="c1"># Basic usage</span>
<span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">ATGCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCG</span>
<span class="s2">AGGCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCGATCG</span>
<span class="s2">&quot;&quot;&quot;</span>
<span class="n">tss_index</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># Transcription start site position</span>

<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">tss_index</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">guides</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
<p><strong>Output DataFrame columns:</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code>: 20bp guide sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">position</span></code>: Distance from TSS (negative = upstream)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">strand</span></code>: <code class="docutils literal notranslate"><span class="pre">+</span></code> or <code class="docutils literal notranslate"><span class="pre">-</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gc_content</span></code>: GC fraction (0-1)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">delta_g</span></code>: Thermodynamic stability</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined_score</span></code>: Weighted composite score</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">go_no_go</span></code>: Classification status</p></li>
</ul>
<p><strong>Advanced usage with custom configuration:</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span><span class="p">,</span> <span class="n">CRISPRConfig</span>

<span class="c1"># Custom configuration</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">CRISPRConfig</span><span class="p">(</span>
    <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">,</span>
    <span class="n">guide_length</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">window_upstream</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>    <span class="c1"># Search 500bp upstream of TSS</span>
    <span class="n">window_downstream</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>  <span class="c1"># Search 100bp downstream</span>
    <span class="n">min_gc</span><span class="o">=</span><span class="mf">0.40</span><span class="p">,</span>
    <span class="n">max_gc</span><span class="o">=</span><span class="mf">0.70</span><span class="p">,</span>
    <span class="n">weights</span><span class="o">=</span><span class="p">{</span>
        <span class="s1">&#39;gc&#39;</span><span class="p">:</span> <span class="mf">0.25</span><span class="p">,</span>
        <span class="s1">&#39;thermo&#39;</span><span class="p">:</span> <span class="mf">0.25</span><span class="p">,</span>
        <span class="s1">&#39;position&#39;</span><span class="p">:</span> <span class="mf">0.25</span><span class="p">,</span>
        <span class="s1">&#39;accessibility&#39;</span><span class="p">:</span> <span class="mf">0.25</span>
    <span class="p">}</span>
<span class="p">)</span>

<span class="c1"># With DNase accessibility peaks</span>
<span class="n">dnase_peaks</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">45</span><span class="p">,</span> <span class="mi">85</span><span class="p">),</span> <span class="p">(</span><span class="mi">120</span><span class="p">,</span> <span class="mi">180</span><span class="p">)]</span>  <span class="c1"># Accessible regions</span>

<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span>
    <span class="n">sequence</span><span class="p">,</span>
    <span class="n">tss_index</span><span class="p">,</span>
    <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span>
    <span class="n">dnase_peaks</span><span class="o">=</span><span class="n">dnase_peaks</span><span class="p">,</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Filter top candidates</span>
<span class="n">top_guides</span> <span class="o">=</span> <span class="n">guides</span><span class="p">[</span><span class="n">guides</span><span class="p">[</span><span class="s1">&#39;go_no_go&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;GO&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="integration-with-spatial-coherence">
<h3>Integration with Spatial Coherence<a class="headerlink" href="#integration-with-spatial-coherence" title="Link to this heading">#</a></h3>
<p>The recommended workflow combines spatial coherence analysis with guide design:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.spatial</span><span class="w"> </span><span class="kn">import</span> <span class="n">analyze_tiling_coherence</span><span class="p">,</span> <span class="n">load_tiling_screen</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_crispra_guides</span>

<span class="c1"># 1. Load your tiling screen data</span>
<span class="n">tiling_data</span> <span class="o">=</span> <span class="n">load_tiling_screen</span><span class="p">(</span>
    <span class="s1">&#39;rai1_tiling_screen.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;position&#39;</span><span class="p">,</span>
    <span class="n">response_col</span><span class="o">=</span><span class="s1">&#39;log2fc&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># 2. Analyze spatial coherence</span>
<span class="n">coherence_result</span> <span class="o">=</span> <span class="n">analyze_tiling_coherence</span><span class="p">(</span><span class="n">tiling_data</span><span class="p">)</span>

<span class="c1"># 3. Get stable region boundaries</span>
<span class="n">stable_regions</span> <span class="o">=</span> <span class="n">coherence_result</span><span class="o">.</span><span class="n">stable_regions</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">stable_regions</span><span class="p">)</span><span class="si">}</span><span class="s2"> stable regions&quot;</span><span class="p">)</span>

<span class="c1"># 4. Design guides restricted to stable regions</span>
<span class="n">guides</span> <span class="o">=</span> <span class="n">design_crispra_guides</span><span class="p">(</span>
    <span class="n">promoter_sequence</span><span class="o">=</span><span class="n">rai1_promoter</span><span class="p">,</span>
    <span class="n">tss_position</span><span class="o">=</span><span class="mi">600</span><span class="p">,</span>
    <span class="c1"># Only consider positions in stable regions</span>
    <span class="n">position_filter</span><span class="o">=</span><span class="k">lambda</span> <span class="n">pos</span><span class="p">:</span> <span class="nb">any</span><span class="p">(</span>
        <span class="n">r</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">pos</span> <span class="o">&lt;=</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">stable_regions</span>
    <span class="p">),</span>
<span class="p">)</span>

<span class="c1"># 5. Annotate guides with local coherence</span>
<span class="k">for</span> <span class="n">guide</span> <span class="ow">in</span> <span class="n">guides</span><span class="o">.</span><span class="n">candidates</span><span class="p">:</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">guide</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span>
    <span class="c1"># Find coherence at this position</span>
    <span class="n">local_coh</span> <span class="o">=</span> <span class="n">coherence_result</span><span class="o">.</span><span class="n">get_coherence_at</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
    <span class="n">guide</span><span class="p">[</span><span class="s1">&#39;local_coherence&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">local_coh</span>
    <span class="n">guide</span><span class="p">[</span><span class="s1">&#39;region_status&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;STABLE&#39;</span> <span class="k">if</span> <span class="n">local_coh</span> <span class="o">&gt;</span> <span class="mf">0.7</span> <span class="k">else</span> <span class="s1">&#39;MIXED&#39;</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="crispr-surf-integration">
<h2>CRISPR-SURF Integration<a class="headerlink" href="#crispr-surf-integration" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<section id="what-is-crispr-surf">
<h3>What is CRISPR-SURF?<a class="headerlink" href="#what-is-crispr-surf" title="Link to this heading">#</a></h3>
<p><a class="reference external" href="https://github.com/pinellolab/CRISPR-SURF">CRISPR-SURF</a> is a deconvolution algorithm that separates true regulatory signal from guide-specific noise in tiling screens. PhaseLab integrates with SURF output to provide coherence analysis on cleaner data.</p>
<p><strong>Why use SURF + PhaseLab together?</strong></p>
<ul class="simple">
<li><p>SURF removes guide-specific artifacts (efficiency, accessibility)</p></li>
<li><p>PhaseLab then measures spatial coherence of the deconvolved signal</p></li>
<li><p>Result: More reliable identification of regulatory regions</p></li>
</ul>
</section>
<section id="coherence-on-deconvolved-data">
<h3>Coherence on Deconvolved Data<a class="headerlink" href="#coherence-on-deconvolved-data" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.surf</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">parse_surf_output</span><span class="p">,</span>
    <span class="n">compute_surf_coherence</span><span class="p">,</span>
    <span class="n">SURFPipeline</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Parse SURF output files</span>
<span class="n">surf_output</span> <span class="o">=</span> <span class="n">parse_surf_output</span><span class="p">(</span>
    <span class="n">beta_file</span><span class="o">=</span><span class="s1">&#39;surf_results/beta_profile.tsv&#39;</span><span class="p">,</span>
    <span class="n">regions_file</span><span class="o">=</span><span class="s1">&#39;surf_results/significant_regions.bed&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Compute coherence on deconvolved data</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">compute_surf_coherence</span><span class="p">(</span>
    <span class="n">surf_output</span><span class="p">,</span>
    <span class="n">window</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">stable_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Compare raw vs deconvolved coherence</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Raw data coherence: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">raw_coherence</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SURF deconvolved coherence: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">deconvolved_coherence</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Improvement: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">coherence_improvement</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># High-confidence targets (stable in deconvolved data)</span>
<span class="k">for</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">high_confidence_targets</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">target</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: β=</span><span class="si">{</span><span class="n">target</span><span class="p">[</span><span class="s1">&#39;beta&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, coh=</span><span class="si">{</span><span class="n">target</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="full-surf-pipeline">
<h3>Full SURF Pipeline<a class="headerlink" href="#full-surf-pipeline" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.surf</span><span class="w"> </span><span class="kn">import</span> <span class="n">SURFPipeline</span><span class="p">,</span> <span class="n">SURFPipelineConfig</span>

<span class="c1"># Configure pipeline</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">SURFPipelineConfig</span><span class="p">(</span>
    <span class="n">surf_path</span><span class="o">=</span><span class="s1">&#39;/path/to/CRISPR-SURF&#39;</span><span class="p">,</span>  <span class="c1"># SURF installation</span>
    <span class="n">window</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">stable_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
    <span class="n">run_surf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Run SURF or use existing output</span>
<span class="p">)</span>

<span class="c1"># Create and run pipeline</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">SURFPipeline</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">input_file</span><span class="o">=</span><span class="s1">&#39;tiling_screen.tsv&#39;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;surf_analysis/&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Results include both SURF and coherence analysis</span>
<span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="omics-module">
<h2>Omics Module<a class="headerlink" href="#omics-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>The omics module applies spatial coherence to common genomics assays.</p>
<section id="atac-seq-coherence">
<h3>ATAC-seq Coherence<a class="headerlink" href="#atac-seq-coherence" title="Link to this heading">#</a></h3>
<p>Identify stably accessible chromatin regions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.omics</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_atac_data</span><span class="p">,</span>
    <span class="n">analyze_atac_coherence</span><span class="p">,</span>
    <span class="n">ATACLandscape</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load ATAC-seq signal</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_atac_data</span><span class="p">(</span>
    <span class="s1">&#39;atac_signal.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;position&#39;</span><span class="p">,</span>
    <span class="n">signal_col</span><span class="o">=</span><span class="s1">&#39;signal&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span>
    <span class="n">cell_type</span><span class="o">=</span><span class="s1">&#39;neurons&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_atac_coherence</span><span class="p">(</span>
    <span class="n">landscape</span><span class="p">,</span>
    <span class="n">window</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">stable_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
    <span class="n">peak_threshold</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>  <span class="c1"># std above mean for peaks</span>
<span class="p">)</span>

<span class="c1"># Find stable accessible regions (good for CRISPR targeting)</span>
<span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">accessible_stable</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable peak: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean signal: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;mean_signal&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">region</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="chip-seq-coherence">
<h3>ChIP-seq Coherence<a class="headerlink" href="#chip-seq-coherence" title="Link to this heading">#</a></h3>
<p>Identify stable protein-DNA binding sites:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.omics</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_chip_data</span><span class="p">,</span> <span class="n">analyze_chip_coherence</span>

<span class="c1"># Load ChIP-seq signal</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_chip_data</span><span class="p">(</span>
    <span class="s1">&#39;h3k27ac_chip.tsv&#39;</span><span class="p">,</span>
    <span class="n">target</span><span class="o">=</span><span class="s1">&#39;H3K27ac&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;MYC&#39;</span><span class="p">,</span>
    <span class="n">cell_type</span><span class="o">=</span><span class="s1">&#39;K562&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_chip_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Stable binding sites</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total peaks: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">peak_regions</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable binding sites: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">n_stable_peaks</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="rna-seq-coherence">
<h3>RNA-seq Coherence<a class="headerlink" href="#rna-seq-coherence" title="Link to this heading">#</a></h3>
<p>Identify genes with reliable expression changes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.omics</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_expression_data</span><span class="p">,</span> <span class="n">analyze_expression_coherence</span>

<span class="c1"># Load expression data (genes as &quot;positions&quot;)</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_expression_data</span><span class="p">(</span>
    <span class="s1">&#39;differential_expression.tsv&#39;</span><span class="p">,</span>
    <span class="n">gene_col</span><span class="o">=</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span>
    <span class="n">expression_col</span><span class="o">=</span><span class="s1">&#39;log2fc&#39;</span><span class="p">,</span>
    <span class="n">condition</span><span class="o">=</span><span class="s1">&#39;treatment_vs_control&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_expression_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Reliable changes (high coherence among similar genes)</span>
<span class="k">for</span> <span class="n">change</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">reliable_changes</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">change</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">change</span><span class="p">[</span><span class="s1">&#39;log2fc&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> (coh=</span><span class="si">{</span><span class="n">change</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="microbiology-module">
<h2>Microbiology Module<a class="headerlink" href="#microbiology-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>Apply spatial coherence to microbial screens and fitness assays.</p>
<section id="tnseq-analysis">
<h3>TnSeq Analysis<a class="headerlink" href="#tnseq-analysis" title="Link to this heading">#</a></h3>
<p>Identify essential genes/domains from transposon screens:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.microbio</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_tnseq_data</span><span class="p">,</span>
    <span class="n">analyze_tnseq_coherence</span><span class="p">,</span>
    <span class="n">identify_essential_domains</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load TnSeq fitness data</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_tnseq_data</span><span class="p">(</span>
    <span class="s1">&#39;tnseq_fitness.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;insertion_site&#39;</span><span class="p">,</span>
    <span class="n">fitness_col</span><span class="o">=</span><span class="s1">&#39;fitness_score&#39;</span><span class="p">,</span>
    <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;essential_gene&#39;</span><span class="p">,</span>
    <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;E. coli&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_tnseq_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find essential domains (low fitness, high coherence)</span>
<span class="n">essential</span> <span class="o">=</span> <span class="n">identify_essential_domains</span><span class="p">(</span>
    <span class="n">result</span><span class="p">,</span>
    <span class="n">fitness_threshold</span><span class="o">=-</span><span class="mf">2.0</span><span class="p">,</span>  <span class="c1"># Log2 fitness</span>
    <span class="n">coherence_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="n">essential</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Essential domain: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean fitness: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;mean_fitness&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="bacterial-crispri-screens">
<h3>Bacterial CRISPRi Screens<a class="headerlink" href="#bacterial-crispri-screens" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.microbio</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_crispri_screen</span><span class="p">,</span> <span class="n">analyze_crispri_coherence</span>

<span class="c1"># Load bacterial CRISPRi screen</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_crispri_screen</span><span class="p">(</span>
    <span class="s1">&#39;crispri_screen.tsv&#39;</span><span class="p">,</span>
    <span class="n">gene</span><span class="o">=</span><span class="s1">&#39;dnaA&#39;</span><span class="p">,</span>
    <span class="n">organism</span><span class="o">=</span><span class="s1">&#39;E. coli&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_crispri_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Rank guides by coherence-weighted score</span>
<span class="k">for</span> <span class="n">guide</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">ranked_guides</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">guide</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: score=</span><span class="si">{</span><span class="n">guide</span><span class="p">[</span><span class="s1">&#39;score&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, coh=</span><span class="si">{</span><span class="n">guide</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="drug-response-landscapes">
<h3>Drug Response Landscapes<a class="headerlink" href="#drug-response-landscapes" title="Link to this heading">#</a></h3>
<p>Identify stable dosing windows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.microbio</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_dose_response</span><span class="p">,</span>
    <span class="n">analyze_drug_coherence</span><span class="p">,</span>
    <span class="n">identify_stable_dosing_window</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load dose-response data</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_dose_response</span><span class="p">(</span>
    <span class="s1">&#39;dose_response.tsv&#39;</span><span class="p">,</span>
    <span class="n">concentration_col</span><span class="o">=</span><span class="s1">&#39;concentration&#39;</span><span class="p">,</span>
    <span class="n">response_col</span><span class="o">=</span><span class="s1">&#39;viability&#39;</span><span class="p">,</span>
    <span class="n">drug_name</span><span class="o">=</span><span class="s1">&#39;Compound_X&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_drug_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find stable dosing window (reliable response)</span>
<span class="n">window</span> <span class="o">=</span> <span class="n">identify_stable_dosing_window</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable dosing window: </span><span class="si">{</span><span class="n">window</span><span class="p">[</span><span class="s1">&#39;min_dose&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">window</span><span class="p">[</span><span class="s1">&#39;max_dose&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected response: </span><span class="si">{</span><span class="n">window</span><span class="p">[</span><span class="s1">&#39;mean_response&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> ± </span><span class="si">{</span><span class="n">window</span><span class="p">[</span><span class="s1">&#39;response_std&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="chemistry-module">
<h2>Chemistry Module<a class="headerlink" href="#chemistry-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>Apply spatial coherence to chemical/biochemical systems.</p>
<section id="binding-landscapes">
<h3>Binding Landscapes<a class="headerlink" href="#binding-landscapes" title="Link to this heading">#</a></h3>
<p>Analyze protein-ligand or protein-protein binding:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.chem</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_binding_data</span><span class="p">,</span>
    <span class="n">analyze_binding_coherence</span><span class="p">,</span>
    <span class="n">hot_spot_analysis</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load binding data (position = residue, response = ΔΔG)</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_binding_data</span><span class="p">(</span>
    <span class="s1">&#39;alanine_scan.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;residue&#39;</span><span class="p">,</span>
    <span class="n">binding_col</span><span class="o">=</span><span class="s1">&#39;ddG&#39;</span><span class="p">,</span>
    <span class="n">protein</span><span class="o">=</span><span class="s1">&#39;antibody_CDR&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_binding_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find stable binding hot spots</span>
<span class="n">hot_spots</span> <span class="o">=</span> <span class="n">hot_spot_analysis</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="k">for</span> <span class="n">spot</span> <span class="ow">in</span> <span class="n">hot_spots</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Hot spot: residues </span><span class="si">{</span><span class="n">spot</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">spot</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean ΔΔG: </span><span class="si">{</span><span class="n">spot</span><span class="p">[</span><span class="s1">&#39;mean_ddG&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> kcal/mol&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">spot</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reaction-optimization">
<h3>Reaction Optimization<a class="headerlink" href="#reaction-optimization" title="Link to this heading">#</a></h3>
<p>Find stable reaction conditions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.chem</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_reaction_data</span><span class="p">,</span>
    <span class="n">analyze_reaction_coherence</span><span class="p">,</span>
    <span class="n">identify_stable_conditions</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load reaction optimization data</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_reaction_data</span><span class="p">(</span>
    <span class="s1">&#39;reaction_screen.tsv&#39;</span><span class="p">,</span>
    <span class="n">condition_col</span><span class="o">=</span><span class="s1">&#39;temperature&#39;</span><span class="p">,</span>  <span class="c1"># or pH, concentration, etc.</span>
    <span class="n">yield_col</span><span class="o">=</span><span class="s1">&#39;yield&#39;</span><span class="p">,</span>
    <span class="n">reaction</span><span class="o">=</span><span class="s1">&#39;Suzuki_coupling&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_reaction_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find stable operating window</span>
<span class="n">stable</span> <span class="o">=</span> <span class="n">identify_stable_conditions</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Optimal temperature range: </span><span class="si">{</span><span class="n">stable</span><span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">°C - </span><span class="si">{</span><span class="n">stable</span><span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">°C&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected yield: </span><span class="si">{</span><span class="n">stable</span><span class="p">[</span><span class="s1">&#39;mean_yield&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">% ± </span><span class="si">{</span><span class="n">stable</span><span class="p">[</span><span class="s1">&#39;yield_std&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="hts-screening">
<h3>HTS Screening<a class="headerlink" href="#hts-screening" title="Link to this heading">#</a></h3>
<p>Identify reliable hits from high-throughput screens:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.chem</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_screening_data</span><span class="p">,</span>
    <span class="n">analyze_screening_coherence</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load HTS data</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_screening_data</span><span class="p">(</span>
    <span class="s1">&#39;hts_plate.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;well&#39;</span><span class="p">,</span>
    <span class="n">activity_col</span><span class="o">=</span><span class="s1">&#39;inhibition&#39;</span><span class="p">,</span>
    <span class="n">assay_name</span><span class="o">=</span><span class="s1">&#39;kinase_inhibition&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_screening_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Reliable hits (active + in coherent region)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total hits: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">reliable_hits</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Hit rate: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">hit_rate</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">reliable_hits</span><span class="p">[:</span><span class="mi">10</span><span class="p">]:</span>
    <span class="n">compound</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;compound_id&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;well_</span><span class="si">{</span><span class="n">hit</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">compound</span><span class="si">}</span><span class="s2">: activity=</span><span class="si">{</span><span class="n">hit</span><span class="p">[</span><span class="s1">&#39;activity&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%, coh=</span><span class="si">{</span><span class="n">hit</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="protein-module">
<h2>Protein Module<a class="headerlink" href="#protein-module" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>Analyze mutational scanning and protein fitness landscapes.</p>
<section id="mutational-scanning">
<h3>Mutational Scanning<a class="headerlink" href="#mutational-scanning" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.protein</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">load_mutscan_data</span><span class="p">,</span>
    <span class="n">analyze_mutscan_coherence</span><span class="p">,</span>
    <span class="n">local_coherence_profile</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Load deep mutational scan data</span>
<span class="n">landscape</span> <span class="o">=</span> <span class="n">load_mutscan_data</span><span class="p">(</span>
    <span class="s1">&#39;dms_data.tsv&#39;</span><span class="p">,</span>
    <span class="n">position_col</span><span class="o">=</span><span class="s1">&#39;position&#39;</span><span class="p">,</span>
    <span class="n">fitness_col</span><span class="o">=</span><span class="s1">&#39;fitness&#39;</span><span class="p">,</span>
    <span class="n">protein</span><span class="o">=</span><span class="s1">&#39;GFP&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Analyze coherence</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">analyze_mutscan_coherence</span><span class="p">(</span><span class="n">landscape</span><span class="p">)</span>

<span class="c1"># Find functional domains (conserved, high coherence)</span>
<span class="k">for</span> <span class="n">domain</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">functional_domains</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Functional domain: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Mean fitness effect: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;mean_fitness&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Conservation: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;conservation&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Coherence: </span><span class="si">{</span><span class="n">domain</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Local coherence profile for structure mapping</span>
<span class="n">profile</span> <span class="o">=</span> <span class="n">local_coherence_profile</span><span class="p">(</span><span class="n">landscape</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="structure-mapping">
<h3>Structure Mapping<a class="headerlink" href="#structure-mapping" title="Link to this heading">#</a></h3>
<p>Map coherence to 3D structure:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.protein</span><span class="w"> </span><span class="kn">import</span> <span class="n">map_coherence_to_structure</span>

<span class="c1"># Map coherence values to B-factor column in PDB</span>
<span class="n">map_coherence_to_structure</span><span class="p">(</span>
    <span class="n">result</span><span class="p">,</span>
    <span class="n">pdb_file</span><span class="o">=</span><span class="s1">&#39;protein.pdb&#39;</span><span class="p">,</span>
    <span class="n">output_file</span><span class="o">=</span><span class="s1">&#39;protein_coherence.pdb&#39;</span><span class="p">,</span>
    <span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Visualize in PyMOL:</span>
<span class="c1"># spectrum b, blue_white_red</span>
<span class="c1"># Low coherence (variable) = blue</span>
<span class="c1"># High coherence (stable) = red</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="quantum-mode-configuration">
<h2>Quantum Mode Configuration<a class="headerlink" href="#quantum-mode-configuration" title="Link to this heading">#</a></h2>
<p><em>Added in v1.0.0</em></p>
<p>Configure how PhaseLab uses quantum computation for coherence validation.</p>
<section id="mode-options">
<h3>Mode Options<a class="headerlink" href="#mode-options" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.quantum</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">QuantumMode</span><span class="p">,</span>
    <span class="n">set_quantum_mode</span><span class="p">,</span>
    <span class="n">get_quantum_mode</span><span class="p">,</span>
    <span class="n">quantum_status</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Available modes</span>
<span class="c1"># QuantumMode.OFF      - Classical only (fastest, default)</span>
<span class="c1"># QuantumMode.AUDIT    - Classical + quantum validation on subset</span>
<span class="c1"># QuantumMode.REQUIRED - Quantum mandatory for all coherence</span>

<span class="c1"># Set mode</span>
<span class="n">set_quantum_mode</span><span class="p">(</span><span class="n">QuantumMode</span><span class="o">.</span><span class="n">OFF</span><span class="p">)</span>      <span class="c1"># Default, fastest</span>
<span class="n">set_quantum_mode</span><span class="p">(</span><span class="n">QuantumMode</span><span class="o">.</span><span class="n">AUDIT</span><span class="p">)</span>    <span class="c1"># Validate with quantum</span>
<span class="n">set_quantum_mode</span><span class="p">(</span><span class="n">QuantumMode</span><span class="o">.</span><span class="n">REQUIRED</span><span class="p">)</span> <span class="c1"># Full quantum (slowest)</span>

<span class="c1"># Check current mode</span>
<span class="n">mode</span> <span class="o">=</span> <span class="n">get_quantum_mode</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current mode: </span><span class="si">{</span><span class="n">mode</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Get detailed status</span>
<span class="n">status</span> <span class="o">=</span> <span class="n">quantum_status</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">status</span><span class="p">)</span>
<span class="c1"># Quantum Mode: AUDIT</span>
<span class="c1"># ATLAS-Q Available: Yes</span>
<span class="c1"># IBM Quantum Token: Configured</span>
<span class="c1"># Default Backend: ibm_torino</span>
</pre></div>
</div>
</section>
<section id="full-configuration">
<h3>Full Configuration<a class="headerlink" href="#full-configuration" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.quantum</span><span class="w"> </span><span class="kn">import</span> <span class="n">configure_quantum</span><span class="p">,</span> <span class="n">QuantumConfig</span>

<span class="c1"># Full configuration</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">QuantumConfig</span><span class="p">(</span>
    <span class="n">mode</span><span class="o">=</span><span class="n">QuantumMode</span><span class="o">.</span><span class="n">AUDIT</span><span class="p">,</span>
    <span class="n">audit_fraction</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>        <span class="c1"># Validate 10% of calculations</span>
    <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;ibm_torino&#39;</span><span class="p">,</span>      <span class="c1"># IBM Quantum backend</span>
    <span class="n">shots</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>                <span class="c1"># Measurements per circuit</span>
    <span class="n">use_gpu</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>              <span class="c1"># GPU acceleration if available</span>
    <span class="n">cache_results</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>        <span class="c1"># Cache quantum results</span>
<span class="p">)</span>

<span class="n">configure_quantum</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="when-to-use-quantum">
<h3>When to Use Quantum<a class="headerlink" href="#when-to-use-quantum" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Mode</p></th>
<th class="head"><p>Use Case</p></th>
<th class="head"><p>Speed</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>OFF</strong></p></td>
<td><p>Development, exploration, large screens</p></td>
<td><p>Fastest</p></td>
</tr>
<tr class="row-odd"><td><p><strong>AUDIT</strong></p></td>
<td><p>Publication, validation, spot-checking</p></td>
<td><p>Medium</p></td>
</tr>
<tr class="row-even"><td><p><strong>REQUIRED</strong></p></td>
<td><p>High-stakes decisions, regulatory</p></td>
<td><p>Slowest</p></td>
</tr>
</tbody>
</table>
</div>
<p>Most users should use <strong>OFF</strong> for exploration and <strong>AUDIT</strong> for final analysis. <strong>REQUIRED</strong> mode is only needed when you need quantum-validated coherence for every calculation.</p>
</section>
</section>
<hr class="docutils" />
<section id="circadian-module">
<h2>Circadian Module<a class="headerlink" href="#circadian-module" title="Link to this heading">#</a></h2>
<p>The circadian module models biological clock systems for Smith-Magenis Syndrome research.</p>
<section id="sms-clock-model">
<h3>SMS Clock Model<a class="headerlink" href="#sms-clock-model" title="Link to this heading">#</a></h3>
<p>Simulate the SMS circadian clock with RAI1 haploinsufficiency:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">simulate_sms_clock</span><span class="p">,</span> <span class="n">SMSClockParams</span>

<span class="c1"># Basic simulation with 50% RAI1 (SMS condition)</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Period: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;period&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hours&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Amplitude: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;amplitude&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Phase shift: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;phase_shift&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> hours&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Coherence R̄: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Status: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Output:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Period</span><span class="p">:</span> <span class="mf">23.45</span> <span class="n">hours</span>
<span class="n">Amplitude</span><span class="p">:</span> <span class="mf">0.6234</span>
<span class="n">Phase</span> <span class="n">shift</span><span class="p">:</span> <span class="mf">2.15</span> <span class="n">hours</span>
<span class="n">Coherence</span> <span class="n">R̄</span><span class="p">:</span> <span class="mf">0.7823</span>
<span class="n">Status</span><span class="p">:</span> <span class="n">GO</span>
</pre></div>
</div>
<p><strong>Custom parameters:</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">simulate_sms_clock</span><span class="p">,</span> <span class="n">SMSClockParams</span>

<span class="c1"># Custom clock parameters</span>
<span class="n">params</span> <span class="o">=</span> <span class="n">SMSClockParams</span><span class="p">(</span>
    <span class="n">tau_P</span><span class="o">=</span><span class="mf">4.0</span><span class="p">,</span>      <span class="c1"># PER delay time constant (hours)</span>
    <span class="n">alpha_P</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>    <span class="c1"># PER suppression strength</span>
    <span class="n">beta_R</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>     <span class="c1"># RORα effect on BMAL1</span>
    <span class="n">beta_V</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>     <span class="c1"># REV-ERBα effect on BMAL1</span>
    <span class="n">K_light</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>    <span class="c1"># Light sensitivity</span>
    <span class="n">omega_0</span><span class="o">=</span><span class="mi">2</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">24</span>  <span class="c1"># Base angular frequency</span>
<span class="p">)</span>

<span class="c1"># Simulate with intervention (CRISPRa restoring RAI1)</span>
<span class="n">results_baseline</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>
<span class="n">results_treated</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="mf">0.85</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Baseline coherence: </span><span class="si">{</span><span class="n">results_baseline</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Treated coherence: </span><span class="si">{</span><span class="n">results_treated</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Full time series output:</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span>
    <span class="n">rai1_level</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">t_end</span><span class="o">=</span><span class="mf">240.0</span><span class="p">,</span>      <span class="c1"># 10 days</span>
    <span class="n">dt</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>           <span class="c1"># Time step (hours)</span>
    <span class="n">return_timeseries</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access time series</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;time&#39;</span><span class="p">]</span>
<span class="n">bmal1</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;BMAL1&#39;</span><span class="p">]</span>
<span class="n">per</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;PER&#39;</span><span class="p">]</span>
<span class="n">rev_erb</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;REV_ERB&#39;</span><span class="p">]</span>
<span class="n">ror</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;ROR&#39;</span><span class="p">]</span>

<span class="c1"># Plot</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">bmal1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;BMAL1&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">per</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;PER&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Time (hours)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Expression level&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="kuramoto-oscillators">
<h3>Kuramoto Oscillators<a class="headerlink" href="#kuramoto-oscillators" title="Link to this heading">#</a></h3>
<p>Model coupled oscillator networks:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">KuramotoNetwork</span>

<span class="c1"># Create network of 100 oscillators</span>
<span class="n">network</span> <span class="o">=</span> <span class="n">KuramotoNetwork</span><span class="p">(</span>
    <span class="n">n_oscillators</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">coupling_strength</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">natural_frequency_spread</span><span class="o">=</span><span class="mf">0.1</span>
<span class="p">)</span>

<span class="c1"># Simulate</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">network</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">t_end</span><span class="o">=</span><span class="mf">100.0</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Order parameter R: </span><span class="si">{</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;order_parameter&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Synchronization: </span><span class="si">{</span><span class="s1">&#39;Yes&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;synchronized&#39;</span><span class="p">]</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;No&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Heterogeneous network:</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">KuramotoNetwork</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># Custom frequency distribution (bimodal)</span>
<span class="n">frequencies</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mi">50</span><span class="p">),</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mf">1.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>
<span class="p">])</span>

<span class="n">network</span> <span class="o">=</span> <span class="n">KuramotoNetwork</span><span class="p">(</span>
    <span class="n">n_oscillators</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">coupling_strength</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">natural_frequencies</span><span class="o">=</span><span class="n">frequencies</span>
<span class="p">)</span>

<span class="c1"># With external forcing (light-dark cycle)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">light_forcing</span><span class="p">(</span><span class="n">t</span><span class="p">):</span>
    <span class="k">return</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span> <span class="o">*</span> <span class="n">t</span> <span class="o">/</span> <span class="mi">24</span><span class="p">)</span>

<span class="n">results</span> <span class="o">=</span> <span class="n">network</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
    <span class="n">t_end</span><span class="o">=</span><span class="mf">240.0</span><span class="p">,</span>
    <span class="n">external_forcing</span><span class="o">=</span><span class="n">light_forcing</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="sms-trials-module">
<h2>SMS Trials Module<a class="headerlink" href="#sms-trials-module" title="Link to this heading">#</a></h2>
<p><em>Added in v0.9.0</em></p>
<p>The SMS Trials module provides a complete therapeutic trial framework for Smith-Magenis Syndrome gene therapy development.</p>
<section id="trial-runners">
<h3>Trial Runners<a class="headerlink" href="#trial-runners" title="Link to this heading">#</a></h3>
<p>Individual trial runners for each therapeutic modality:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.trials.sms</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">run_sms_crispra_trial</span><span class="p">,</span>
    <span class="n">run_sms_crispri_trial</span><span class="p">,</span>
    <span class="n">run_sms_knockout_trial</span><span class="p">,</span>
    <span class="n">run_sms_base_editing_trial</span><span class="p">,</span>
    <span class="n">run_sms_prime_editing_trial</span><span class="p">,</span>
    <span class="n">run_circadian_rescue_simulation</span><span class="p">,</span>
    <span class="n">run_delivery_assessment</span><span class="p">,</span>
    <span class="n">SMSTrialConfig</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Configure trials</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">SMSTrialConfig</span><span class="p">(</span>
    <span class="n">therapeutic_window</span><span class="o">=</span><span class="p">(</span><span class="mf">0.70</span><span class="p">,</span> <span class="mf">1.10</span><span class="p">),</span>  <span class="c1"># 70-110% of normal RAI1</span>
    <span class="n">optimal_expression</span><span class="o">=</span><span class="mf">0.80</span><span class="p">,</span>          <span class="c1"># Target 80%</span>
    <span class="n">baseline_expression</span><span class="o">=</span><span class="mf">0.50</span><span class="p">,</span>         <span class="c1"># SMS baseline is ~50%</span>
    <span class="n">use_virtual_assay</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>           <span class="c1"># Use enhanced pipeline</span>
    <span class="n">coherence_mode</span><span class="o">=</span><span class="s2">&quot;heuristic&quot;</span><span class="p">,</span>       <span class="c1"># Fast mode</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># CRISPRa trial for RAI1 activation</span>
<span class="n">crispra_result</span> <span class="o">=</span> <span class="n">run_sms_crispra_trial</span><span class="p">(</span>
    <span class="n">promoter_sequence</span><span class="o">=</span><span class="n">rai1_promoter</span><span class="p">,</span>  <span class="c1"># Your promoter sequence</span>
    <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Status: </span><span class="si">{</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Candidates: </span><span class="si">{</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">n_candidates</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Claim level: </span><span class="si">{</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">claim_level</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">crispra_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Best guide: </span><span class="si">{</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<section id="crispri-modifier-suppression">
<h4>CRISPRi Modifier Suppression<a class="headerlink" href="#crispri-modifier-suppression" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Target circadian modifier genes (PER1, PER2, CRY1)</span>
<span class="n">crispri_result</span> <span class="o">=</span> <span class="n">run_sms_crispri_trial</span><span class="p">(</span>
    <span class="n">target_gene</span><span class="o">=</span><span class="s2">&quot;PER1&quot;</span><span class="p">,</span>
    <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Check suppression levels (target: 30-60% suppression)</span>
<span class="k">if</span> <span class="n">crispri_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">:</span>
    <span class="n">supp</span> <span class="o">=</span> <span class="n">crispri_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">[</span><span class="s1">&#39;expected_suppression&#39;</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected suppression: </span><span class="si">{</span><span class="n">supp</span><span class="si">:</span><span class="s2">.0%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="circadian-rescue-simulation">
<h4>Circadian Rescue Simulation<a class="headerlink" href="#circadian-rescue-simulation" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Predict circadian rescue from RAI1 boost</span>
<span class="n">circadian_result</span> <span class="o">=</span> <span class="n">run_circadian_rescue_simulation</span><span class="p">(</span>
    <span class="n">predicted_rai1_expression</span><span class="o">=</span><span class="mf">0.80</span><span class="p">,</span>  <span class="c1"># Expected RAI1 level after treatment</span>
    <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Rescue status: </span><span class="si">{</span><span class="n">circadian_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;rescue_status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Final R̄: </span><span class="si">{</span><span class="n">circadian_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;final_R_bar&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Sleep quality: </span><span class="si">{</span><span class="n">circadian_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;sleep_quality_prediction&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="delivery-assessment">
<h4>Delivery Assessment<a class="headerlink" href="#delivery-assessment" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assess AAV delivery feasibility for CNS</span>
<span class="n">delivery_result</span> <span class="o">=</span> <span class="n">run_delivery_assessment</span><span class="p">(</span>
    <span class="n">modality</span><span class="o">=</span><span class="s2">&quot;CRISPRa_VP64&quot;</span><span class="p">,</span>
    <span class="n">target_tissue</span><span class="o">=</span><span class="s2">&quot;brain&quot;</span><span class="p">,</span>
    <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feasibility: </span><span class="si">{</span><span class="n">delivery_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;delivery_feasibility&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Payload size: </span><span class="si">{</span><span class="n">delivery_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;payload_size&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">bp&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Recommended serotype: </span><span class="si">{</span><span class="n">delivery_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">[</span><span class="s1">&#39;recommended_serotype&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Capacity utilization: </span><span class="si">{</span><span class="n">delivery_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;packaging_utilization&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.0%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="sms-pipeline">
<h3>SMS Pipeline<a class="headerlink" href="#sms-pipeline" title="Link to this heading">#</a></h3>
<p>The SMSPipeline orchestrates all trials and provides integrated GO/NO-GO decisions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.trials.sms</span><span class="w"> </span><span class="kn">import</span> <span class="n">SMSPipeline</span><span class="p">,</span> <span class="n">SMSTrialConfig</span>

<span class="c1"># Configure pipeline</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">SMSTrialConfig</span><span class="p">(</span>
    <span class="n">therapeutic_window</span><span class="o">=</span><span class="p">(</span><span class="mf">0.70</span><span class="p">,</span> <span class="mf">1.10</span><span class="p">),</span>
    <span class="n">use_virtual_assay</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">simulation_hours</span><span class="o">=</span><span class="mf">120.0</span><span class="p">,</span>
    <span class="n">n_circadian_trials</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Create and run pipeline</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">SMSPipeline</span><span class="p">(</span><span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">)</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">run_full_pipeline</span><span class="p">(</span>
    <span class="n">include_modifiers</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>   <span class="c1"># Include CRISPRi trials</span>
    <span class="n">include_editing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>     <span class="c1"># Include base/prime editing</span>
<span class="p">)</span>

<span class="c1"># Overall decision</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overall GO/NO-GO: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">overall_go_nogo</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Claim level: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">overall_claim_level</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Individual results</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CRISPRa candidates: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">n_candidates</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Circadian rescue: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">circadian_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;rescue_status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Delivery feasible: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">delivery_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;delivery_feasibility&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Wet lab recommendations</span>
<span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">wet_lab_recommendations</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- </span><span class="si">{</span><span class="n">rec</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="falsification-tests">
<h3>Falsification Tests<a class="headerlink" href="#falsification-tests" title="Link to this heading">#</a></h3>
<p>The pipeline automatically generates falsification tests for wet-lab validation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get falsification tests</span>
<span class="k">for</span> <span class="n">test</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">falsification_tests</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Test </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Description: </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;description&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Failure condition: </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;failure_condition&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Required for: </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;required_for&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Available Tests:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Test</p></th>
<th class="head"><p>Name</p></th>
<th class="head"><p>Purpose</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>A</strong></p></td>
<td><p>Ranking Validity</p></td>
<td><p>PhaseLab-ranked guides must outperform random CRISPOR-acceptable controls</p></td>
</tr>
<tr class="row-odd"><td><p><strong>B</strong></p></td>
<td><p>Risk Prediction</p></td>
<td><p>High-risk (CAUTION) guides should fail safety screen more often</p></td>
</tr>
<tr class="row-even"><td><p><strong>C</strong></p></td>
<td><p>Dosage Prediction</p></td>
<td><p>Predicted expression levels should correlate (r &gt; 0.6) with observed</p></td>
</tr>
<tr class="row-odd"><td><p><strong>D</strong></p></td>
<td><p>UNKNOWN Bucket</p></td>
<td><p>UNKNOWN-labeled guides should fail at approximately random rate</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="trial-result-structure">
<h3>Trial Result Structure<a class="headerlink" href="#trial-result-structure" title="Link to this heading">#</a></h3>
<p>All trial runners return <code class="docutils literal notranslate"><span class="pre">SMSTrialResult</span></code> with consistent structure:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">SMSTrialResult</span><span class="p">:</span>
    <span class="n">trial_type</span><span class="p">:</span> <span class="n">TrialType</span>          <span class="c1"># Type of trial (CRISPRA_RAI1, CRISPRI_MODIFIER, etc.)</span>
    <span class="n">status</span><span class="p">:</span> <span class="n">TrialStatus</span>            <span class="c1"># COMPLETED, FAILED, PENDING</span>
    <span class="n">summary</span><span class="p">:</span> <span class="nb">str</span>                   <span class="c1"># Human-readable summary</span>
    <span class="n">candidates</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span>         <span class="c1"># Ranked candidates</span>
    <span class="n">best_candidate</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="c1"># Top candidate</span>
    <span class="n">metrics</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span>        <span class="c1"># Trial-specific metrics</span>
    <span class="n">claim_level</span><span class="p">:</span> <span class="nb">str</span>               <span class="c1"># STRONG_COMPUTATIONAL, CONTEXT_DEPENDENT, etc.</span>
    <span class="n">claim_description</span><span class="p">:</span> <span class="nb">str</span>         <span class="c1"># Explanation of claim level</span>
    <span class="n">warnings</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>            <span class="c1"># Important warnings</span>
    <span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>              <span class="c1"># Any errors encountered</span>

    <span class="c1"># Helper properties</span>
    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">is_successful</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span> <span class="o">...</span>
    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">has_viable_candidates</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span> <span class="o">...</span>
    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">n_candidates</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span> <span class="o">...</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="quantum-discriminator">
<h2>Quantum Discriminator<a class="headerlink" href="#quantum-discriminator" title="Link to this heading">#</a></h2>
<p><em>Added in v0.9.5</em></p>
<p>The Quantum Discriminator is a late-stage guide selection tool that uses quantum chemistry on IBM Quantum hardware to discriminate between elite CRISPRa guides whose predicted effectiveness is classically indistinguishable.</p>
<p><strong>Scientific Claim (defensible):</strong></p>
<blockquote>
<div><p>“IR-enhanced quantum VQE on current IBM hardware can resolve binding energy differences between CRISPRa guides that are indistinguishable under classical scoring, providing a physically grounded late-stage discriminator for therapeutic guide selection.”</p>
</div></blockquote>
<p><strong>What this IS:</strong></p>
<ul class="simple">
<li><p>Quantum resolves energetic degeneracy</p></li>
<li><p>Quantum increases hit rate per wet-lab experiment</p></li>
<li><p>Quantum reduces wasted biological trials</p></li>
</ul>
<p><strong>What this is NOT:</strong></p>
<ul class="simple">
<li><p>Not “quantum finds cures”</p></li>
<li><p>Not “quantum replaces biology”</p></li>
<li><p>Not “quantum predicts expression directly”</p></li>
</ul>
<section id="when-to-use-quantum-discrimination">
<h3>When to Use Quantum Discrimination<a class="headerlink" href="#when-to-use-quantum-discrimination" title="Link to this heading">#</a></h3>
<p>Use the quantum discriminator when:</p>
<ol class="arabic simple">
<li><p><strong>Classical methods saturate</strong>: Top guides have combined scores within 5% of each other</p></li>
<li><p><strong>Elite tier only</strong>: Only for guides that pass all classical gates</p></li>
<li><p><strong>Late-stage</strong>: After multi-evidence pipeline (binding, phase, geometry) has ranked candidates</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">DISCRIMINATOR_GATES</span>

<span class="c1"># Pre-quantum gates that must be passed:</span>
<span class="nb">print</span><span class="p">(</span><span class="n">DISCRIMINATOR_GATES</span><span class="p">)</span>
<span class="c1"># {</span>
<span class="c1">#     &#39;min_mit_score&#39;: 50,</span>
<span class="c1">#     &#39;max_exonic_ot&#39;: 0,</span>
<span class="c1">#     &#39;min_delta_r&#39;: 0.30,</span>
<span class="c1">#     &#39;min_phase_coherence&#39;: 0.90,</span>
<span class="c1">#     &#39;min_guides_for_quantum&#39;: 2,</span>
<span class="c1"># }</span>
</pre></div>
</div>
</section>
<section id="effective-binding-hamiltonian">
<h3>Effective Binding Hamiltonian<a class="headerlink" href="#effective-binding-hamiltonian" title="Link to this heading">#</a></h3>
<p>The discriminator constructs an effective Hamiltonian for RNA-DNA binding:</p>
<div class="math notranslate nohighlight">
\[H = H_{HB} + H_{stack} + H_{charge} + H_{constraint}\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><strong>H_HB</strong>: Watson-Crick hydrogen bonding (G-C: -0.18 eV, A-T: -0.12 eV)</p></li>
<li><p><strong>H_stack</strong>: π-π stacking stabilization between adjacent bases</p></li>
<li><p><strong>H_charge</strong>: Backbone electrostatic interactions with 0.7 screening</p></li>
<li><p><strong>H_constraint</strong>: Prevents unphysical strand separation</p></li>
</ul>
<p>The Hamiltonian uses the seed region (12bp PAM-proximal) encoded as Pauli operators:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr.quantum_discriminator</span><span class="w"> </span><span class="kn">import</span> <span class="n">_build_effective_binding_hamiltonian</span>

<span class="n">guide</span> <span class="o">=</span> <span class="s2">&quot;GCGCGCGCGCGCGCGCGCGC&quot;</span>
<span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;CGCGCGCGCGCGCGCGCGCG&quot;</span>

<span class="n">coefficients</span><span class="p">,</span> <span class="n">paulis</span><span class="p">,</span> <span class="n">n_qubits</span> <span class="o">=</span> <span class="n">_build_effective_binding_hamiltonian</span><span class="p">(</span><span class="n">guide</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Qubits: </span><span class="si">{</span><span class="n">n_qubits</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>        <span class="c1"># 12</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pauli terms: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">paulis</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="c1"># ~44</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;First term: </span><span class="si">{</span><span class="n">coefficients</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2"> * </span><span class="si">{</span><span class="n">paulis</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># -0.1800 * ZIIIIIIIIIII</span>
</pre></div>
</div>
</section>
<section id="running-the-discriminator">
<h3>Running the Discriminator<a class="headerlink" href="#running-the-discriminator" title="Link to this heading">#</a></h3>
<section id="basic-usage-simulation">
<h4>Basic Usage (Simulation)<a class="headerlink" href="#basic-usage-simulation" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">run_quantum_discriminator</span><span class="p">,</span> <span class="n">DiscriminatorStatus</span>

<span class="c1"># Elite guides from classical pipeline (scores are degenerate)</span>
<span class="n">guides</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">{</span><span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="s1">&#39;GCGCGCGCGCGCGCGCGCGC&#39;</span><span class="p">,</span> <span class="s1">&#39;combined_score&#39;</span><span class="p">:</span> <span class="mf">0.92</span><span class="p">},</span>
    <span class="p">{</span><span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="s1">&#39;ATCGATCGATCGATCGATCG&#39;</span><span class="p">,</span> <span class="s1">&#39;combined_score&#39;</span><span class="p">:</span> <span class="mf">0.91</span><span class="p">},</span>
    <span class="p">{</span><span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="s1">&#39;CCGGCCGGCCGGCCGGCCGG&#39;</span><span class="p">,</span> <span class="s1">&#39;combined_score&#39;</span><span class="p">:</span> <span class="mf">0.90</span><span class="p">},</span>
<span class="p">]</span>

<span class="c1"># DNA target context</span>
<span class="n">dna_context</span> <span class="o">=</span> <span class="s2">&quot;CGCGCGCGCGCGCGCGCGCG...&quot;</span>  <span class="c1"># Promoter region</span>

<span class="c1"># Run discriminator (simulation mode)</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">run_quantum_discriminator</span><span class="p">(</span>
    <span class="n">guides</span><span class="o">=</span><span class="n">guides</span><span class="p">,</span>
    <span class="n">dna_context</span><span class="o">=</span><span class="n">dna_context</span><span class="p">,</span>
    <span class="n">backend_name</span><span class="o">=</span><span class="s2">&quot;ibm_torino&quot;</span><span class="p">,</span>
    <span class="n">use_hardware</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Simulation</span>
    <span class="n">shots</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">max_iterations</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">degeneracy_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="o">.</span><span class="n">summary</span><span class="p">())</span>
<span class="c1"># ======================================================================</span>
<span class="c1"># QUANTUM DISCRIMINATOR RESULT</span>
<span class="c1"># ======================================================================</span>
<span class="c1"># Status: quantum_success</span>
<span class="c1"># Guides evaluated: 3</span>
<span class="c1"># Quantum advantage: YES</span>
<span class="c1">#</span>
<span class="c1"># Ranking by quantum binding energy:</span>
<span class="c1">#   #1: CCGGCCGGCCGGCCG... E=-1.993470 Ha [GO]</span>
<span class="c1">#   #2: ATCGATCGATCGATC... E=-1.681406 Ha [GO]</span>
<span class="c1">#   #3: GCGCGCGCGCGCGCG... E=-0.556050 Ha [GO]</span>
</pre></div>
</div>
</section>
<section id="running-on-ibm-quantum-hardware">
<h4>Running on IBM Quantum Hardware<a class="headerlink" href="#running-on-ibm-quantum-hardware" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;IBM_QUANTUM_TOKEN&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;your-token-here&#39;</span>

<span class="n">result</span> <span class="o">=</span> <span class="n">run_quantum_discriminator</span><span class="p">(</span>
    <span class="n">guides</span><span class="o">=</span><span class="n">guides</span><span class="p">,</span>
    <span class="n">dna_context</span><span class="o">=</span><span class="n">dna_context</span><span class="p">,</span>
    <span class="n">backend_name</span><span class="o">=</span><span class="s2">&quot;ibm_torino&quot;</span><span class="p">,</span>
    <span class="n">use_hardware</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Real hardware</span>
    <span class="n">shots</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">max_iterations</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Check individual guide results</span>
<span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">ranked_guides</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">g</span><span class="o">.</span><span class="n">guide_sequence</span><span class="p">[:</span><span class="mi">15</span><span class="p">]</span><span class="si">}</span><span class="s2">... E=</span><span class="si">{</span><span class="n">g</span><span class="o">.</span><span class="n">binding_energy</span><span class="si">:</span><span class="s2">.6f</span><span class="si">}</span><span class="s2"> Ha, &quot;</span>
          <span class="sa">f</span><span class="s2">&quot;R̄=</span><span class="si">{</span><span class="n">g</span><span class="o">.</span><span class="n">coherence</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="s1">&#39;GO&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">g</span><span class="o">.</span><span class="n">is_go</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;NO-GO&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="understanding-results">
<h4>Understanding Results<a class="headerlink" href="#understanding-results" title="Link to this heading">#</a></h4>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Result structure</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Status: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">status</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>              <span class="c1"># DiscriminatorStatus.QUANTUM_SUCCESS</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Quantum advantage: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">quantum_advantage</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># True if ordering differs</span>

<span class="c1"># Energy separations between guide pairs</span>
<span class="k">for</span> <span class="n">pair</span><span class="p">,</span> <span class="n">delta_e</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">energy_separations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">is_sig</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">significant_separations</span><span class="p">[</span><span class="n">pair</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pair</span><span class="si">}</span><span class="s2">: ΔE = </span><span class="si">{</span><span class="n">delta_e</span><span class="si">:</span><span class="s2">.6f</span><span class="si">}</span><span class="s2"> Ha </span><span class="si">{</span><span class="s1">&#39;*&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">is_sig</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Classical vs quantum comparison</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Classical scores: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">classical_scores</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="phaselab-api-integration">
<h3>PhaseLab API Integration<a class="headerlink" href="#phaselab-api-integration" title="Link to this heading">#</a></h3>
<p>The high-level API integrates quantum discrimination into the full design pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides_with_quantum_discriminator</span>

<span class="c1"># After running classical pipeline...</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">design_guides_with_quantum_discriminator</span><span class="p">(</span>
    <span class="n">gene</span><span class="o">=</span><span class="s2">&quot;RAI1&quot;</span><span class="p">,</span>
    <span class="n">guides</span><span class="o">=</span><span class="n">top_classical_guides</span><span class="p">,</span>  <span class="c1"># From multi-evidence pipeline</span>
    <span class="n">dna_context</span><span class="o">=</span><span class="n">rai1_promoter</span><span class="p">,</span>
    <span class="n">quantum_stage</span><span class="o">=</span><span class="s2">&quot;late&quot;</span><span class="p">,</span>          <span class="c1"># Only use at end</span>
    <span class="n">quantum_backend</span><span class="o">=</span><span class="s2">&quot;ibm_torino&quot;</span><span class="p">,</span>
    <span class="n">use_hardware</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>            <span class="c1"># True for real hardware</span>
    <span class="n">max_quantum_guides</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>          <span class="c1"># Limit quantum evaluations</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Quantum advantage: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;quantum_advantage&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Recommendation: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;recommendation&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Final ranking (quantum-ordered if advantage detected)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">g</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;final_ranking&#39;</span><span class="p">],</span> <span class="mi">1</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;ranking_source&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;quantum&#39;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;#</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][:</span><span class="mi">15</span><span class="p">]</span><span class="si">}</span><span class="s2">... E=</span><span class="si">{</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;quantum_energy&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.6f</span><span class="si">}</span><span class="s2"> Ha&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;#</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">][:</span><span class="mi">15</span><span class="p">]</span><span class="si">}</span><span class="s2">... score=</span><span class="si">{</span><span class="n">g</span><span class="p">[</span><span class="s1">&#39;classical_score&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="discriminator-status-codes">
<h3>Discriminator Status Codes<a class="headerlink" href="#discriminator-status-codes" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Status</p></th>
<th class="head"><p>Meaning</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">QUANTUM_SUCCESS</span></code></p></td>
<td><p>VQE completed, results available</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">NO_DEGENERACY</span></code></p></td>
<td><p>Classical scores differ enough, quantum not needed</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">INSUFFICIENT_GUIDES</span></code></p></td>
<td><p>Need ≥2 guides to discriminate</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">QUANTUM_FAILED</span></code></p></td>
<td><p>VQE failed, fell back to classical</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="go-no-go-threshold">
<h3>GO/NO-GO Threshold<a class="headerlink" href="#go-no-go-threshold" title="Link to this heading">#</a></h3>
<p>Quantum execution quality is validated using IR coherence:</p>
<div class="math notranslate nohighlight">
\[\text{GO if } \bar{R} &gt; e^{-2} \approx 0.135\]</div>
<p>Guides with coherence below this threshold during VQE execution are marked NO-GO and excluded from final ranking.</p>
</section>
<section id="api-reference">
<h3>API Reference<a class="headerlink" href="#api-reference" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Function</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">run_quantum_discriminator(guides,</span> <span class="pre">dna_context,</span> <span class="pre">...)</span></code></p></td>
<td><p>Main discriminator entry point</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">design_guides_with_quantum_discriminator(gene,</span> <span class="pre">guides,</span> <span class="pre">...)</span></code></p></td>
<td><p>High-level PhaseLab API</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DiscriminatorStatus</span></code></p></td>
<td><p>Enum for result status</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">QuantumGuideResult</span></code></p></td>
<td><p>Single guide quantum result</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DiscriminatorResult</span></code></p></td>
<td><p>Complete discriminator result</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">DISCRIMINATOR_GATES</span></code></p></td>
<td><p>Pre-quantum gate thresholds</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<hr class="docutils" />
<section id="quantum-integration">
<h2>Quantum Integration<a class="headerlink" href="#quantum-integration" title="Link to this heading">#</a></h2>
<p>PhaseLab integrates with IBM Quantum for hardware validation.</p>
<section id="setting-up-ibm-quantum">
<h3>Setting Up IBM Quantum<a class="headerlink" href="#setting-up-ibm-quantum" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;IBM_QUANTUM_TOKEN&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;your-token-here&#39;</span>

<span class="c1"># Or use .env file</span>
<span class="c1"># IBM_QUANTUM_TOKEN=your-token-here</span>
</pre></div>
</div>
</section>
<section id="running-on-quantum-hardware">
<h3>Running on Quantum Hardware<a class="headerlink" href="#running-on-quantum-hardware" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.quantum</span><span class="w"> </span><span class="kn">import</span> <span class="n">QuantumCoherenceValidator</span>

<span class="c1"># Initialize validator</span>
<span class="n">validator</span> <span class="o">=</span> <span class="n">QuantumCoherenceValidator</span><span class="p">(</span>
    <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;ibm_torino&#39;</span><span class="p">,</span>  <span class="c1"># or &#39;ibm_brisbane&#39;, &#39;ibm_kyoto&#39;</span>
    <span class="n">shots</span><span class="o">=</span><span class="mi">4096</span>
<span class="p">)</span>

<span class="c1"># Validate a guide RNA</span>
<span class="n">guide</span> <span class="o">=</span> <span class="s2">&quot;GAAGGAGAGCAAGAGCGCGA&quot;</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">validator</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">guide</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Hardware R̄: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Status: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Job ID: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;job_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="simulator-testing">
<h3>Simulator Testing<a class="headerlink" href="#simulator-testing" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.quantum</span><span class="w"> </span><span class="kn">import</span> <span class="n">QuantumCoherenceValidator</span>

<span class="c1"># Use Aer simulator</span>
<span class="n">validator</span> <span class="o">=</span> <span class="n">QuantumCoherenceValidator</span><span class="p">(</span><span class="n">backend</span><span class="o">=</span><span class="s1">&#39;aer_simulator&#39;</span><span class="p">)</span>

<span class="n">guides</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;GAAGGAGAGCAAGAGCGCGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AACTGCAAAGAAGTGGGCAC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TACAGGAGCTTCCAGCGTCA&quot;</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">guide</span> <span class="ow">in</span> <span class="n">guides</span><span class="p">:</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">validator</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">guide</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">guide</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span><span class="si">}</span><span class="s2">... R̄=</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> [</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="complete-examples">
<h2>Complete Examples<a class="headerlink" href="#complete-examples" title="Link to this heading">#</a></h2>
<section id="example-1-sms-gene-therapy-pipeline">
<h3>Example 1: SMS Gene Therapy Pipeline<a class="headerlink" href="#example-1-sms-gene-therapy-pipeline" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Complete SMS gene therapy gRNA design pipeline.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">phaselab</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span><span class="p">,</span> <span class="n">CRISPRConfig</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">simulate_sms_clock</span>

<span class="c1"># RAI1 promoter sequence (truncated for example)</span>
<span class="n">rai1_promoter</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">GCGCGCTCGCGCGCTCGCGCGAAGGAGAGCAAGAGCGCGACGGCTAGCTAGCT</span>
<span class="s2">AGCTAGCTAGCTACAGGAGCTTCCAGCGTCAGGGCTAGCTAGCTAGCTAGCTA</span>
<span class="s2">GCTAGCTAGCTAACTGCAAAGAAGTGGGCACGCGCTAGCTAGCTAGCTAGCT</span>
<span class="s2">&quot;&quot;&quot;</span>

<span class="c1"># TSS position</span>
<span class="n">tss_index</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rai1_promoter</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>

<span class="c1"># Configure pipeline</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">CRISPRConfig</span><span class="p">(</span>
    <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">,</span>
    <span class="n">window_upstream</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span>
    <span class="n">window_downstream</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">min_gc</span><span class="o">=</span><span class="mf">0.40</span><span class="p">,</span>
    <span class="n">max_gc</span><span class="o">=</span><span class="mf">0.70</span>
<span class="p">)</span>

<span class="c1"># Design guides</span>
<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">rai1_promoter</span><span class="p">,</span> <span class="n">tss_index</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">)</span>

<span class="c1"># Filter GO candidates</span>
<span class="n">go_candidates</span> <span class="o">=</span> <span class="n">guides</span><span class="p">[</span><span class="n">guides</span><span class="p">[</span><span class="s1">&#39;go_no_go&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;GO&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">go_candidates</span><span class="p">)</span><span class="si">}</span><span class="s2"> GO candidates&quot;</span><span class="p">)</span>

<span class="c1"># Simulate therapeutic effect</span>
<span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">go_candidates</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Position: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> bp from TSS&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Score: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;combined_score&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Simulate with restored RAI1 (hypothetical 80% restoration)</span>
    <span class="n">sim</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Predicted period: </span><span class="si">{</span><span class="n">sim</span><span class="p">[</span><span class="s1">&#39;period&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">h&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Predicted coherence: </span><span class="si">{</span><span class="n">sim</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="example-2-comparative-analysis">
<h3>Example 2: Comparative Analysis<a class="headerlink" href="#example-2-comparative-analysis" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Compare coherence across multiple conditions.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">phaselab</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">simulate_sms_clock</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># RAI1 levels to test</span>
<span class="n">rai1_levels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>

<span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">level</span> <span class="ow">in</span> <span class="n">rai1_levels</span><span class="p">:</span>
    <span class="n">sim</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="n">level</span><span class="p">)</span>
    <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
        <span class="s1">&#39;rai1_level&#39;</span><span class="p">:</span> <span class="n">level</span><span class="p">,</span>
        <span class="s1">&#39;period&#39;</span><span class="p">:</span> <span class="n">sim</span><span class="p">[</span><span class="s1">&#39;period&#39;</span><span class="p">],</span>
        <span class="s1">&#39;coherence&#39;</span><span class="p">:</span> <span class="n">sim</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">],</span>
        <span class="s1">&#39;status&#39;</span><span class="p">:</span> <span class="n">sim</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span>
    <span class="p">})</span>

<span class="c1"># Find therapeutic threshold</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
<span class="n">go_threshold</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;GO&#39;</span><span class="p">][</span><span class="s1">&#39;rai1_level&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Minimum RAI1 for GO status: </span><span class="si">{</span><span class="n">go_threshold</span><span class="si">:</span><span class="s2">.1%</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Visualize</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;rai1_level&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">],</span> <span class="s1">&#39;bo-&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">E_MINUS_2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;GO threshold&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;RAI1 Level&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Coherence R̄&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;rai1_level&#39;</span><span class="p">],</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;period&#39;</span><span class="p">],</span> <span class="s1">&#39;go-&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">24</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;24h&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;RAI1 Level&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Period (hours)&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="example-3-batch-processing">
<h3>Example 3: Batch Processing<a class="headerlink" href="#example-3-batch-processing" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Batch process multiple sequences.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Multiple target genes</span>
<span class="n">targets</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;RAI1&#39;</span><span class="p">:</span> <span class="s1">&#39;ATGCGATCG...&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CLOCK&#39;</span><span class="p">:</span> <span class="s1">&#39;GCTAGCTAG...&#39;</span><span class="p">,</span>
    <span class="s1">&#39;BMAL1&#39;</span><span class="p">:</span> <span class="s1">&#39;TAGCTAGCT...&#39;</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">all_guides</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">tss_index</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">guides</span><span class="p">[</span><span class="s1">&#39;target_gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span>
    <span class="n">all_guides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">guides</span><span class="p">)</span>

<span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">all_guides</span><span class="p">)</span>
<span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;combined_score&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Top guide per gene</span>
<span class="n">top_per_gene</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;target_gene&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">top_per_gene</span><span class="p">[[</span><span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;combined_score&#39;</span><span class="p">,</span> <span class="s1">&#39;go_no_go&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="id1">
<h2>API Reference<a class="headerlink" href="#id1" title="Link to this heading">#</a></h2>
<section id="core-functions">
<h3>Core Functions<a class="headerlink" href="#core-functions" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Function</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">coherence_score(data,</span> <span class="pre">mode='auto')</span></code></p></td>
<td><p>Compute R̄ coherence score</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">phase_variance(phases)</span></code></p></td>
<td><p>Compute phase variance V_φ</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">go_no_go(R_bar,</span> <span class="pre">threshold=E_MINUS_2)</span></code></p></td>
<td><p>Binary GO/NO-GO classification</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">classify_coherence(R_bar)</span></code></p></td>
<td><p>Detailed category classification</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="crispr-functions">
<h3>CRISPR Functions<a class="headerlink" href="#crispr-functions" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Function</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">design_guides(seq,</span> <span class="pre">tss,</span> <span class="pre">config,</span> <span class="pre">dnase_peaks)</span></code></p></td>
<td><p>High-level guide design pipeline</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">find_pam_sites(seq,</span> <span class="pre">pam,</span> <span class="pre">guide_length)</span></code></p></td>
<td><p>Find PAM sites in sequence</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">score_guide(guide,</span> <span class="pre">weights)</span></code></p></td>
<td><p>Score individual guide</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">gc_content(seq)</span></code></p></td>
<td><p>Calculate GC fraction</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">santalucia_delta_g(seq)</span></code></p></td>
<td><p>Calculate thermodynamic ΔG</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="circadian-functions">
<h3>Circadian Functions<a class="headerlink" href="#circadian-functions" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Function</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">simulate_sms_clock(rai1_level,</span> <span class="pre">params,</span> <span class="pre">t_end)</span></code></p></td>
<td><p>Simulate SMS clock model</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">KuramotoNetwork(n,</span> <span class="pre">coupling,</span> <span class="pre">frequencies)</span></code></p></td>
<td><p>Create Kuramoto oscillator network</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="constants">
<h3>Constants<a class="headerlink" href="#constants" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Constant</p></th>
<th class="head"><p>Value</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">E_MINUS_2</span></code></p></td>
<td><p>0.1353…</p></td>
<td><p>e⁻² GO/NO-GO threshold</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">FOUR_PI_SQUARED</span></code></p></td>
<td><p>39.478…</p></td>
<td><p>4π² universal constant</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<hr class="docutils" />
<section id="citation">
<h2>Citation<a class="headerlink" href="#citation" title="Link to this heading">#</a></h2>
<p>If you use PhaseLab in your research, please cite:</p>
<div class="highlight-bibtex notranslate"><div class="highlight"><pre><span></span><span class="nc">@software</span><span class="p">{</span><span class="nl">phaselab</span><span class="p">,</span>
<span class="w">  </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Vaca, Dylan}</span><span class="p">,</span>
<span class="w">  </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{PhaseLab: Phase-coherence analysis framework}</span><span class="p">,</span>
<span class="w">  </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2024}</span><span class="p">,</span>
<span class="w">  </span><span class="na">url</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://github.com/followthesapper/phaselab}</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<hr class="docutils" />
<section id="license">
<h2>License<a class="headerlink" href="#license" title="Link to this heading">#</a></h2>
<p>MIT License - see <a class="reference internal" href="#LICENSE"><span class="xref myst">LICENSE</span></a> for details.</p>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-is-phaselab">What is PhaseLab?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#installation">Installation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-installation">Basic Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#with-quantum-support-ibm-quantum">With Quantum Support (IBM Quantum)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#with-plotting-support">With Plotting Support</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#full-installation">Full Installation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#development-installation">Development Installation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quick-start">Quick Start</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#core-concepts">Core Concepts</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#what-is-spatial-coherence">What is Spatial Coherence?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#stable-vs-amplifying-regions">Stable vs Amplifying Regions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-trust-predictions">When to Trust Predictions</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#landscapes-module">Landscapes Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#response-landscapes">Response Landscapes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coherence-profiles">Coherence Profiles</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#region-classification">Region Classification</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#spatial-analysis-module">Spatial Analysis Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tiling-screen-analysis">Tiling Screen Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#validated-results">Validated Results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#crispr-module">CRISPR Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#important-deprecated-features">Important: Deprecated Features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispra-guide-design">CRISPRa Guide Design</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pam-scanning">PAM Scanning</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#guide-scoring">Guide Scoring</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#design-pipeline">Design Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#integration-with-spatial-coherence">Integration with Spatial Coherence</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#crispr-surf-integration">CRISPR-SURF Integration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#what-is-crispr-surf">What is CRISPR-SURF?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#coherence-on-deconvolved-data">Coherence on Deconvolved Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#full-surf-pipeline">Full SURF Pipeline</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#omics-module">Omics Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#atac-seq-coherence">ATAC-seq Coherence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#chip-seq-coherence">ChIP-seq Coherence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#rna-seq-coherence">RNA-seq Coherence</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#microbiology-module">Microbiology Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tnseq-analysis">TnSeq Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bacterial-crispri-screens">Bacterial CRISPRi Screens</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#drug-response-landscapes">Drug Response Landscapes</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chemistry-module">Chemistry Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#binding-landscapes">Binding Landscapes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reaction-optimization">Reaction Optimization</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hts-screening">HTS Screening</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#protein-module">Protein Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mutational-scanning">Mutational Scanning</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#structure-mapping">Structure Mapping</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-mode-configuration">Quantum Mode Configuration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mode-options">Mode Options</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#full-configuration">Full Configuration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-use-quantum">When to Use Quantum</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#circadian-module">Circadian Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sms-clock-model">SMS Clock Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#kuramoto-oscillators">Kuramoto Oscillators</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sms-trials-module">SMS Trials Module</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trial-runners">Trial Runners</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#crispri-modifier-suppression">CRISPRi Modifier Suppression</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#circadian-rescue-simulation">Circadian Rescue Simulation</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#delivery-assessment">Delivery Assessment</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sms-pipeline">SMS Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#falsification-tests">Falsification Tests</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#trial-result-structure">Trial Result Structure</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-discriminator">Quantum Discriminator</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#when-to-use-quantum-discrimination">When to Use Quantum Discrimination</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#effective-binding-hamiltonian">Effective Binding Hamiltonian</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#running-the-discriminator">Running the Discriminator</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-usage-simulation">Basic Usage (Simulation)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#running-on-ibm-quantum-hardware">Running on IBM Quantum Hardware</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#understanding-results">Understanding Results</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phaselab-api-integration">PhaseLab API Integration</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#discriminator-status-codes">Discriminator Status Codes</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#go-no-go-threshold">GO/NO-GO Threshold</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#api-reference">API Reference</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-integration">Quantum Integration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-ibm-quantum">Setting Up IBM Quantum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#running-on-quantum-hardware">Running on Quantum Hardware</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simulator-testing">Simulator Testing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#complete-examples">Complete Examples</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-sms-gene-therapy-pipeline">Example 1: SMS Gene Therapy Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-comparative-analysis">Example 2: Comparative Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-3-batch-processing">Example 3: Batch Processing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">API Reference</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#core-functions">Core Functions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispr-functions">CRISPR Functions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#circadian-functions">Circadian Functions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#constants">Constants</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citation">Citation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#license">License</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Dylan Vaca.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>