
<!DOCTYPE html>


<html lang="en" data-content_root="./" data-theme="auto">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Smith-Magenis Syndrome Gene Therapy Research &#8212; PhaseLab 0.6.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "auto";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "auto";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=09465d15" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=2bd037f0"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=30646c52"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'SMS_GENE_THERAPY_RESEARCH';</script>
    <link rel="icon" href="_static/favicon.svg"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="auto">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo-lm.svg" class="logo__image only-light" alt=""/>
    <img src="_static/logo.svg" class="logo__image only-dark pst-js-only" alt=""/>
  
  
    <p class="title logo__title">PhaseLab</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/explanations/index.html">
    Explanations
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="reference/index.html">
    API Reference
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="research/index.html">
    Research Applications
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="developer/index.html">
    Developer Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="citing.html">
    Citing PhaseLab
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="user_guide/explanations/index.html">
    Explanations
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="reference/index.html">
    API Reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="research/index.html">
    Research Applications
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="developer/index.html">
    Developer Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="citing.html">
    Citing PhaseLab
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Smith-Magenis Syndrome Gene Therapy Research</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="smith-magenis-syndrome-gene-therapy-research">
<h1>Smith-Magenis Syndrome Gene Therapy Research<a class="headerlink" href="#smith-magenis-syndrome-gene-therapy-research" title="Link to this heading">#</a></h1>
<p><strong>CRISPRa Guide RNA Design Using Quantum-Validated Coherence Metrics</strong></p>
<p><em>Dylan Vaca | December 2025</em></p>
<hr class="docutils" />
<section id="abstract">
<h2>Abstract<a class="headerlink" href="#abstract" title="Link to this heading">#</a></h2>
<p>We present a novel approach to CRISPRa guide RNA design for Smith-Magenis Syndrome (SMS) gene therapy, combining quantum computing validation with the Informational Relativity (IR) coherence framework. Using IBM Quantum hardware (Torino, 133 qubits) and genome-wide off-target analysis (CRISPOR), we identified and validated optimal guide RNAs for upregulating RAI1 expression. Our top candidate (<code class="docutils literal notranslate"><span class="pre">TACAGGAGCTTCCAGCGTCA</span></code>) achieved excellent specificity scores (MIT 83, CFD 93) with zero off-targets at ≤2 mismatches, validated by quantum coherence R̄ = 0.839 on real hardware.</p>
</section>
<hr class="docutils" />
<section id="background">
<h2>1. Background<a class="headerlink" href="#background" title="Link to this heading">#</a></h2>
<section id="smith-magenis-syndrome">
<h3>1.1 Smith-Magenis Syndrome<a class="headerlink" href="#smith-magenis-syndrome" title="Link to this heading">#</a></h3>
<p>Smith-Magenis Syndrome is a rare genetic disorder affecting approximately 1 in 15,000-25,000 births, caused by haploinsufficiency of the RAI1 gene at chromosome 17p11.2.</p>
<p><strong>Clinical Features:</strong></p>
<ul class="simple">
<li><p>Intellectual disability</p></li>
<li><p>Characteristic sleep disturbance (inverted melatonin rhythm)</p></li>
<li><p>Behavioral challenges</p></li>
<li><p>Distinctive facial features</p></li>
</ul>
<p><strong>Genetic Basis:</strong>
Most SMS patients have a deletion encompassing RAI1, leaving only one functional copy. This results in ~50% of normal RAI1 protein levels.</p>
</section>
<section id="therapeutic-approach">
<h3>1.2 Therapeutic Approach<a class="headerlink" href="#therapeutic-approach" title="Link to this heading">#</a></h3>
<p><strong>CRISPRa (CRISPR Activation)</strong> offers a promising therapeutic strategy:</p>
<ul class="simple">
<li><p>Does NOT cut DNA (unlike traditional CRISPR)</p></li>
<li><p>Uses catalytically dead Cas9 (dCas9) fused to transcriptional activators (VP64)</p></li>
<li><p>Boosts expression of the remaining functional RAI1 copy</p></li>
<li><p>Goal: Restore RAI1 to 60-80% of normal diploid levels</p></li>
</ul>
</section>
<section id="the-ir-coherence-framework">
<h3>1.3 The IR Coherence Framework<a class="headerlink" href="#the-ir-coherence-framework" title="Link to this heading">#</a></h3>
<p>This research applies the <strong>Informational Relativity (IR) framework</strong> to assess simulation reliability:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Component</p></th>
<th class="head"><p>Application</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>R̄ = exp(-V_φ/2)</strong></p></td>
<td><p>Coherence metric for binding quality</p></td>
</tr>
<tr class="row-odd"><td><p><strong>e⁻² threshold</strong></p></td>
<td><p>GO/NO-GO decision boundary (≈0.135)</p></td>
</tr>
<tr class="row-even"><td><p><strong>VQE grouping</strong></p></td>
<td><p>Reduced quantum measurements</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Kuramoto dynamics</strong></p></td>
<td><p>Circadian clock modeling</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>
<hr class="docutils" />
<section id="methods">
<h2>2. Methods<a class="headerlink" href="#methods" title="Link to this heading">#</a></h2>
<section id="genomic-data">
<h3>2.1 Genomic Data<a class="headerlink" href="#genomic-data" title="Link to this heading">#</a></h3>
<p><strong>Source:</strong> NCBI RefSeq NC_000017.11 (GRCh38.p14)</p>
<p><strong>Region Analyzed:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Chromosome</span><span class="p">:</span>  <span class="mi">17</span>
<span class="n">Coordinates</span><span class="p">:</span> <span class="mi">17</span><span class="p">,</span><span class="mi">680</span><span class="p">,</span><span class="mi">958</span> <span class="o">-</span> <span class="mi">17</span><span class="p">,</span><span class="mi">681</span><span class="p">,</span><span class="mi">958</span> <span class="p">(</span><span class="mi">1000</span> <span class="n">bp</span> <span class="n">promoter</span><span class="p">)</span>
<span class="n">TSS</span><span class="p">:</span>         <span class="mi">17</span><span class="p">,</span><span class="mi">681</span><span class="p">,</span><span class="mi">458</span>
<span class="n">Gene</span><span class="p">:</span>        <span class="n">RAI1</span> <span class="p">(</span><span class="n">Retinoic</span> <span class="n">Acid</span> <span class="n">Induced</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="grna-design-pipeline">
<h3>2.2 gRNA Design Pipeline<a class="headerlink" href="#grna-design-pipeline" title="Link to this heading">#</a></h3>
<p>Our multi-layer validation pipeline:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Layer</p></th>
<th class="head"><p>Method</p></th>
<th class="head"><p>Purpose</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>PAM Scanning</p></td>
<td><p>NGG (SpCas9)</p></td>
<td><p>Find Cas9 binding sites</p></td>
</tr>
<tr class="row-odd"><td><p>GC Content</p></td>
<td><p>40-70% filter</p></td>
<td><p>Optimal hybridization</p></td>
</tr>
<tr class="row-even"><td><p>Thermodynamics</p></td>
<td><p>SantaLucia ΔG</p></td>
<td><p>Binding energy prediction</p></td>
</tr>
<tr class="row-odd"><td><p>Chromatin</p></td>
<td><p>DNase HS model</p></td>
<td><p>Accessibility assessment</p></td>
</tr>
<tr class="row-even"><td><p>MIT Score</p></td>
<td><p>Position-weighted</p></td>
<td><p>Off-target specificity</p></td>
</tr>
<tr class="row-odd"><td><p>CFD Score</p></td>
<td><p>Mismatch penalty</p></td>
<td><p>Cutting frequency</p></td>
</tr>
<tr class="row-even"><td><p><strong>IR Coherence</strong></p></td>
<td><p>R̄ metric</p></td>
<td><p>Simulation reliability</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="quantum-simulation">
<h3>2.3 Quantum Simulation<a class="headerlink" href="#quantum-simulation" title="Link to this heading">#</a></h3>
<p><strong>Hamiltonian Encoding:</strong></p>
<ul class="simple">
<li><p>DNA-RNA base pairing → Pauli ZZ terms</p></li>
<li><p>Stacking interactions → XX/YY terms</p></li>
<li><p>GC content bonus → Identity term</p></li>
</ul>
<p><strong>Hardware:</strong></p>
<ul class="simple">
<li><p>IBM Torino (133 superconducting qubits)</p></li>
<li><p>12 circuits (3 gRNAs × 4 Pauli terms)</p></li>
<li><p>2,000 shots per circuit</p></li>
</ul>
</section>
<section id="circadian-clock-model">
<h3>2.4 Circadian Clock Model<a class="headerlink" href="#circadian-clock-model" title="Link to this heading">#</a></h3>
<p>We modeled the circadian gene network using coupled Kuramoto oscillators:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>RAI1 → CLOCK ↔ BMAL1 → PER1/2/3 ↔ CRY1/2 → (feedback)
              ↓
         REV-ERBα ↔ RORα
</pre></div>
</div>
<p><strong>Model Features:</strong></p>
<ul class="simple">
<li><p>PER delayed negative feedback (τ_P = 4h)</p></li>
<li><p>REV-ERBα/RORα modulation of BMAL1</p></li>
<li><p>RAI1 dosage affects CLOCK-BMAL1 coupling strength</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="results">
<h2>3. Results<a class="headerlink" href="#results" title="Link to this heading">#</a></h2>
<section id="candidate-identification">
<h3>3.1 Candidate Identification<a class="headerlink" href="#candidate-identification" title="Link to this heading">#</a></h3>
<p>From 274 PAM sites in the RAI1 promoter, we identified 91 candidates in the optimal CRISPRa window (-400 to -50 bp from TSS).</p>
<p><strong>Initial Screening Results:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Metric</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Total PAM sites</p></td>
<td><p>274</p></td>
</tr>
<tr class="row-odd"><td><p>CRISPRa window candidates</p></td>
<td><p>91</p></td>
</tr>
<tr class="row-even"><td><p>GO classification rate</p></td>
<td><p>100% (15/15 tested)</p></td>
</tr>
<tr class="row-odd"><td><p>Average coherence (R̄)</p></td>
<td><p>0.944</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="ibm-quantum-hardware-validation">
<h3>3.2 IBM Quantum Hardware Validation<a class="headerlink" href="#ibm-quantum-hardware-validation" title="Link to this heading">#</a></h3>
<p><strong>Run Details:</strong></p>
<ul class="simple">
<li><p>Date: December 10, 2025</p></li>
<li><p>Backend: IBM Torino (133 qubits)</p></li>
<li><p>Job ID: <code class="docutils literal notranslate"><span class="pre">d4suvd7t3pms7399mq8g</span></code></p></li>
<li><p>Cost: FREE (IBM Quantum Open Plan)</p></li>
</ul>
<p><strong>Hardware Results:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>Position</p></th>
<th class="head"><p>Energy</p></th>
<th class="head"><p>Hardware R̄</p></th>
<th class="head"><p>Status</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">GAAGGAGAGCAAGAGCGCGA</span></code></p></td>
<td><p>-81 bp</p></td>
<td><p>-1.894</p></td>
<td><p><strong>0.854</strong></p></td>
<td><p><strong>GO</strong></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">AACTGCAAAGAAGTGGGCAC</span></code></p></td>
<td><p>-334 bp</p></td>
<td><p>-1.563</p></td>
<td><p><strong>0.840</strong></p></td>
<td><p><strong>GO</strong></p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">TACAGGAGCTTCCAGCGTCA</span></code></p></td>
<td><p>-294 bp</p></td>
<td><p>-2.009</p></td>
<td><p><strong>0.839</strong></p></td>
<td><p><strong>GO</strong></p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Simulator-Hardware Agreement:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Simulator R̄</p></th>
<th class="head"><p>Hardware R̄</p></th>
<th class="head"><p>Δ</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>0.833</p></td>
<td><p>0.854</p></td>
<td><p>+2.1%</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>0.820</p></td>
<td><p>0.840</p></td>
<td><p>+2.0%</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>0.827</p></td>
<td><p>0.839</p></td>
<td><p>+1.2%</p></td>
</tr>
</tbody>
</table>
</div>
<p>All candidates showed excellent agreement (within 1-2%) between simulator and real quantum hardware.</p>
</section>
<section id="crispor-genome-wide-validation">
<h3>3.3 CRISPOR Genome-Wide Validation<a class="headerlink" href="#crispor-genome-wide-validation" title="Link to this heading">#</a></h3>
<p><strong>Analysis Details:</strong></p>
<ul class="simple">
<li><p>Tool: CRISPOR v5.2 (crispor.gi.ucsc.edu)</p></li>
<li><p>Genome: Human hg38</p></li>
<li><p>Batch ID: <code class="docutils literal notranslate"><span class="pre">QkJQAaLNZB1vU53T54MY</span></code></p></li>
</ul>
<p><strong>CRISPOR Results:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>MIT</p></th>
<th class="head"><p>CFD</p></th>
<th class="head"><p>Off-targets (0-1-2-3-4 mm)</p></th>
<th class="head"><p>Verdict</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>3</strong></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">TACAGGAGCTTCCAGCGTCA</span></code></p></td>
<td><p><strong>83</strong></p></td>
<td><p><strong>93</strong></p></td>
<td><p><strong>0-0-0-8-77</strong></p></td>
<td><p><strong>BEST</strong></p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">GAAGGAGAGCAAGAGCGCGA</span></code></p></td>
<td><p>51</p></td>
<td><p>65</p></td>
<td><p>0-0-6-72-650</p></td>
<td><p>GOOD</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">AACTGCAAAGAAGTGGGCAC</span></code></p></td>
<td><p>60</p></td>
<td><p>83</p></td>
<td><p>0-1-4-27-203</p></td>
<td><p>CAUTION</p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Key Finding:</strong> The quantum binding energy ranking correctly predicted CRISPOR specificity:</p>
<ul class="simple">
<li><p>gRNA_3: Best energy (-2.009) → Best specificity (MIT 83)</p></li>
<li><p>gRNA_1: Medium energy (-1.894) → Medium specificity (MIT 51)</p></li>
<li><p>gRNA_2: Weakest energy (-1.563) → Has 1-mismatch off-target</p></li>
</ul>
</section>
<section id="circadian-clock-simulation">
<h3>3.4 Circadian Clock Simulation<a class="headerlink" href="#circadian-clock-simulation" title="Link to this heading">#</a></h3>
<p><strong>RAI1 Dosage vs. Circadian Synchronization:</strong></p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>RAI1 Level</p></th>
<th class="head"><p>R̄ (sync)</p></th>
<th class="head"><p>Classification</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>30% (severe)</p></td>
<td><p>0.27</p></td>
<td><p>WEAKLY_SYNCHRONIZED</p></td>
</tr>
<tr class="row-odd"><td><p>50% (SMS)</p></td>
<td><p>0.73</p></td>
<td><p>PARTIALLY_SYNCHRONIZED</p></td>
</tr>
<tr class="row-even"><td><p>60-80%</p></td>
<td><p>0.90-0.99</p></td>
<td><p><strong>SYNCHRONIZED</strong></p></td>
</tr>
<tr class="row-odd"><td><p>100% (normal)</p></td>
<td><p>0.77</p></td>
<td><p>PARTIALLY_SYNCHRONIZED</p></td>
</tr>
</tbody>
</table>
</div>
<p><strong>Therapeutic Insight:</strong> SMS patients (50% RAI1) show partial circadian desynchronization. Boosting RAI1 to 60-80% restores near-perfect synchronization.</p>
</section>
</section>
<hr class="docutils" />
<section id="discussion">
<h2>4. Discussion<a class="headerlink" href="#discussion" title="Link to this heading">#</a></h2>
<section id="primary-candidate-selection">
<h3>4.1 Primary Candidate Selection<a class="headerlink" href="#primary-candidate-selection" title="Link to this heading">#</a></h3>
<p>Based on combined quantum and CRISPOR validation, we recommend:</p>
<p><strong>Primary: gRNA_3</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Sequence:    TACAGGAGCTTCCAGCGTCA
Position:    -294 bp from TSS
MIT Score:   83 (highest)
CFD Score:   93 (highest)
Off-targets: 0 at ≤2 mismatches
Hardware R̄: 0.839
</pre></div>
</div>
<p><strong>Rationale:</strong></p>
<ol class="arabic simple">
<li><p>Highest CRISPOR specificity scores</p></li>
<li><p>Zero off-targets with ≤2 mismatches (safest)</p></li>
<li><p>Best quantum binding energy (-2.009)</p></li>
<li><p>Confirmed GO on IBM hardware</p></li>
</ol>
</section>
<section id="framework-validation">
<h3>4.2 Framework Validation<a class="headerlink" href="#framework-validation" title="Link to this heading">#</a></h3>
<p>This study demonstrates:</p>
<ol class="arabic simple">
<li><p><strong>IR Framework Generalization:</strong> Successfully transferred from molecular (H₂) to biological (gRNA) quantum simulations</p></li>
<li><p><strong>Hardware Reproducibility:</strong> Real quantum hardware results match simulator predictions within 1-2%</p></li>
<li><p><strong>Predictive Power:</strong> Quantum binding energy correlates with empirical specificity metrics</p></li>
</ol>
</section>
<section id="therapeutic-window">
<h3>4.3 Therapeutic Window<a class="headerlink" href="#therapeutic-window" title="Link to this heading">#</a></h3>
<p>Our circadian model predicts:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Target</p></th>
<th class="head"><p>RAI1 Level</p></th>
<th class="head"><p>Required Boost</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Minimum effective</p></td>
<td><p>60%</p></td>
<td><p>+20% from SMS baseline</p></td>
</tr>
<tr class="row-odd"><td><p>Optimal</p></td>
<td><p>80%</p></td>
<td><p>+60% from SMS baseline</p></td>
</tr>
<tr class="row-even"><td><p>Maximum safe</p></td>
<td><p>100%</p></td>
<td><p>+100% from SMS baseline</p></td>
</tr>
</tbody>
</table>
</div>
<p>Published CRISPRa studies report 2-3× activation, making this window achievable.</p>
</section>
</section>
<hr class="docutils" />
<section id="recommended-next-steps">
<h2>5. Recommended Next Steps<a class="headerlink" href="#recommended-next-steps" title="Link to this heading">#</a></h2>
<section id="use-sms-trials-module-v0-9-0">
<h3>5.1 Use SMS Trials Module (v0.9.0+)<a class="headerlink" href="#use-sms-trials-module-v0-9-0" title="Link to this heading">#</a></h3>
<p>PhaseLab v0.9.0 introduces a complete SMS therapeutic trial framework:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.trials.sms</span><span class="w"> </span><span class="kn">import</span> <span class="n">SMSPipeline</span><span class="p">,</span> <span class="n">SMSTrialConfig</span>

<span class="c1"># Configure and run full pipeline</span>
<span class="n">config</span> <span class="o">=</span> <span class="n">SMSTrialConfig</span><span class="p">(</span><span class="n">therapeutic_window</span><span class="o">=</span><span class="p">(</span><span class="mf">0.70</span><span class="p">,</span> <span class="mf">1.10</span><span class="p">))</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">SMSPipeline</span><span class="p">(</span><span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">)</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">run_full_pipeline</span><span class="p">()</span>

<span class="c1"># Get GO/NO-GO decision</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Overall: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">overall_go_nogo</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Top guide: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">crispra_result</span><span class="o">.</span><span class="n">best_candidate</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Circadian rescue: </span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">circadian_result</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s1">&#39;rescue_status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Generate falsification tests for wet lab</span>
<span class="k">for</span> <span class="n">test</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">falsification_tests</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Test </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">test</span><span class="p">[</span><span class="s1">&#39;failure_condition&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="wet-lab-validation">
<h3>5.2 Wet Lab Validation<a class="headerlink" href="#wet-lab-validation" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p><strong>Order Synthesis</strong></p>
<ul class="simple">
<li><p>Primary: <code class="docutils literal notranslate"><span class="pre">TACAGGAGCTTCCAGCGTCA</span></code> (gRNA_3)</p></li>
<li><p>Backup: <code class="docutils literal notranslate"><span class="pre">GAAGGAGAGCAAGAGCGCGA</span></code> (gRNA_1)</p></li>
<li><p>Source: IDT, Synthego, or similar</p></li>
</ul>
</li>
<li><p><strong>Clone into Vector</strong></p>
<ul class="simple">
<li><p>dCas9-VP64 system (e.g., Addgene #135338)</p></li>
<li><p>Verify by Sanger sequencing</p></li>
</ul>
</li>
<li><p><strong>Cell Culture Testing</strong></p>
<ul class="simple">
<li><p>Initial: HEK293T</p></li>
<li><p>Disease-relevant: SH-SY5Y or iPSC-derived neurons</p></li>
<li><p>Patient-derived: iPSCs from SMS patients</p></li>
</ul>
</li>
<li><p><strong>Readouts</strong></p>
<ul class="simple">
<li><p>qPCR for RAI1 mRNA (48-72h post-transfection)</p></li>
<li><p>Western blot for RAI1 protein</p></li>
<li><p>Target: 150-200% of baseline expression</p></li>
</ul>
</li>
</ol>
</section>
<section id="falsification-tests">
<h3>5.3 Falsification Tests<a class="headerlink" href="#falsification-tests" title="Link to this heading">#</a></h3>
<p>Run these validation experiments to confirm PhaseLab predictions:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Test</p></th>
<th class="head"><p>Experiment</p></th>
<th class="head"><p>Pass Criterion</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>A</strong></p></td>
<td><p>Compare PhaseLab top 10 vs random 10 guides</p></td>
<td><p>PhaseLab outperforms by ≥1.5σ</p></td>
</tr>
<tr class="row-odd"><td><p><strong>B</strong></p></td>
<td><p>Test CAUTION-labeled guides</p></td>
<td><p>Higher failure rate than GO guides</p></td>
</tr>
<tr class="row-even"><td><p><strong>C</strong></p></td>
<td><p>Measure expression at predicted doses</p></td>
<td><p>Correlation r &gt; 0.6</p></td>
</tr>
<tr class="row-odd"><td><p><strong>D</strong></p></td>
<td><p>Test UNKNOWN-labeled guides</p></td>
<td><p>Random pass/fail rate</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="research-partnerships">
<h3>5.4 Research Partnerships<a class="headerlink" href="#research-partnerships" title="Link to this heading">#</a></h3>
<p>We recommend contacting:</p>
<ul class="simple">
<li><p>SMS Research Foundation</p></li>
<li><p>Academic groups studying RAI1 biology</p></li>
<li><p>Gene therapy companies with CRISPRa platforms</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="limitations">
<h2>6. Limitations<a class="headerlink" href="#limitations" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Computational only</strong> - No wet lab validation yet</p></li>
<li><p><strong>Chromatin model</strong> - Generic, not patient/tissue-specific</p></li>
<li><p><strong>Off-target prediction</strong> - Algorithmic, not empirical</p></li>
<li><p><strong>Binding ≠ Activation</strong> - Strong binding doesn’t guarantee efficient transcriptional activation</p></li>
<li><p><strong>Delivery</strong> - In vivo delivery to neurons remains a challenge</p></li>
</ol>
</section>
<hr class="docutils" />
<section id="conclusion">
<h2>7. Conclusion<a class="headerlink" href="#conclusion" title="Link to this heading">#</a></h2>
<p>We have identified and validated a lead CRISPRa guide RNA candidate for RAI1 upregulation in Smith-Magenis Syndrome:</p>
<p><strong><code class="docutils literal notranslate"><span class="pre">TACAGGAGCTTCCAGCGTCA</span></code></strong></p>
<p>This candidate was validated through:</p>
<ul class="simple">
<li><p>Quantum coherence (R̄ = 0.839 on IBM Torino)</p></li>
<li><p>CRISPOR specificity (MIT 83, CFD 93)</p></li>
<li><p>Zero off-targets with ≤2 mismatches</p></li>
<li><p>Best quantum binding energy</p></li>
</ul>
<p>The IR coherence framework provides a novel reliability metric for guide RNA design that complements traditional computational methods. All tested candidates achieved GO classification, and quantum predictions correlated with empirical specificity scores.</p>
</section>
<hr class="docutils" />
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>PMC9762195 - rAAV-CRISPRa therapy for SMS in mice</p></li>
<li><p>Hsu et al. (2013) - MIT specificity score</p></li>
<li><p>Doench et al. (2016) - CFD score methodology</p></li>
<li><p>SantaLucia (1998) - Nearest-neighbor thermodynamic parameters</p></li>
<li><p>CRISPOR - https://crispor.tefor.net/</p></li>
<li><p>IBM Quantum - https://quantum.ibm.com/</p></li>
</ol>
</section>
<hr class="docutils" />
<section id="appendix-a-data-availability">
<h2>Appendix A: Data Availability<a class="headerlink" href="#appendix-a-data-availability" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Resource</p></th>
<th class="head"><p>Location</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Quantum results</p></td>
<td><p>IBM Quantum: Job <code class="docutils literal notranslate"><span class="pre">d4suvd7t3pms7399mq8g</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>CRISPOR analysis</p></td>
<td><p>Batch <code class="docutils literal notranslate"><span class="pre">QkJQAaLNZB1vU53T54MY</span></code></p></td>
</tr>
<tr class="row-even"><td><p>PhaseLab package</p></td>
<td><p>https://github.com/followthesapper/phaselab</p></td>
</tr>
<tr class="row-odd"><td><p>Raw data</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">experiments/E200_SMS_Gene_Therapy/Data/</span></code></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="appendix-b-code-availability">
<h2>Appendix B: Code Availability<a class="headerlink" href="#appendix-b-code-availability" title="Link to this heading">#</a></h2>
<p>All code is available in the PhaseLab Python package:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>phaselab<span class="o">[</span>quantum<span class="o">]</span>
</pre></div>
</div>
<p>Example usage:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.circadian</span><span class="w"> </span><span class="kn">import</span> <span class="n">simulate_sms_clock</span>

<span class="c1"># Design guides for RAI1 promoter</span>
<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">rai1_promoter</span><span class="p">,</span> <span class="n">tss_index</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="n">top_candidates</span> <span class="o">=</span> <span class="n">guides</span><span class="p">[</span><span class="n">guides</span><span class="p">[</span><span class="s1">&#39;go_no_go&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;GO&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>

<span class="c1"># Predict therapeutic window</span>
<span class="k">for</span> <span class="n">level</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">]:</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">simulate_sms_clock</span><span class="p">(</span><span class="n">rai1_level</span><span class="o">=</span><span class="n">level</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;RAI1 </span><span class="si">{</span><span class="n">level</span><span class="si">:</span><span class="s2">.0%</span><span class="si">}</span><span class="s2">: R̄=</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<hr class="docutils" />
<p><em>This research was conducted as part of the Informational Relativity experimental series.</em></p>
<p><em>Hardware validation: IBM Torino, December 2025</em></p>
<p><em>CRISPOR validation: Batch QkJQAaLNZB1vU53T54MY</em></p>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#abstract">Abstract</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#background">1. Background</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#smith-magenis-syndrome">1.1 Smith-Magenis Syndrome</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#therapeutic-approach">1.2 Therapeutic Approach</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-ir-coherence-framework">1.3 The IR Coherence Framework</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods">2. Methods</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genomic-data">2.1 Genomic Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#grna-design-pipeline">2.2 gRNA Design Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-simulation">2.3 Quantum Simulation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#circadian-clock-model">2.4 Circadian Clock Model</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#results">3. Results</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#candidate-identification">3.1 Candidate Identification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ibm-quantum-hardware-validation">3.2 IBM Quantum Hardware Validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispor-genome-wide-validation">3.3 CRISPOR Genome-Wide Validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#circadian-clock-simulation">3.4 Circadian Clock Simulation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#discussion">4. Discussion</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#primary-candidate-selection">4.1 Primary Candidate Selection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#framework-validation">4.2 Framework Validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#therapeutic-window">4.3 Therapeutic Window</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recommended-next-steps">5. Recommended Next Steps</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#use-sms-trials-module-v0-9-0">5.1 Use SMS Trials Module (v0.9.0+)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#wet-lab-validation">5.2 Wet Lab Validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#falsification-tests">5.3 Falsification Tests</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#research-partnerships">5.4 Research Partnerships</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#limitations">6. Limitations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">7. Conclusion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#appendix-a-data-availability">Appendix A: Data Availability</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#appendix-b-code-availability">Appendix B: Code Availability</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Dylan Vaca.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>