
<!DOCTYPE html>


<html lang="en" data-content_root="../" data-theme="auto">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Quantum-Informed CRISPRa gRNA Design for RAI1 Activation in Smith-Magenis Syndrome &#8212; PhaseLab 0.6.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "auto";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "auto";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=09465d15" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=2bd037f0"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=30646c52"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'papers/PAPER_2_CRISPRA_GRNA_DESIGN';</script>
    <link rel="icon" href="../_static/favicon.svg"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="auto">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo-lm.svg" class="logo__image only-light" alt=""/>
    <img src="../_static/logo.svg" class="logo__image only-dark pst-js-only" alt=""/>
  
  
    <p class="title logo__title">PhaseLab</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/explanations/index.html">
    Explanations
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../reference/index.html">
    API Reference
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../research/index.html">
    Research Applications
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../examples/index.html">
    Examples
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../citing.html">
    Citing PhaseLab
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../user_guide/explanations/index.html">
    Explanations
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../reference/index.html">
    API Reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../research/index.html">
    Research Applications
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../citing.html">
    Citing PhaseLab
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Quantum-Informed CRISPRa gRNA Design for RAI1 Activation in Smith-Magenis Syndrome</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="quantum-informed-crispra-grna-design-for-rai1-activation-in-smith-magenis-syndrome">
<h1>Quantum-Informed CRISPRa gRNA Design for RAI1 Activation in Smith-Magenis Syndrome<a class="headerlink" href="#quantum-informed-crispra-grna-design-for-rai1-activation-in-smith-magenis-syndrome" title="Link to this heading">#</a></h1>
<p>Dylan Vaca</p>
<p>December 2025</p>
<hr class="docutils" />
<section id="abstract">
<h2>Abstract<a class="headerlink" href="#abstract" title="Link to this heading">#</a></h2>
<p>Guide RNA design for CRISPR activation requires balancing multiple factors including target specificity, chromatin accessibility, thermodynamic stability, and transcriptional activation efficiency. Current computational tools evaluate these factors independently without a unified reliability metric. This paper presents a quantum-informed approach using PhaseLab to design CRISPRa guides for RAI1 upregulation in Smith-Magenis Syndrome.</p>
<p>We encoded DNA-RNA interactions as Pauli Hamiltonians and computed binding energies through quantum simulation. The resulting candidates were validated on IBM Quantum hardware (Torino, 133 qubits) and through CRISPOR genome-wide off-target analysis. The top candidate (TACAGGAGCTTCCAGCGTCA) achieved MIT specificity of 83, CFD score of 93, and zero off-targets with two or fewer mismatches, representing the highest specificity among all tested guides.</p>
<p>The quantum binding energy rankings correctly predicted CRISPOR specificity scores, suggesting that quantum Hamiltonian encoding captures biologically relevant sequence features. This work represents the first application of quantum computing to CRISPRa guide design and establishes a new approach for reliability-assessed genomic engineering.</p>
<p>Keywords: CRISPRa, guide RNA, quantum computing, Smith-Magenis Syndrome, RAI1, off-target, specificity</p>
</section>
<hr class="docutils" />
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<section id="crispra-for-gene-therapy">
<h3>1.1 CRISPRa for Gene Therapy<a class="headerlink" href="#crispra-for-gene-therapy" title="Link to this heading">#</a></h3>
<p>CRISPR activation uses catalytically dead Cas9 fused to transcriptional activators to increase gene expression without cutting DNA. This approach is particularly relevant for haploinsufficiency disorders, where boosting expression from the remaining functional allele could restore normal protein levels.</p>
</section>
<section id="smith-magenis-syndrome">
<h3>1.2 Smith-Magenis Syndrome<a class="headerlink" href="#smith-magenis-syndrome" title="Link to this heading">#</a></h3>
<p>Smith-Magenis Syndrome is a rare genetic disorder affecting approximately 1 in 15,000 to 25,000 births, caused by haploinsufficiency of RAI1 at chromosome 17p11.2. Patients experience intellectual disability, inverted circadian rhythm with disrupted sleep patterns, and characteristic behavioral phenotypes.</p>
<p>Most patients have deletions that include RAI1, leaving one functional copy that produces approximately 50% of normal protein levels. CRISPRa targeting the remaining alleleâ€™s promoter could potentially restore expression to therapeutic levels.</p>
</section>
<section id="the-guide-design-challenge">
<h3>1.3 The Guide Design Challenge<a class="headerlink" href="#the-guide-design-challenge" title="Link to this heading">#</a></h3>
<p>Effective CRISPRa requires guides that bind specifically to minimize off-target effects, access open chromatin regions, position optimally relative to the transcription start site (typically 50 to 400 base pairs upstream for CRISPRa), and bind with favorable thermodynamics.</p>
<p>Current tools such as CRISPOR and Benchling score these factors independently. This paper proposes a unified approach using quantum coherence as a reliability metric.</p>
</section>
<section id="quantum-informed-design">
<h3>1.4 Quantum-Informed Design<a class="headerlink" href="#quantum-informed-design" title="Link to this heading">#</a></h3>
<p>By encoding gRNA-DNA interactions as quantum Hamiltonians, we can compute binding energies through VQE-style simulation, assess simulation reliability through the coherence metric, validate predictions on real quantum hardware, and correlate quantum observables with empirical specificity measures.</p>
<p>This paper presents the first quantum-informed CRISPRa guide design pipeline, applied to RAI1 activation for SMS therapy.</p>
</section>
</section>
<hr class="docutils" />
<section id="methods">
<h2>2. Methods<a class="headerlink" href="#methods" title="Link to this heading">#</a></h2>
<section id="target-region">
<h3>2.1 Target Region<a class="headerlink" href="#target-region" title="Link to this heading">#</a></h3>
<p>We analyzed the RAI1 gene from human genome assembly GRCh38.p14. The promoter region spans chromosome 17 coordinates 17,680,958 to 17,681,958 (1000 base pairs), with the transcription start site at position 17,681,458. The sequence was retrieved from NCBI RefSeq (NC_000017.11).</p>
</section>
<section id="pam-site-identification">
<h3>2.2 PAM Site Identification<a class="headerlink" href="#pam-site-identification" title="Link to this heading">#</a></h3>
<p>We scanned for SpCas9 PAM sites (NGG) using PhaseLab:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">find_pam_sites</span>

<span class="n">sites</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span><span class="n">promoter_sequence</span><span class="p">,</span> <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">,</span> <span class="n">guide_length</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
<p>This identified 274 total PAM sites, with 91 located within the CRISPRa window (50 to 400 base pairs upstream of the TSS).</p>
</section>
<section id="multi-layer-scoring">
<h3>2.3 Multi-Layer Scoring<a class="headerlink" href="#multi-layer-scoring" title="Link to this heading">#</a></h3>
<p>Each candidate was evaluated across multiple criteria:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Layer</p></th>
<th class="head"><p>Method</p></th>
<th class="head"><p>Weight</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>GC Content</p></td>
<td><p>40-70% optimal range</p></td>
<td><p>0.15</p></td>
</tr>
<tr class="row-odd"><td><p>Thermodynamics</p></td>
<td><p>SantaLucia free energy</p></td>
<td><p>0.20</p></td>
</tr>
<tr class="row-even"><td><p>Chromatin</p></td>
<td><p>DNase hypersensitivity model</p></td>
<td><p>0.20</p></td>
</tr>
<tr class="row-odd"><td><p>MIT Score</p></td>
<td><p>Position-weighted mismatches</p></td>
<td><p>0.20</p></td>
</tr>
<tr class="row-even"><td><p>Quantum Coherence</p></td>
<td><p>Coherence metric</p></td>
<td><p>0.25</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="quantum-hamiltonian-encoding">
<h3>2.4 Quantum Hamiltonian Encoding<a class="headerlink" href="#quantum-hamiltonian-encoding" title="Link to this heading">#</a></h3>
<p>DNA-RNA interactions were encoded as:</p>
<div class="math notranslate nohighlight">
\[H = H_{WC} + H_{stack} + H_{GC}\]</div>
<p>The Watson-Crick term uses:</p>
<div class="math notranslate nohighlight">
\[H_{WC} = \sum_{i=1}^{20} \alpha_i Z_i \otimes Z'_i\]</div>
<p>The coefficient <span class="math notranslate nohighlight">\(\alpha_i\)</span> depends on the match type: G-C pairs receive -1.5 (three hydrogen bonds), A-T/U pairs receive -1.0 (two hydrogen bonds), and mismatches receive +0.5 (destabilizing).</p>
<p>The stacking term captures nearest-neighbor interactions:</p>
<div class="math notranslate nohighlight">
\[H_{stack} = \sum_{i=1}^{19} \beta_i (X_i X_{i+1} + Y_i Y_{i+1})\]</div>
<p>Stacking parameters follow SantaLucia nearest-neighbor values.</p>
<p>The GC bonus term adds:</p>
<div class="math notranslate nohighlight">
\[H_{GC} = -0.1 \times (\text{GC count})\]</div>
</section>
<section id="simulation-protocol">
<h3>2.5 Simulation Protocol<a class="headerlink" href="#simulation-protocol" title="Link to this heading">#</a></h3>
<p>Simulations used the Qiskit Aer statevector backend with an EfficientSU2 ansatz at one repetition. We measured four Pauli terms (ZIII, IZII, ZZII, XXII) with 2,000 shots per term.</p>
</section>
<section id="hardware-validation">
<h3>2.6 Hardware Validation<a class="headerlink" href="#hardware-validation" title="Link to this heading">#</a></h3>
<p>Hardware validation used IBM Torino (133 superconducting qubits) on December 10, 2025. Job ID: d4suvd7t3pms7399mq8g. We ran 12 circuits (3 gRNAs times 4 Pauli terms) with 2,000 shots per circuit.</p>
</section>
<section id="crispor-validation">
<h3>2.7 CRISPOR Validation<a class="headerlink" href="#crispor-validation" title="Link to this heading">#</a></h3>
<p>Top candidates were submitted to CRISPOR for genome-wide off-target analysis using human genome hg38 with the 20bp-NGG PAM setting for SpCas9. Batch ID: QkJQAaLNZB1vU53T54MY.</p>
</section>
</section>
<hr class="docutils" />
<section id="results">
<h2>3. Results<a class="headerlink" href="#results" title="Link to this heading">#</a></h2>
<section id="candidate-screening">
<h3>3.1 Candidate Screening<a class="headerlink" href="#candidate-screening" title="Link to this heading">#</a></h3>
<p>From 91 candidates in the CRISPRa window, the top 15 were selected for detailed analysis based on combined scoring.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Metric</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Candidates screened</p></td>
<td><p>91</p></td>
</tr>
<tr class="row-odd"><td><p>GC content 40-70%</p></td>
<td><p>78 (86%)</p></td>
</tr>
<tr class="row-even"><td><p>Open chromatin</p></td>
<td><p>34 (37%)</p></td>
</tr>
<tr class="row-odd"><td><p>GO classification</p></td>
<td><p>91 (100%)</p></td>
</tr>
</tbody>
</table>
</div>
<p>All candidates achieved GO classification (coherence above the threshold), indicating reliable simulations throughout.</p>
</section>
<section id="top-candidates">
<h3>3.2 Top Candidates<a class="headerlink" href="#top-candidates" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Rank</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>Position</p></th>
<th class="head"><p>GC</p></th>
<th class="head"><p>Free Energy</p></th>
<th class="head"><p>Chromatin</p></th>
<th class="head"><p>Coherence</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>GAGGAAAAAAGTTCCCGGGC</p></td>
<td><p>-99</p></td>
<td><p>55%</p></td>
<td><p>-21.3</p></td>
<td><p>Open</p></td>
<td><p>0.945</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>GAAGGAGAGCAAGAGCGCGA</p></td>
<td><p>-81</p></td>
<td><p>60%</p></td>
<td><p>-22.1</p></td>
<td><p>Open</p></td>
<td><p>0.943</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>TACAGGAGCTTCCAGCGTCA</p></td>
<td><p>-294</p></td>
<td><p>55%</p></td>
<td><p>-20.8</p></td>
<td><p>Moderate</p></td>
<td><p>0.944</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>AGTGTCCTCCCAGGCCGCAC</p></td>
<td><p>-243</p></td>
<td><p>70%</p></td>
<td><p>-23.5</p></td>
<td><p>Moderate</p></td>
<td><p>0.942</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p>AACTGCAAAGAAGTGGGCAC</p></td>
<td><p>-334</p></td>
<td><p>50%</p></td>
<td><p>-19.2</p></td>
<td><p>CpG island</p></td>
<td><p>0.944</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="hardware-validation-results">
<h3>3.3 Hardware Validation Results<a class="headerlink" href="#hardware-validation-results" title="Link to this heading">#</a></h3>
<p>Three candidates were validated on IBM Torino:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>Energy</p></th>
<th class="head"><p>Simulator</p></th>
<th class="head"><p>Hardware</p></th>
<th class="head"><p>Difference</p></th>
<th class="head"><p>Status</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>GAAGGAGAGCAAGAGCGCGA</p></td>
<td><p>-1.894</p></td>
<td><p>0.833</p></td>
<td><p>0.854</p></td>
<td><p>+2.1%</p></td>
<td><p>GO</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>AACTGCAAAGAAGTGGGCAC</p></td>
<td><p>-1.563</p></td>
<td><p>0.820</p></td>
<td><p>0.840</p></td>
<td><p>+2.0%</p></td>
<td><p>GO</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>TACAGGAGCTTCCAGCGTCA</p></td>
<td><p>-2.009</p></td>
<td><p>0.827</p></td>
<td><p>0.839</p></td>
<td><p>+1.2%</p></td>
<td><p>GO</p></td>
</tr>
</tbody>
</table>
</div>
<p>All candidates were confirmed as GO on real quantum hardware. The agreement between simulator and hardware was excellent, differing by only 1-2%. Binding energy rankings remained consistent across both platforms.</p>
</section>
<section id="crispor-genome-wide-analysis">
<h3>3.4 CRISPOR Genome-Wide Analysis<a class="headerlink" href="#crispor-genome-wide-analysis" title="Link to this heading">#</a></h3>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>MIT</p></th>
<th class="head"><p>CFD</p></th>
<th class="head"><p>Off-targets by mismatch (0-1-2-3-4)</p></th>
<th class="head"><p>Verdict</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>3</p></td>
<td><p>TACAGGAGCTTCCAGCGTCA</p></td>
<td><p>83</p></td>
<td><p>93</p></td>
<td><p>0-0-0-8-77</p></td>
<td><p>Best</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>GAAGGAGAGCAAGAGCGCGA</p></td>
<td><p>51</p></td>
<td><p>65</p></td>
<td><p>0-0-6-72-650</p></td>
<td><p>Good</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>AACTGCAAAGAAGTGGGCAC</p></td>
<td><p>60</p></td>
<td><p>83</p></td>
<td><p>0-1-4-27-203</p></td>
<td><p>Caution</p></td>
</tr>
</tbody>
</table>
</div>
<p>Candidate 2 has one off-target with only one mismatch (in an MYPN intron), making it less suitable for therapeutic use.</p>
</section>
<section id="quantum-specificity-correlation">
<h3>3.5 Quantum-Specificity Correlation<a class="headerlink" href="#quantum-specificity-correlation" title="Link to this heading">#</a></h3>
<p>The quantum binding energy rankings correctly predicted CRISPOR specificity:</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>gRNA</p></th>
<th class="head"><p>Binding Energy</p></th>
<th class="head"><p>MIT Score</p></th>
<th class="head"><p>Prediction Correct</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>3</p></td>
<td><p>-2.009 (best)</p></td>
<td><p>83 (best)</p></td>
<td><p>Yes</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>-1.894 (middle)</p></td>
<td><p>51 (middle)</p></td>
<td><p>Yes</p></td>
</tr>
<tr class="row-even"><td><p>2</p></td>
<td><p>-1.563 (worst)</p></td>
<td><p>60 with off-target</p></td>
<td><p>Yes</p></td>
</tr>
</tbody>
</table>
</div>
<p>Lower (more favorable) quantum binding energy correlates with higher specificity. The Hamiltonian encoding appears to capture sequence features relevant to off-target binding.</p>
</section>
<section id="final-recommendation">
<h3>3.6 Final Recommendation<a class="headerlink" href="#final-recommendation" title="Link to this heading">#</a></h3>
<p>Based on combined quantum and CRISPOR validation, we recommend:</p>
<p>Primary candidate:</p>
<ul class="simple">
<li><p>Sequence: TACAGGAGCTTCCAGCGTCA</p></li>
<li><p>Position: 294 base pairs upstream of TSS</p></li>
<li><p>Strand: Minus</p></li>
<li><p>MIT Score: 83 (highest)</p></li>
<li><p>CFD Score: 93 (highest)</p></li>
<li><p>Off-targets: Zero at two or fewer mismatches</p></li>
<li><p>Hardware coherence: 0.839</p></li>
<li><p>Binding energy: -2.009 (most favorable)</p></li>
</ul>
<p>Backup candidate:</p>
<ul class="simple">
<li><p>Sequence: GAAGGAGAGCAAGAGCGCGA</p></li>
<li><p>Position: 81 base pairs upstream of TSS</p></li>
<li><p>MIT Score: 51</p></li>
<li><p>Hardware coherence: 0.854 (highest)</p></li>
<li><p>Off-targets: Zero at one or fewer mismatches</p></li>
</ul>
</section>
</section>
<hr class="docutils" />
<section id="discussion">
<h2>4. Discussion<a class="headerlink" href="#discussion" title="Link to this heading">#</a></h2>
<section id="advantages-of-quantum-informed-design">
<h3>4.1 Advantages of Quantum-Informed Design<a class="headerlink" href="#advantages-of-quantum-informed-design" title="Link to this heading">#</a></h3>
<p>Traditional gRNA design tools treat scoring components independently. The quantum approach provides a unified energy landscape where all interactions appear in one Hamiltonian, a reliability metric indicating simulation trustworthiness, hardware validation on real quantum processors, and predictive power where energy correlates with specificity.</p>
</section>
<section id="the-binding-specificity-connection">
<h3>4.2 The Binding-Specificity Connection<a class="headerlink" href="#the-binding-specificity-connection" title="Link to this heading">#</a></h3>
<p>The correlation between quantum binding energy and CRISPOR specificity scores is notable. We hypothesize that guides with more favorable binding energies have stronger on-target affinity, which creates selectivity that reduces effective off-target binding. The quantum Hamiltonian captures nucleotide-level determinants of specificity.</p>
</section>
<section id="comparison-to-published-work">
<h3>4.3 Comparison to Published Work<a class="headerlink" href="#comparison-to-published-work" title="Link to this heading">#</a></h3>
<p>A previous study (PMC9762195) achieved successful CRISPRa in SMS mice using SaCas9 with a different PAM sequence (NNGRRT) and targeting the mouse Rai1 promoter. They demonstrated that 2-3 fold activation was sufficient for phenotype rescue.</p>
<p>Our analysis uses SpCas9 (NGG PAM), which is more widely used, and targets the human RAI1 promoter for direct translatability. The quantum-validated specificity provides a novel quality metric.</p>
</section>
<section id="clinical-translation-path">
<h3>4.4 Clinical Translation Path<a class="headerlink" href="#clinical-translation-path" title="Link to this heading">#</a></h3>
<p>For therapeutic development, we recommend the following steps. First, synthesize gRNA candidate 3 from a commercial provider such as IDT or Synthego. Second, clone into a dCas9-VP64 vector such as Addgene 135338. Third, test in HEK293T cells initially, then in SH-SY5Y neurons. Fourth, measure outcomes through qPCR for mRNA and Western blot for protein. The target is 150-200% of baseline RAI1 expression.</p>
</section>
<section id="limitations">
<h3>4.5 Limitations<a class="headerlink" href="#limitations" title="Link to this heading">#</a></h3>
<p>Several limitations should be acknowledged. No wet lab data are available; these are computational predictions only. The chromatin model is generic rather than neural-specific. The 4-qubit encoding is simplified; full 20 base pair encoding would require larger circuits. Finally, strong binding does not guarantee efficient transcriptional activation.</p>
</section>
</section>
<hr class="docutils" />
<section id="conclusion">
<h2>5. Conclusion<a class="headerlink" href="#conclusion" title="Link to this heading">#</a></h2>
<p>This paper presents the first quantum-informed approach to CRISPRa guide design, applied to RAI1 activation for Smith-Magenis Syndrome. We identified an optimal guide (TACAGGAGCTTCCAGCGTCA) with MIT specificity of 83 and CFD score of 93, zero critical off-targets at two or fewer mismatches, hardware-validated coherence of 0.839 on IBM Torino, and predictive correlation between quantum energy and empirical specificity.</p>
<p>The PhaseLab framework enables quantum-informed genomic engineering with built-in reliability assessment. This approach could extend to other CRISPR applications, protein engineering, and drug design.</p>
</section>
<hr class="docutils" />
<section id="methods-extended">
<h2>6. Methods (Extended)<a class="headerlink" href="#methods-extended" title="Link to this heading">#</a></h2>
<section id="sequence-retrieval">
<h3>6.1 Sequence Retrieval<a class="headerlink" href="#sequence-retrieval" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>

<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;researcher@institution.edu&quot;</span>
<span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;NC_000017.11&quot;</span><span class="p">,</span>
                       <span class="n">seq_start</span><span class="o">=</span><span class="mi">17680958</span><span class="p">,</span> <span class="n">seq_stop</span><span class="o">=</span><span class="mi">17681958</span><span class="p">,</span>
                       <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="n">promoter</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>
</pre></div>
</div>
</section>
<section id="phaselab-pipeline">
<h3>6.2 PhaseLab Pipeline<a class="headerlink" href="#phaselab-pipeline" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.crispr</span><span class="w"> </span><span class="kn">import</span> <span class="n">design_guides</span><span class="p">,</span> <span class="n">CRISPRConfig</span>

<span class="n">config</span> <span class="o">=</span> <span class="n">CRISPRConfig</span><span class="p">(</span>
    <span class="n">pam</span><span class="o">=</span><span class="s2">&quot;NGG&quot;</span><span class="p">,</span>
    <span class="n">guide_length</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">window_upstream</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span>
    <span class="n">window_downstream</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">min_gc</span><span class="o">=</span><span class="mf">0.40</span><span class="p">,</span>
    <span class="n">max_gc</span><span class="o">=</span><span class="mf">0.70</span>
<span class="p">)</span>

<span class="n">guides</span> <span class="o">=</span> <span class="n">design_guides</span><span class="p">(</span><span class="n">promoter</span><span class="p">,</span> <span class="n">tss_index</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">config</span><span class="o">=</span><span class="n">config</span><span class="p">)</span>
<span class="n">top_candidates</span> <span class="o">=</span> <span class="n">guides</span><span class="p">[</span><span class="n">guides</span><span class="p">[</span><span class="s1">&#39;go_no_go&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;GO&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="hardware-execution">
<h3>6.3 Hardware Execution<a class="headerlink" href="#hardware-execution" title="Link to this heading">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">phaselab.quantum</span><span class="w"> </span><span class="kn">import</span> <span class="n">QuantumCoherenceValidator</span>

<span class="n">validator</span> <span class="o">=</span> <span class="n">QuantumCoherenceValidator</span><span class="p">(</span>
    <span class="n">backend</span><span class="o">=</span><span class="s1">&#39;ibm_torino&#39;</span><span class="p">,</span>
    <span class="n">shots</span><span class="o">=</span><span class="mi">2000</span>
<span class="p">)</span>

<span class="k">for</span> <span class="n">guide</span> <span class="ow">in</span> <span class="n">top_candidates</span><span class="p">[</span><span class="s1">&#39;sequence&#39;</span><span class="p">]:</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">validator</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">guide</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">guide</span><span class="si">}</span><span class="s2">: coherence=</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;coherence&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2"> [</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="data-availability">
<h2>7. Data Availability<a class="headerlink" href="#data-availability" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Resource</p></th>
<th class="head"><p>Identifier</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>IBM Quantum Job</p></td>
<td><p>d4suvd7t3pms7399mq8g</p></td>
</tr>
<tr class="row-odd"><td><p>CRISPOR Batch</p></td>
<td><p>QkJQAaLNZB1vU53T54MY</p></td>
</tr>
<tr class="row-even"><td><p>PhaseLab Package</p></td>
<td><p>https://pypi.org/project/phaselab/</p></td>
</tr>
<tr class="row-odd"><td><p>Source Code</p></td>
<td><p>https://github.com/followthesapper/phaselab</p></td>
</tr>
<tr class="row-even"><td><p>RAI1 RefSeq</p></td>
<td><p>NC_000017.11</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<hr class="docutils" />
<section id="figures">
<h2>8. Figures<a class="headerlink" href="#figures" title="Link to this heading">#</a></h2>
<p>Figure 1 presents a genomic map of the RAI1 promoter region (chromosome 17, positions 17,680,958 to 17,681,958) showing the transcription start site, CRISPRa window (50 to 400 base pairs upstream), CpG islands, and DNase hypersensitivity sites. Top candidate positions are marked.</p>
<p>Figure 2 shows a histogram of 274 NGG PAM sites across the promoter region, colored by strand orientation, with the CRISPRa window highlighted.</p>
<p>Figure 3 displays dual heatmaps of SantaLucia free energy values and MIT/CFD specificity scores for all 91 candidates in the CRISPRa window, with the top 5 candidates highlighted.</p>
<p>Figure 4 presents a scatter plot of quantum binding energy versus CRISPOR MIT score for the three hardware-validated candidates, with a linear regression line showing the correlation.</p>
<p>Figure 5 compares coherence values on the Aer simulator versus IBM Torino hardware for all three candidates, with error bars from shot noise and a dashed line indicating perfect agreement.</p>
<p>Figure 6 shows off-target counts at 0, 1, 2, 3, and 4 mismatches for each candidate, highlighting candidate 3 (TACAGGAGCTTCCAGCGTCA) as optimal with zero off-targets at two or fewer mismatches.</p>
<p>Figure 7 illustrates the complete CRISPRa design pipeline as a flowchart: promoter sequence to PAM scanning to multi-layer filtering (GC, free energy, chromatin, MIT/CFD) to quantum Hamiltonian simulation to hardware validation to CRISPOR verification to final candidate selection.</p>
</section>
<hr class="docutils" />
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Chavez A, et al. Highly efficient Cas9-mediated transcriptional programming. Nature Methods. 2015.</p></li>
<li><p>Gilbert LA, et al. CRISPR-mediated modular RNA-guided regulation of transcription in eukaryotes. Cell. 2013.</p></li>
<li><p>Hsu PD, et al. DNA targeting specificity of RNA-guided Cas9 nucleases. Nature Biotechnology. 2013.</p></li>
<li><p>Doench JG, et al. Optimized sgRNA design to maximize activity and minimize off-target effects of CRISPR-Cas9. Nature Biotechnology. 2016.</p></li>
<li><p>Concordet JP, Haeussler M. CRISPOR: intuitive guide selection for CRISPR/Cas9 genome editing experiments and screens. Nucleic Acids Research. 2018.</p></li>
<li><p>Elsea SH, Girirajan S. Smith-Magenis syndrome. European Journal of Human Genetics. 2008.</p></li>
<li><p>Johansson AS, et al. rAAV-CRISPRa therapy for Smith-Magenis Syndrome. 2022. PMC9762195.</p></li>
<li><p>SantaLucia J. A unified view of polymer, dumbbell, and oligonucleotide DNA nearest-neighbor thermodynamics. PNAS. 1998.</p></li>
</ol>
<hr class="docutils" />
<p>Corresponding author: Dylan Vaca</p>
<p>Hardware validation performed on IBM Torino, December 2025</p>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#abstract">Abstract</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispra-for-gene-therapy">1.1 CRISPRa for Gene Therapy</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#smith-magenis-syndrome">1.2 Smith-Magenis Syndrome</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-guide-design-challenge">1.3 The Guide Design Challenge</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-informed-design">1.4 Quantum-Informed Design</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods">2. Methods</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#target-region">2.1 Target Region</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#pam-site-identification">2.2 PAM Site Identification</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-layer-scoring">2.3 Multi-Layer Scoring</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-hamiltonian-encoding">2.4 Quantum Hamiltonian Encoding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simulation-protocol">2.5 Simulation Protocol</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hardware-validation">2.6 Hardware Validation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispor-validation">2.7 CRISPOR Validation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#results">3. Results</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#candidate-screening">3.1 Candidate Screening</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#top-candidates">3.2 Top Candidates</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hardware-validation-results">3.3 Hardware Validation Results</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#crispor-genome-wide-analysis">3.4 CRISPOR Genome-Wide Analysis</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#quantum-specificity-correlation">3.5 Quantum-Specificity Correlation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#final-recommendation">3.6 Final Recommendation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#discussion">4. Discussion</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#advantages-of-quantum-informed-design">4.1 Advantages of Quantum-Informed Design</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-binding-specificity-connection">4.2 The Binding-Specificity Connection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#comparison-to-published-work">4.3 Comparison to Published Work</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clinical-translation-path">4.4 Clinical Translation Path</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#limitations">4.5 Limitations</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#conclusion">5. Conclusion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-extended">6. Methods (Extended)</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-retrieval">6.1 Sequence Retrieval</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#phaselab-pipeline">6.2 PhaseLab Pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hardware-execution">6.3 Hardware Execution</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-availability">7. Data Availability</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#figures">8. Figures</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      Â© Copyright 2025, Dylan Vaca.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>