
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" data-theme="auto">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>phaselab.crispr.enhancer &#8212; PhaseLab 0.6.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "auto";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "auto";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=09465d15" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=2bd037f0"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=30646c52"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/phaselab/crispr/enhancer';</script>
    <link rel="icon" href="../../../_static/favicon.svg"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="auto">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/logo-lm.svg" class="logo__image only-light" alt=""/>
    <img src="../../../_static/logo.svg" class="logo__image only-dark pst-js-only" alt=""/>
  
  
    <p class="title logo__title">PhaseLab</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/explanations/index.html">
    Explanations
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../reference/index.html">
    API Reference
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../research/index.html">
    Research Applications
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../examples/index.html">
    Examples
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../citing.html">
    Citing PhaseLab
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/explanations/index.html">
    Explanations
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../reference/index.html">
    API Reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../research/index.html">
    Research Applications
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../citing.html">
    Citing PhaseLab
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">phaselab.crispr.enhancer</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for phaselab.crispr.enhancer</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">PhaseLab Enhancer Targeting: CRISPRa guide design for enhancer activation.</span>

<span class="sd">Implements:</span>
<span class="sd">- Enhancer region identification and scoring</span>
<span class="sd">- eQTL-based enhancer-gene linking</span>
<span class="sd">- CRISPRa enhancer activation prediction</span>
<span class="sd">- Multi-enhancer combinatorial activation</span>
<span class="sd">- Tissue-specific enhancer models</span>

<span class="sd">References:</span>
<span class="sd">- crisprQTL: Mapping enhancer-gene pairs via CRISPR screens</span>
<span class="sd">- pgBoost: Probabilistic scoring for enhancer-gene links</span>
<span class="sd">- Fulco et al. Activity-by-contact model for enhancer prediction</span>
<span class="sd">- ENCODE enhancer annotations (cCREs)</span>

<span class="sd">IMPORTANT: Many genes are difficult to activate through promoter-only</span>
<span class="sd">CRISPRa. Targeting enhancers can provide stronger or tissue-specific</span>
<span class="sd">activation for genes like MYC, NRXN1, SCN2A, and MECP2.</span>

<span class="sd">Version: 0.5.0</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Any</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">dataclasses</span><span class="w"> </span><span class="kn">import</span> <span class="n">dataclass</span><span class="p">,</span> <span class="n">field</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">..core.coherence</span><span class="w"> </span><span class="kn">import</span> <span class="n">coherence_score</span><span class="p">,</span> <span class="n">go_no_go</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">..core.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">E_MINUS_2</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<span class="c1"># ENCODE cCRE (candidate Cis-Regulatory Element) types</span>
<span class="n">CCRE_TYPES</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;pELS&#39;</span><span class="p">:</span> <span class="s1">&#39;proximal enhancer-like signature&#39;</span><span class="p">,</span>   <span class="c1"># &lt;2kb from TSS</span>
    <span class="s1">&#39;dELS&#39;</span><span class="p">:</span> <span class="s1">&#39;distal enhancer-like signature&#39;</span><span class="p">,</span>     <span class="c1"># &gt;2kb from TSS</span>
    <span class="s1">&#39;PLS&#39;</span><span class="p">:</span> <span class="s1">&#39;promoter-like signature&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CTCF-only&#39;</span><span class="p">:</span> <span class="s1">&#39;CTCF-bound only&#39;</span><span class="p">,</span>
    <span class="s1">&#39;DNase-H3K4me3&#39;</span><span class="p">:</span> <span class="s1">&#39;promoter-proximal&#39;</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Activity-by-contact model parameters</span>
<span class="n">ABC_MODEL_PARAMS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;activity_weight&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">,</span>      <span class="c1"># Weight for enhancer activity signal</span>
    <span class="s1">&#39;contact_weight&#39;</span><span class="p">:</span> <span class="mf">0.4</span><span class="p">,</span>       <span class="c1"># Weight for 3D contact frequency</span>
    <span class="s1">&#39;distance_decay&#39;</span><span class="p">:</span> <span class="mi">5000</span><span class="p">,</span>      <span class="c1"># Distance for contact decay (bp)</span>
    <span class="s1">&#39;min_abc_score&#39;</span><span class="p">:</span> <span class="mf">0.02</span><span class="p">,</span>       <span class="c1"># Minimum ABC score for significance</span>
<span class="p">}</span>


<div class="viewcode-block" id="Enhancer">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.Enhancer">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">Enhancer</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Enhancer element annotation.&quot;&quot;&quot;</span>
    <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="c1"># Type and activity</span>
    <span class="n">element_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;dELS&quot;</span>   <span class="c1"># pELS, dELS, etc.</span>
    <span class="n">activity_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span>  <span class="c1"># H3K27ac/ATAC signal</span>

    <span class="c1"># Target gene information</span>
    <span class="n">target_gene</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">target_gene_tss</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">distance_to_tss</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Linking scores</span>
    <span class="n">abc_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span>       <span class="c1"># Activity-by-contact</span>
    <span class="n">eqtl_support</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>   <span class="c1"># Has eQTL evidence</span>
    <span class="n">hi_c_contact</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span>    <span class="c1"># Hi-C contact frequency</span>

    <span class="c1"># Tissue specificity</span>
    <span class="n">tissue</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;generic&quot;</span>
    <span class="n">tissue_specific</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">center</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">is_proximal</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">abs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">distance_to_tss</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2000</span></div>



<div class="viewcode-block" id="EnhancerGuideResult">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.EnhancerGuideResult">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">EnhancerGuideResult</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Result from enhancer-targeting guide design.&quot;&quot;&quot;</span>
    <span class="n">guide_sequence</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">guide_position</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">enhancer</span><span class="p">:</span> <span class="n">Enhancer</span>

    <span class="c1"># Scores</span>
    <span class="n">activation_potential</span><span class="p">:</span> <span class="nb">float</span>
    <span class="n">guide_quality</span><span class="p">:</span> <span class="nb">float</span>
    <span class="n">combined_score</span><span class="p">:</span> <span class="nb">float</span>

    <span class="c1"># Coherence</span>
    <span class="n">coherence_R</span><span class="p">:</span> <span class="nb">float</span>
    <span class="n">go_no_go</span><span class="p">:</span> <span class="nb">str</span>

    <span class="c1"># Recommendations</span>
    <span class="n">recommendation</span><span class="p">:</span> <span class="nb">str</span></div>



<div class="viewcode-block" id="EnhancerConfig">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.EnhancerConfig">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">EnhancerConfig</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Configuration for enhancer targeting.&quot;&quot;&quot;</span>

    <span class="c1"># Enhancer definition</span>
    <span class="n">min_enhancer_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span>
    <span class="n">max_enhancer_distance</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1_000_000</span>  <span class="c1"># 1Mb max</span>

    <span class="c1"># Activity thresholds</span>
    <span class="n">min_activity_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.3</span>
    <span class="n">min_abc_score</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.02</span>

    <span class="c1"># Guide design</span>
    <span class="n">guides_per_enhancer</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="n">prefer_enhancer_center</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="c1"># Tissue specificity</span>
    <span class="n">require_tissue_match</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">target_tissue</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;generic&quot;</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">calculate_abc_score</span><span class="p">(</span>
    <span class="n">activity</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">contact</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate Activity-by-Contact (ABC) score for enhancer-gene link.</span>

<span class="sd">    ABC model: Score = Activity × Contact / Σ(Activity × Contact)</span>

<span class="sd">    Based on Fulco et al. (2019) - predicts enhancer-gene pairs.</span>

<span class="sd">    Args:</span>
<span class="sd">        activity: Enhancer activity signal (H3K27ac, ATAC)</span>
<span class="sd">        contact: 3D contact frequency (Hi-C, if available)</span>
<span class="sd">        distance: Distance to target gene TSS</span>
<span class="sd">        config: ABC model parameters</span>

<span class="sd">    Returns:</span>
<span class="sd">        ABC score (higher = stronger link)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">config</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">config</span> <span class="o">=</span> <span class="n">ABC_MODEL_PARAMS</span>

    <span class="c1"># If no contact data, estimate from distance</span>
    <span class="k">if</span> <span class="n">contact</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Power-law decay with distance</span>
        <span class="n">contact</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="p">(</span><span class="n">distance</span> <span class="o">/</span> <span class="n">config</span><span class="p">[</span><span class="s1">&#39;distance_decay&#39;</span><span class="p">])</span> <span class="o">**</span> <span class="mf">1.5</span><span class="p">)</span>

    <span class="c1"># ABC score (not normalized - would need all enhancers for normalization)</span>
    <span class="n">abc</span> <span class="o">=</span> <span class="n">activity</span> <span class="o">*</span> <span class="n">contact</span>

    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="n">abc</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">estimate_enhancer_activity</span><span class="p">(</span>
    <span class="n">h3k27ac_signal</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="n">atac_signal</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="n">dnase_signal</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Estimate enhancer activity from epigenetic signals.</span>

<span class="sd">    Args:</span>
<span class="sd">        h3k27ac_signal: H3K27ac ChIP-seq signal (active mark)</span>
<span class="sd">        atac_signal: ATAC-seq signal</span>
<span class="sd">        dnase_signal: DNase-seq signal</span>

<span class="sd">    Returns:</span>
<span class="sd">        Activity score 0-1</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">signals</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># H3K27ac is the strongest indicator of active enhancers</span>
    <span class="k">if</span> <span class="n">h3k27ac_signal</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">signals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">h3k27ac_signal</span> <span class="o">*</span> <span class="mf">1.2</span><span class="p">)</span>  <span class="c1"># Higher weight</span>

    <span class="k">if</span> <span class="n">atac_signal</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">signals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atac_signal</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">dnase_signal</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">signals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dnase_signal</span> <span class="o">*</span> <span class="mf">0.8</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">signals</span><span class="p">:</span>
        <span class="k">return</span> <span class="mf">0.0</span>

    <span class="c1"># Combine signals (weighted average, capped at 1.0)</span>
    <span class="n">activity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">signals</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">float</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">activity</span><span class="p">))</span>


<div class="viewcode-block" id="score_enhancer_for_activation">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.score_enhancer_for_activation">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">score_enhancer_for_activation</span><span class="p">(</span>
    <span class="n">enhancer</span><span class="p">:</span> <span class="n">Enhancer</span><span class="p">,</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">EnhancerConfig</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Score an enhancer&#39;s potential for CRISPRa activation.</span>

<span class="sd">    Considers:</span>
<span class="sd">    - Enhancer activity (already active enhancers harder to boost)</span>
<span class="sd">    - Distance to target gene</span>
<span class="sd">    - ABC score (link confidence)</span>
<span class="sd">    - Tissue specificity</span>

<span class="sd">    Args:</span>
<span class="sd">        enhancer: Enhancer to score</span>
<span class="sd">        config: EnhancerConfig</span>

<span class="sd">    Returns:</span>
<span class="sd">        Scoring dictionary</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">config</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">config</span> <span class="o">=</span> <span class="n">EnhancerConfig</span><span class="p">()</span>

    <span class="n">scores</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Activity-based potential</span>
    <span class="c1"># Moderately active enhancers have highest activation potential</span>
    <span class="c1"># Very active = already maxed, inactive = hard to activate</span>
    <span class="k">if</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">activity_score</span> <span class="o">&lt;</span> <span class="mf">0.2</span><span class="p">:</span>
        <span class="n">activity_potential</span> <span class="o">=</span> <span class="mf">0.4</span>  <span class="c1"># Weak enhancer, harder to activate</span>
    <span class="k">elif</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">activity_score</span> <span class="o">&lt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="n">activity_potential</span> <span class="o">=</span> <span class="mf">0.9</span>  <span class="c1"># Moderate, good potential</span>
    <span class="k">elif</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">activity_score</span> <span class="o">&lt;</span> <span class="mf">0.8</span><span class="p">:</span>
        <span class="n">activity_potential</span> <span class="o">=</span> <span class="mf">0.7</span>  <span class="c1"># Already active, some headroom</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">activity_potential</span> <span class="o">=</span> <span class="mf">0.3</span>  <span class="c1"># Near maximum, little room to boost</span>

    <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;activity_potential&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">activity_potential</span>

    <span class="c1"># Distance factor</span>
    <span class="c1"># Proximal enhancers easier to link to target gene</span>
    <span class="n">distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">enhancer</span><span class="o">.</span><span class="n">distance_to_tss</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">distance</span> <span class="o">&lt;</span> <span class="mi">2000</span><span class="p">:</span>
        <span class="n">distance_factor</span> <span class="o">=</span> <span class="mf">1.0</span>
    <span class="k">elif</span> <span class="n">distance</span> <span class="o">&lt;</span> <span class="mi">10000</span><span class="p">:</span>
        <span class="n">distance_factor</span> <span class="o">=</span> <span class="mf">0.9</span>
    <span class="k">elif</span> <span class="n">distance</span> <span class="o">&lt;</span> <span class="mi">100000</span><span class="p">:</span>
        <span class="n">distance_factor</span> <span class="o">=</span> <span class="mf">0.7</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">distance_factor</span> <span class="o">=</span> <span class="mf">0.5</span>

    <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;distance_factor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">distance_factor</span>

    <span class="c1"># ABC score contribution</span>
    <span class="k">if</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">abc_score</span> <span class="o">&gt;=</span> <span class="n">config</span><span class="o">.</span><span class="n">min_abc_score</span><span class="p">:</span>
        <span class="n">abc_factor</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">abc_score</span> <span class="o">/</span> <span class="mf">0.1</span><span class="p">)</span>  <span class="c1"># Normalize</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">abc_factor</span> <span class="o">=</span> <span class="mf">0.3</span>  <span class="c1"># Below threshold, uncertain link</span>

    <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;abc_factor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">abc_factor</span>

    <span class="c1"># eQTL evidence boost</span>
    <span class="k">if</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">eqtl_support</span><span class="p">:</span>
        <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;eqtl_boost&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.2</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;eqtl_boost&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="c1"># Combined score</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="p">(</span>
        <span class="mf">0.35</span> <span class="o">*</span> <span class="n">activity_potential</span> <span class="o">+</span>
        <span class="mf">0.25</span> <span class="o">*</span> <span class="n">distance_factor</span> <span class="o">+</span>
        <span class="mf">0.25</span> <span class="o">*</span> <span class="n">abc_factor</span> <span class="o">+</span>
        <span class="mf">0.15</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="k">if</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">eqtl_support</span> <span class="k">else</span> <span class="mf">0.5</span><span class="p">)</span>
    <span class="p">)</span> <span class="o">*</span> <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;eqtl_boost&#39;</span><span class="p">]</span>

    <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;combined_potential&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">combined</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">scores</span></div>



<div class="viewcode-block" id="identify_target_enhancers">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.identify_target_enhancers">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">identify_target_enhancers</span><span class="p">(</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_tss</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">gene_chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">candidate_enhancers</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Enhancer</span><span class="p">],</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">EnhancerConfig</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">Enhancer</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identify enhancers likely to regulate a target gene.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_symbol: Target gene symbol</span>
<span class="sd">        gene_tss: Gene TSS position</span>
<span class="sd">        gene_chrom: Gene chromosome</span>
<span class="sd">        candidate_enhancers: List of candidate enhancers</span>
<span class="sd">        config: EnhancerConfig</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of (enhancer, link_score) tuples, sorted by score</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">config</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">config</span> <span class="o">=</span> <span class="n">EnhancerConfig</span><span class="p">()</span>

    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">enh</span> <span class="ow">in</span> <span class="n">candidate_enhancers</span><span class="p">:</span>
        <span class="c1"># Check chromosome match</span>
        <span class="k">if</span> <span class="n">enh</span><span class="o">.</span><span class="n">chrom</span> <span class="o">!=</span> <span class="n">gene_chrom</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="c1"># Calculate distance</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">enh</span><span class="o">.</span><span class="n">center</span> <span class="o">-</span> <span class="n">gene_tss</span><span class="p">)</span>

        <span class="c1"># Check maximum distance</span>
        <span class="k">if</span> <span class="n">distance</span> <span class="o">&gt;</span> <span class="n">config</span><span class="o">.</span><span class="n">max_enhancer_distance</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="c1"># Update enhancer with distance</span>
        <span class="n">enh</span><span class="o">.</span><span class="n">distance_to_tss</span> <span class="o">=</span> <span class="n">distance</span>
        <span class="n">enh</span><span class="o">.</span><span class="n">target_gene</span> <span class="o">=</span> <span class="n">gene_symbol</span>
        <span class="n">enh</span><span class="o">.</span><span class="n">target_gene_tss</span> <span class="o">=</span> <span class="n">gene_tss</span>

        <span class="c1"># Calculate ABC score if not already set</span>
        <span class="k">if</span> <span class="n">enh</span><span class="o">.</span><span class="n">abc_score</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">enh</span><span class="o">.</span><span class="n">abc_score</span> <span class="o">=</span> <span class="n">calculate_abc_score</span><span class="p">(</span>
                <span class="n">activity</span><span class="o">=</span><span class="n">enh</span><span class="o">.</span><span class="n">activity_score</span><span class="p">,</span>
                <span class="n">contact</span><span class="o">=</span><span class="n">enh</span><span class="o">.</span><span class="n">hi_c_contact</span><span class="p">,</span>
                <span class="n">distance</span><span class="o">=</span><span class="n">distance</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># Score for CRISPRa potential</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="n">score_enhancer_for_activation</span><span class="p">(</span><span class="n">enh</span><span class="p">,</span> <span class="n">config</span><span class="p">)</span>
        <span class="n">link_score</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[</span><span class="s1">&#39;combined_potential&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">link_score</span> <span class="o">&gt;=</span> <span class="mf">0.3</span><span class="p">:</span>  <span class="c1"># Minimum threshold</span>
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">enh</span><span class="p">,</span> <span class="n">link_score</span><span class="p">))</span>

    <span class="c1"># Sort by link score</span>
    <span class="n">results</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">results</span></div>



<div class="viewcode-block" id="design_enhancer_guides">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.design_enhancer_guides">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">design_enhancer_guides</span><span class="p">(</span>
    <span class="n">enhancer</span><span class="p">:</span> <span class="n">Enhancer</span><span class="p">,</span>
    <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">enhancer_start_in_seq</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">EnhancerConfig</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">guide_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">pam</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;NGG&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">EnhancerGuideResult</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Design CRISPRa guides targeting an enhancer.</span>

<span class="sd">    For enhancer activation, guides should target:</span>
<span class="sd">    - Near the enhancer center (peak of activity)</span>
<span class="sd">    - Accessible regions within enhancer</span>
<span class="sd">    - Avoid repressive elements</span>

<span class="sd">    Args:</span>
<span class="sd">        enhancer: Target enhancer</span>
<span class="sd">        sequence: DNA sequence containing enhancer</span>
<span class="sd">        enhancer_start_in_seq: Position of enhancer start in sequence</span>
<span class="sd">        config: EnhancerConfig</span>
<span class="sd">        guide_length: Guide length (default 20)</span>
<span class="sd">        pam: PAM sequence</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of EnhancerGuideResult objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">config</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">config</span> <span class="o">=</span> <span class="n">EnhancerConfig</span><span class="p">()</span>

    <span class="c1"># Import here to avoid circular imports</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.pam_scan</span><span class="w"> </span><span class="kn">import</span> <span class="n">find_pam_sites</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.scoring</span><span class="w"> </span><span class="kn">import</span> <span class="n">gc_content</span><span class="p">,</span> <span class="n">delta_g_santalucia</span><span class="p">,</span> <span class="n">mit_specificity_score</span>

    <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Define search region within enhancer</span>
    <span class="n">enh_start</span> <span class="o">=</span> <span class="n">enhancer_start_in_seq</span>
    <span class="n">enh_end</span> <span class="o">=</span> <span class="n">enh_start</span> <span class="o">+</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">length</span>

    <span class="c1"># Find PAM sites within enhancer</span>
    <span class="n">all_hits</span> <span class="o">=</span> <span class="n">find_pam_sites</span><span class="p">(</span>
        <span class="n">sequence</span><span class="p">,</span>
        <span class="n">pam</span><span class="o">=</span><span class="n">pam</span><span class="p">,</span>
        <span class="n">guide_length</span><span class="o">=</span><span class="n">guide_length</span><span class="p">,</span>
        <span class="n">both_strands</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Filter to enhancer region</span>
    <span class="n">enhancer_hits</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">h</span> <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">all_hits</span>
        <span class="k">if</span> <span class="n">enh_start</span> <span class="o">&lt;=</span> <span class="n">h</span><span class="o">.</span><span class="n">guide_start</span> <span class="o">&lt;</span> <span class="n">enh_end</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">enhancer_hits</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No guides found in enhancer </span><span class="si">{</span><span class="n">enhancer</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>

    <span class="c1"># Score enhancer potential</span>
    <span class="n">enh_scores</span> <span class="o">=</span> <span class="n">score_enhancer_for_activation</span><span class="p">(</span><span class="n">enhancer</span><span class="p">,</span> <span class="n">config</span><span class="p">)</span>

    <span class="c1"># Score each guide</span>
    <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">enhancer_hits</span><span class="p">:</span>
        <span class="n">guide_seq</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">guide</span>
        <span class="n">guide_pos</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">guide_start</span>

        <span class="c1"># Basic guide quality</span>
        <span class="n">gc</span> <span class="o">=</span> <span class="n">gc_content</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gc</span> <span class="o">&lt;</span> <span class="mf">0.35</span> <span class="ow">or</span> <span class="n">gc</span> <span class="o">&gt;</span> <span class="mf">0.75</span><span class="p">:</span>
            <span class="k">continue</span>  <span class="c1"># Skip bad GC</span>

        <span class="n">delta_g</span> <span class="o">=</span> <span class="n">delta_g_santalucia</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">)</span>
        <span class="n">mit</span> <span class="o">=</span> <span class="n">mit_specificity_score</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">)</span>

        <span class="c1"># Position within enhancer (prefer center)</span>
        <span class="n">enh_center</span> <span class="o">=</span> <span class="n">enh_start</span> <span class="o">+</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">length</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="n">distance_from_center</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">guide_pos</span> <span class="o">-</span> <span class="n">enh_center</span><span class="p">)</span>
        <span class="n">center_factor</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">distance_from_center</span> <span class="o">/</span> <span class="p">(</span><span class="n">enhancer</span><span class="o">.</span><span class="n">length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span>

        <span class="c1"># Guide quality score</span>
        <span class="n">guide_quality</span> <span class="o">=</span> <span class="p">(</span>
            <span class="mf">0.3</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="k">if</span> <span class="mf">0.45</span> <span class="o">&lt;=</span> <span class="n">gc</span> <span class="o">&lt;=</span> <span class="mf">0.65</span> <span class="k">else</span> <span class="mf">0.6</span><span class="p">)</span> <span class="o">+</span>
            <span class="mf">0.3</span> <span class="o">*</span> <span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="n">delta_g</span> <span class="o">/</span> <span class="mf">25.0</span><span class="p">)</span> <span class="o">+</span>
            <span class="mf">0.2</span> <span class="o">*</span> <span class="p">(</span><span class="n">mit</span> <span class="o">/</span> <span class="mf">100.0</span><span class="p">)</span> <span class="o">+</span>
            <span class="mf">0.2</span> <span class="o">*</span> <span class="n">center_factor</span>
        <span class="p">)</span>

        <span class="c1"># Activation potential (enhancer × guide quality)</span>
        <span class="n">activation_potential</span> <span class="o">=</span> <span class="n">enh_scores</span><span class="p">[</span><span class="s1">&#39;combined_potential&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">guide_quality</span>

        <span class="c1"># IR coherence (simplified)</span>
        <span class="n">R_bar</span> <span class="o">=</span> <span class="n">_compute_guide_coherence</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">)</span>
        <span class="n">status</span> <span class="o">=</span> <span class="n">go_no_go</span><span class="p">(</span><span class="n">R_bar</span><span class="p">)</span>

        <span class="c1"># Combined score</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="p">(</span>
            <span class="mf">0.4</span> <span class="o">*</span> <span class="n">activation_potential</span> <span class="o">+</span>
            <span class="mf">0.3</span> <span class="o">*</span> <span class="n">guide_quality</span> <span class="o">+</span>
            <span class="mf">0.3</span> <span class="o">*</span> <span class="n">R_bar</span>
        <span class="p">)</span>

        <span class="c1"># Recommendation</span>
        <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="s2">&quot;NO-GO&quot;</span><span class="p">:</span>
            <span class="n">recommendation</span> <span class="o">=</span> <span class="s2">&quot;LOW COHERENCE: Guide prediction unreliable&quot;</span>
        <span class="k">elif</span> <span class="n">activation_potential</span> <span class="o">&gt;=</span> <span class="mf">0.6</span><span class="p">:</span>
            <span class="n">recommendation</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;EXCELLENT: High activation potential (</span><span class="si">{</span><span class="n">activation_potential</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">)&quot;</span>
        <span class="k">elif</span> <span class="n">activation_potential</span> <span class="o">&gt;=</span> <span class="mf">0.4</span><span class="p">:</span>
            <span class="n">recommendation</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;GOOD: Moderate activation potential (</span><span class="si">{</span><span class="n">activation_potential</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">)&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">recommendation</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;MARGINAL: Low activation potential (</span><span class="si">{</span><span class="n">activation_potential</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">)&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="n">EnhancerGuideResult</span><span class="p">(</span>
            <span class="n">guide_sequence</span><span class="o">=</span><span class="n">guide_seq</span><span class="p">,</span>
            <span class="n">guide_position</span><span class="o">=</span><span class="n">guide_pos</span><span class="p">,</span>
            <span class="n">enhancer</span><span class="o">=</span><span class="n">enhancer</span><span class="p">,</span>
            <span class="n">activation_potential</span><span class="o">=</span><span class="n">activation_potential</span><span class="p">,</span>
            <span class="n">guide_quality</span><span class="o">=</span><span class="n">guide_quality</span><span class="p">,</span>
            <span class="n">combined_score</span><span class="o">=</span><span class="n">combined</span><span class="p">,</span>
            <span class="n">coherence_R</span><span class="o">=</span><span class="n">R_bar</span><span class="p">,</span>
            <span class="n">go_no_go</span><span class="o">=</span><span class="n">status</span><span class="p">,</span>
            <span class="n">recommendation</span><span class="o">=</span><span class="n">recommendation</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

    <span class="c1"># Sort by combined score</span>
    <span class="n">results</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">r</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">combined_score</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Return top guides</span>
    <span class="k">return</span> <span class="n">results</span><span class="p">[:</span><span class="n">config</span><span class="o">.</span><span class="n">guides_per_enhancer</span><span class="p">]</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">_compute_guide_coherence</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute IR coherence using ATLAS-Q enhanced backend (v0.6.0+).&quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">.coherence_utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">compute_guide_coherence</span>
    <span class="k">return</span> <span class="n">compute_guide_coherence</span><span class="p">(</span><span class="n">guide_seq</span><span class="p">,</span> <span class="n">use_atlas_q</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


<div class="viewcode-block" id="predict_enhancer_activation_effect">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.predict_enhancer_activation_effect">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">predict_enhancer_activation_effect</span><span class="p">(</span>
    <span class="n">enhancer</span><span class="p">:</span> <span class="n">Enhancer</span><span class="p">,</span>
    <span class="n">guide_quality</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">baseline_expression</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Predict effect of enhancer CRISPRa on target gene expression.</span>

<span class="sd">    Args:</span>
<span class="sd">        enhancer: Target enhancer</span>
<span class="sd">        guide_quality: Quality score of guide used</span>
<span class="sd">        baseline_expression: Baseline gene expression level</span>

<span class="sd">    Returns:</span>
<span class="sd">        Prediction dictionary with fold-change estimates</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Enhancer activation typically gives 1.5-5x fold change</span>
    <span class="c1"># depending on enhancer strength and gene responsiveness</span>

    <span class="c1"># Base fold change from enhancer activity</span>
    <span class="n">base_fold</span> <span class="o">=</span> <span class="mf">1.5</span> <span class="o">+</span> <span class="mf">2.5</span> <span class="o">*</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">abc_score</span>

    <span class="c1"># Adjust for guide quality</span>
    <span class="n">fold_change</span> <span class="o">=</span> <span class="n">base_fold</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.5</span> <span class="o">+</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">guide_quality</span><span class="p">)</span>

    <span class="c1"># Distance penalty</span>
    <span class="n">distance_penalty</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="mf">0.3</span> <span class="o">*</span> <span class="nb">min</span><span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">distance_to_tss</span> <span class="o">/</span> <span class="mi">500000</span><span class="p">)</span>
    <span class="n">fold_change</span> <span class="o">*=</span> <span class="n">distance_penalty</span>

    <span class="c1"># Confidence based on evidence</span>
    <span class="k">if</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">eqtl_support</span><span class="p">:</span>
        <span class="n">confidence</span> <span class="o">=</span> <span class="mf">0.8</span>
    <span class="k">elif</span> <span class="n">enhancer</span><span class="o">.</span><span class="n">abc_score</span> <span class="o">&gt;=</span> <span class="mf">0.05</span><span class="p">:</span>
        <span class="n">confidence</span> <span class="o">=</span> <span class="mf">0.6</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">confidence</span> <span class="o">=</span> <span class="mf">0.4</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;predicted_fold_change&#39;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">fold_change</span><span class="p">),</span>
        <span class="s1">&#39;fold_change_low&#39;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">fold_change</span> <span class="o">*</span> <span class="mf">0.6</span><span class="p">),</span>
        <span class="s1">&#39;fold_change_high&#39;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">fold_change</span> <span class="o">*</span> <span class="mf">1.5</span><span class="p">),</span>
        <span class="s1">&#39;confidence&#39;</span><span class="p">:</span> <span class="n">confidence</span><span class="p">,</span>
        <span class="s1">&#39;baseline_expression&#39;</span><span class="p">:</span> <span class="n">baseline_expression</span><span class="p">,</span>
        <span class="s1">&#39;predicted_expression&#39;</span><span class="p">:</span> <span class="n">baseline_expression</span> <span class="o">*</span> <span class="n">fold_change</span><span class="p">,</span>
    <span class="p">}</span></div>



<div class="viewcode-block" id="compare_promoter_vs_enhancer">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.compare_promoter_vs_enhancer">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compare_promoter_vs_enhancer</span><span class="p">(</span>
    <span class="n">promoter_guides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">],</span>
    <span class="n">enhancer_guides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">EnhancerGuideResult</span><span class="p">],</span>
    <span class="n">target_expression</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">baseline_expression</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compare promoter CRISPRa vs enhancer CRISPRa strategies.</span>

<span class="sd">    Args:</span>
<span class="sd">        promoter_guides: Promoter-targeting guide results</span>
<span class="sd">        enhancer_guides: Enhancer-targeting guide results</span>
<span class="sd">        target_expression: Desired expression level</span>
<span class="sd">        baseline_expression: Current baseline expression</span>

<span class="sd">    Returns:</span>
<span class="sd">        Comparison and recommendation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;promoter_option&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s1">&#39;enhancer_option&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s1">&#39;recommendation&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="c1"># Analyze promoter option</span>
    <span class="k">if</span> <span class="n">promoter_guides</span><span class="p">:</span>
        <span class="n">best_promoter</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span>
            <span class="n">promoter_guides</span><span class="p">,</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">g</span><span class="p">:</span> <span class="n">g</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;combined_score&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">promoter_fold</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">+</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">best_promoter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;combined_score&#39;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>
        <span class="n">promoter_expression</span> <span class="o">=</span> <span class="n">baseline_expression</span> <span class="o">*</span> <span class="n">promoter_fold</span>

        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;promoter_option&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;best_guide&#39;</span><span class="p">:</span> <span class="n">best_promoter</span><span class="p">,</span>
            <span class="s1">&#39;predicted_fold_change&#39;</span><span class="p">:</span> <span class="n">promoter_fold</span><span class="p">,</span>
            <span class="s1">&#39;predicted_expression&#39;</span><span class="p">:</span> <span class="n">promoter_expression</span><span class="p">,</span>
            <span class="s1">&#39;reaches_target&#39;</span><span class="p">:</span> <span class="n">promoter_expression</span> <span class="o">&gt;=</span> <span class="n">target_expression</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="c1"># Analyze enhancer option</span>
    <span class="k">if</span> <span class="n">enhancer_guides</span><span class="p">:</span>
        <span class="n">best_enhancer</span> <span class="o">=</span> <span class="n">enhancer_guides</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># Already sorted</span>
        <span class="n">enh_pred</span> <span class="o">=</span> <span class="n">predict_enhancer_activation_effect</span><span class="p">(</span>
            <span class="n">best_enhancer</span><span class="o">.</span><span class="n">enhancer</span><span class="p">,</span>
            <span class="n">best_enhancer</span><span class="o">.</span><span class="n">guide_quality</span><span class="p">,</span>
            <span class="n">baseline_expression</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;enhancer_option&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;best_guide&#39;</span><span class="p">:</span> <span class="n">best_enhancer</span><span class="p">,</span>
            <span class="s1">&#39;predicted_fold_change&#39;</span><span class="p">:</span> <span class="n">enh_pred</span><span class="p">[</span><span class="s1">&#39;predicted_fold_change&#39;</span><span class="p">],</span>
            <span class="s1">&#39;predicted_expression&#39;</span><span class="p">:</span> <span class="n">enh_pred</span><span class="p">[</span><span class="s1">&#39;predicted_expression&#39;</span><span class="p">],</span>
            <span class="s1">&#39;confidence&#39;</span><span class="p">:</span> <span class="n">enh_pred</span><span class="p">[</span><span class="s1">&#39;confidence&#39;</span><span class="p">],</span>
            <span class="s1">&#39;reaches_target&#39;</span><span class="p">:</span> <span class="n">enh_pred</span><span class="p">[</span><span class="s1">&#39;predicted_expression&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">target_expression</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="c1"># Generate recommendation</span>
    <span class="n">promoter_reaches</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;promoter_option&#39;</span><span class="p">]</span> <span class="ow">and</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;promoter_option&#39;</span><span class="p">][</span><span class="s1">&#39;reaches_target&#39;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">enhancer_reaches</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;enhancer_option&#39;</span><span class="p">]</span> <span class="ow">and</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;enhancer_option&#39;</span><span class="p">][</span><span class="s1">&#39;reaches_target&#39;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">promoter_reaches</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">enhancer_reaches</span><span class="p">:</span>
        <span class="n">recommendation</span> <span class="o">=</span> <span class="s2">&quot;PROMOTER: Sufficient activation from promoter CRISPRa&quot;</span>
    <span class="k">elif</span> <span class="n">enhancer_reaches</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">promoter_reaches</span><span class="p">:</span>
        <span class="n">recommendation</span> <span class="o">=</span> <span class="s2">&quot;ENHANCER: Enhancer targeting needed to reach target&quot;</span>
    <span class="k">elif</span> <span class="n">promoter_reaches</span> <span class="ow">and</span> <span class="n">enhancer_reaches</span><span class="p">:</span>
        <span class="c1"># Both work - recommend based on confidence/simplicity</span>
        <span class="n">recommendation</span> <span class="o">=</span> <span class="s2">&quot;BOTH VIABLE: Promoter simpler, enhancer may give higher activation&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">recommendation</span> <span class="o">=</span> <span class="s2">&quot;COMBINATION: May need multi-guide or promoter+enhancer strategy&quot;</span>

    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;recommendation&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">recommendation</span>

    <span class="k">return</span> <span class="n">results</span></div>



<span class="c1"># Pre-defined enhancer databases for common therapeutic targets</span>
<span class="n">KNOWN_ENHANCERS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;RAI1&#39;</span><span class="p">:</span> <span class="p">[</span>
        <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RAI1_enh1&#39;</span><span class="p">,</span>
            <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="s1">&#39;chr17&#39;</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="mi">17679000</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="mi">17680000</span><span class="p">,</span>
            <span class="s1">&#39;distance_to_tss&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">2500</span><span class="p">,</span>
            <span class="s1">&#39;activity_score&#39;</span><span class="p">:</span> <span class="mf">0.65</span><span class="p">,</span>
            <span class="s1">&#39;tissue&#39;</span><span class="p">:</span> <span class="s1">&#39;brain&#39;</span><span class="p">,</span>
        <span class="p">},</span>
    <span class="p">],</span>
    <span class="s1">&#39;SCN2A&#39;</span><span class="p">:</span> <span class="p">[</span>
        <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;SCN2A_enh1&#39;</span><span class="p">,</span>
            <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="s1">&#39;chr2&#39;</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="mi">165235000</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="mi">165236500</span><span class="p">,</span>
            <span class="s1">&#39;distance_to_tss&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">4400</span><span class="p">,</span>
            <span class="s1">&#39;activity_score&#39;</span><span class="p">:</span> <span class="mf">0.72</span><span class="p">,</span>
            <span class="s1">&#39;tissue&#39;</span><span class="p">:</span> <span class="s1">&#39;brain&#39;</span><span class="p">,</span>
        <span class="p">},</span>
    <span class="p">],</span>
    <span class="s1">&#39;MECP2&#39;</span><span class="p">:</span> <span class="p">[</span>
        <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;MECP2_enh1&#39;</span><span class="p">,</span>
            <span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="s1">&#39;chrX&#39;</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="mi">154025000</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="mi">154026000</span><span class="p">,</span>
            <span class="s1">&#39;distance_to_tss&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">15000</span><span class="p">,</span>
            <span class="s1">&#39;activity_score&#39;</span><span class="p">:</span> <span class="mf">0.58</span><span class="p">,</span>
            <span class="s1">&#39;tissue&#39;</span><span class="p">:</span> <span class="s1">&#39;brain&#39;</span><span class="p">,</span>
        <span class="p">},</span>
    <span class="p">],</span>
<span class="p">}</span>


<div class="viewcode-block" id="get_known_enhancers">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.get_known_enhancers">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_known_enhancers</span><span class="p">(</span><span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Enhancer</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get known enhancers for a gene from built-in database.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene symbol</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of Enhancer objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">KNOWN_ENHANCERS</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>

    <span class="n">enhancers</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">enh_data</span> <span class="ow">in</span> <span class="n">KNOWN_ENHANCERS</span><span class="p">[</span><span class="n">gene</span><span class="p">]:</span>
        <span class="n">enh</span> <span class="o">=</span> <span class="n">Enhancer</span><span class="p">(</span>
            <span class="n">chrom</span><span class="o">=</span><span class="n">enh_data</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span>
            <span class="n">start</span><span class="o">=</span><span class="n">enh_data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span>
            <span class="n">end</span><span class="o">=</span><span class="n">enh_data</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span>
            <span class="n">name</span><span class="o">=</span><span class="n">enh_data</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
            <span class="n">activity_score</span><span class="o">=</span><span class="n">enh_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;activity_score&#39;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span>
            <span class="n">distance_to_tss</span><span class="o">=</span><span class="n">enh_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;distance_to_tss&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">tissue</span><span class="o">=</span><span class="n">enh_data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;tissue&#39;</span><span class="p">,</span> <span class="s1">&#39;generic&#39;</span><span class="p">),</span>
            <span class="n">target_gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">enhancers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">enh</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">enhancers</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Dylan Vaca.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>