
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" data-theme="auto">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>phaselab.crispr.region &#8212; PhaseLab 0.6.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "auto";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "auto";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=09465d15" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=2bd037f0"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=30646c52"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/phaselab/crispr/region';</script>
    <link rel="icon" href="../../../_static/favicon.svg"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.6.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="auto">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/logo-lm.svg" class="logo__image only-light" alt=""/>
    <img src="../../../_static/logo.svg" class="logo__image only-dark pst-js-only" alt=""/>
  
  
    <p class="title logo__title">PhaseLab</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/explanations/index.html">
    Explanations
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../reference/index.html">
    API Reference
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../research/index.html">
    Research Applications
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../examples/index.html">
    Examples
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../citing.html">
    Citing PhaseLab
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../quickstart.html">
    Quick Start
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/tutorials/index.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/howtos/index.html">
    How-To Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../user_guide/explanations/index.html">
    Explanations
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../reference/index.html">
    API Reference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../research/index.html">
    Research Applications
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../developer/index.html">
    Developer Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../examples/index.html">
    Examples
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../citing.html">
    Citing PhaseLab
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/followthesapper/phaselab" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github-square fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/phaselab/" title="PyPI" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fas fa-box fa-lg" aria-hidden="true"></i>
            <span class="sr-only">PyPI</span></a>
        </li>
</ul></div>
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">phaselab.crispr.region</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for phaselab.crispr.region</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">PhaseLab Region Declaration Module (v0.9.2+).</span>

<span class="sd">Handles promoter/region definition for guide enumeration.</span>

<span class="sd">KEY PRINCIPLE: PhaseLab evaluates guides *conditional on* a declared</span>
<span class="sd">promoter hypothesis. It does NOT decide which TSS is &quot;correct&quot; - that</span>
<span class="sd">is a biological declaration made by the user.</span>

<span class="sd">This module provides:</span>
<span class="sd">- TSS coordinate loading from GENCODE/RefSeq</span>
<span class="sd">- Multi-TSS hypothesis support</span>
<span class="sd">- Window-based region definition</span>
<span class="sd">- Region validation and metadata</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">dataclasses</span><span class="w"> </span><span class="kn">import</span> <span class="n">dataclass</span><span class="p">,</span> <span class="n">field</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">enum</span><span class="w"> </span><span class="kn">import</span> <span class="n">Enum</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="GenomeBuild">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.GenomeBuild">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">GenomeBuild</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Supported genome builds.&quot;&quot;&quot;</span>
    <span class="n">HG38</span> <span class="o">=</span> <span class="s2">&quot;hg38&quot;</span>
    <span class="n">HG19</span> <span class="o">=</span> <span class="s2">&quot;hg19&quot;</span>
    <span class="n">MM10</span> <span class="o">=</span> <span class="s2">&quot;mm10&quot;</span>
    <span class="n">MM39</span> <span class="o">=</span> <span class="s2">&quot;mm39&quot;</span></div>



<div class="viewcode-block" id="TSSSource">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.TSSSource">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">TSSSource</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;TSS annotation sources.&quot;&quot;&quot;</span>
    <span class="n">GENCODE</span> <span class="o">=</span> <span class="s2">&quot;gencode&quot;</span>
    <span class="n">REFSEQ</span> <span class="o">=</span> <span class="s2">&quot;refseq&quot;</span>
    <span class="n">ENSEMBL</span> <span class="o">=</span> <span class="s2">&quot;ensembl&quot;</span>
    <span class="n">CUSTOM</span> <span class="o">=</span> <span class="s2">&quot;custom&quot;</span>  <span class="c1"># User-provided coordinates</span></div>



<span class="k">class</span><span class="w"> </span><span class="nc">Modality</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;CRISPR modality - affects window positioning.&quot;&quot;&quot;</span>
    <span class="n">CRISPRA</span> <span class="o">=</span> <span class="s2">&quot;CRISPRa&quot;</span>      <span class="c1"># Activation: typically -400 to -50 from TSS</span>
    <span class="n">CRISPRI</span> <span class="o">=</span> <span class="s2">&quot;CRISPRi&quot;</span>      <span class="c1"># Interference: typically -50 to +300 from TSS</span>
    <span class="n">KNOCKOUT</span> <span class="o">=</span> <span class="s2">&quot;knockout&quot;</span>    <span class="c1"># Cutting: typically exon-focused</span>
    <span class="n">PRIME</span> <span class="o">=</span> <span class="s2">&quot;prime&quot;</span>          <span class="c1"># Prime editing: position-specific</span>
    <span class="n">BASE</span> <span class="o">=</span> <span class="s2">&quot;base&quot;</span>            <span class="c1"># Base editing: position-specific</span>


<div class="viewcode-block" id="TSSAnnotation">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.TSSAnnotation">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">TSSAnnotation</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transcription Start Site annotation.</span>

<span class="sd">    Represents a single TSS from an annotation source.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">transcript_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">position</span><span class="p">:</span> <span class="nb">int</span>           <span class="c1"># 0-based genomic coordinate</span>
    <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span>             <span class="c1"># &#39;+&#39; or &#39;-&#39;</span>
    <span class="n">source</span><span class="p">:</span> <span class="n">TSSSource</span>
    <span class="n">source_version</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Confidence/evidence (if available)</span>
    <span class="n">support_level</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># GENCODE support level (1-5)</span>
    <span class="n">is_canonical</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>           <span class="c1"># MANE/canonical transcript</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;TSS(</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">transcript_id</span><span class="si">}</span><span class="s2"> @ </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="si">}{</span><span class="bp">self</span><span class="o">.</span><span class="n">strand</span><span class="si">}</span><span class="s2">)&quot;</span></div>



<div class="viewcode-block" id="Window">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.Window">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">Window</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Relative window around a TSS.</span>

<span class="sd">    Coordinates are relative to TSS (negative = upstream).</span>
<span class="sd">    For minus-strand genes, upstream/downstream are automatically flipped.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">upstream</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">400</span>     <span class="c1"># bp upstream of TSS (negative direction)</span>
    <span class="n">downstream</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">50</span>    <span class="c1"># bp downstream of TSS (positive direction)</span>
    <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="Window.to_genomic">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.Window.to_genomic">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">to_genomic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tss_position</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert relative window to absolute genomic coordinates.</span>

<span class="sd">        Args:</span>
<span class="sd">            tss_position: Absolute TSS position.</span>
<span class="sd">            strand: &#39;+&#39; or &#39;-&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            (start, end) as 0-based half-open interval.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">tss_position</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">upstream</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">tss_position</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">downstream</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Minus strand: upstream is in positive direction</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">tss_position</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">downstream</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">tss_position</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">upstream</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span></div>
</div>



<span class="c1"># Default windows for each modality</span>
<span class="n">DEFAULT_WINDOWS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">Modality</span><span class="o">.</span><span class="n">CRISPRA</span><span class="p">:</span> <span class="n">Window</span><span class="p">(</span><span class="n">upstream</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CRISPRa_standard&quot;</span><span class="p">),</span>
    <span class="n">Modality</span><span class="o">.</span><span class="n">CRISPRI</span><span class="p">:</span> <span class="n">Window</span><span class="p">(</span><span class="n">upstream</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;CRISPRi_standard&quot;</span><span class="p">),</span>
    <span class="n">Modality</span><span class="o">.</span><span class="n">KNOCKOUT</span><span class="p">:</span> <span class="n">Window</span><span class="p">(</span><span class="n">upstream</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;exon1&quot;</span><span class="p">),</span>
<span class="p">}</span>


<div class="viewcode-block" id="Region">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.Region">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">Region</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A genomic region for guide enumeration.</span>

<span class="sd">    Represents a specific interval to scan for candidate guides,</span>
<span class="sd">    with full provenance tracking.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">start</span><span class="p">:</span> <span class="nb">int</span>              <span class="c1"># 0-based</span>
    <span class="n">end</span><span class="p">:</span> <span class="nb">int</span>                <span class="c1"># exclusive</span>
    <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span>

    <span class="c1"># Provenance</span>
    <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">tss_id</span><span class="p">:</span> <span class="nb">str</span>             <span class="c1"># Which TSS this region is derived from</span>
    <span class="n">tss_position</span><span class="p">:</span> <span class="nb">int</span>
    <span class="n">window</span><span class="p">:</span> <span class="n">Window</span>
    <span class="n">modality</span><span class="p">:</span> <span class="n">Modality</span>

    <span class="c1"># Metadata</span>
    <span class="n">genome_build</span><span class="p">:</span> <span class="n">GenomeBuild</span> <span class="o">=</span> <span class="n">GenomeBuild</span><span class="o">.</span><span class="n">HG38</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start</span>

    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">region_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Unique identifier for this region.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">window</span><span class="o">.</span><span class="n">name</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="s1">&#39;custom&#39;</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;Region(</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">start</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="si">}</span><span class="s2"> [</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2">])&quot;</span></div>



<div class="viewcode-block" id="RegionSet">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionSet">[docs]</a>
<span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">RegionSet</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Collection of regions for a single target.</span>

<span class="sd">    Supports multiple TSS hypotheses and multiple windows per TSS.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">genome_build</span><span class="p">:</span> <span class="n">GenomeBuild</span>
    <span class="n">modality</span><span class="p">:</span> <span class="n">Modality</span>

    <span class="n">regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Region</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>
    <span class="n">tss_annotations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">TSSAnnotation</span><span class="p">]</span> <span class="o">=</span> <span class="n">field</span><span class="p">(</span><span class="n">default_factory</span><span class="o">=</span><span class="nb">list</span><span class="p">)</span>

    <span class="c1"># Metadata for reproducibility</span>
    <span class="n">tss_source</span><span class="p">:</span> <span class="n">TSSSource</span> <span class="o">=</span> <span class="n">TSSSource</span><span class="o">.</span><span class="n">GENCODE</span>
    <span class="n">tss_source_version</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="RegionSet.add_region">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionSet.add_region">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_region</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">region</span><span class="p">:</span> <span class="n">Region</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a region to the set.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">region</span><span class="p">)</span></div>


    <span class="nd">@property</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">total_length</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Total bp covered (may overlap).&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">regions</span><span class="p">)</span>

<div class="viewcode-block" id="RegionSet.to_dict">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionSet.to_dict">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">to_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Serialize for manifest.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
            <span class="s1">&#39;gene_symbol&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span>
            <span class="s1">&#39;genome_build&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome_build</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="s1">&#39;modality&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">modality</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="s1">&#39;tss_source&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_source</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="s1">&#39;tss_source_version&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_source_version</span><span class="p">,</span>
            <span class="s1">&#39;num_regions&#39;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">regions</span><span class="p">),</span>
            <span class="s1">&#39;total_length&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span><span class="p">,</span>
            <span class="s1">&#39;regions&#39;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">{</span>
                    <span class="s1">&#39;region_id&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">region_id</span><span class="p">,</span>
                    <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                    <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">start</span><span class="p">,</span>
                    <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
                    <span class="s1">&#39;strand&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                    <span class="s1">&#39;tss_id&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">tss_id</span><span class="p">,</span>
                    <span class="s1">&#39;tss_position&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">tss_position</span><span class="p">,</span>
                    <span class="s1">&#39;window&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="s1">&#39;upstream&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">window</span><span class="o">.</span><span class="n">upstream</span><span class="p">,</span>
                        <span class="s1">&#39;downstream&#39;</span><span class="p">:</span> <span class="n">r</span><span class="o">.</span><span class="n">window</span><span class="o">.</span><span class="n">downstream</span><span class="p">,</span>
                    <span class="p">},</span>
                <span class="p">}</span>
                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">regions</span>
            <span class="p">],</span>
        <span class="p">}</span></div>
</div>



<span class="c1"># =============================================================================</span>
<span class="c1"># REGION BUILDER</span>
<span class="c1"># =============================================================================</span>

<div class="viewcode-block" id="RegionBuilder">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">RegionBuilder</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Build regions from TSS annotations and window specifications.</span>

<span class="sd">    This is the main interface for region declaration.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; builder = RegionBuilder(GenomeBuild.HG38, Modality.CRISPRA)</span>
<span class="sd">        &gt;&gt;&gt; builder.add_tss(TSSAnnotation(...))</span>
<span class="sd">        &gt;&gt;&gt; builder.add_window(Window(upstream=400, downstream=50))</span>
<span class="sd">        &gt;&gt;&gt; region_set = builder.build(&quot;RAI1&quot;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">genome_build</span><span class="p">:</span> <span class="n">GenomeBuild</span> <span class="o">=</span> <span class="n">GenomeBuild</span><span class="o">.</span><span class="n">HG38</span><span class="p">,</span>
        <span class="n">modality</span><span class="p">:</span> <span class="n">Modality</span> <span class="o">=</span> <span class="n">Modality</span><span class="o">.</span><span class="n">CRISPRA</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome_build</span> <span class="o">=</span> <span class="n">genome_build</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">modality</span> <span class="o">=</span> <span class="n">modality</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">TSSAnnotation</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Window</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tss_source</span> <span class="o">=</span> <span class="n">TSSSource</span><span class="o">.</span><span class="n">CUSTOM</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tss_source_version</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="RegionBuilder.add_tss">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder.add_tss">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_tss</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tss</span><span class="p">:</span> <span class="n">TSSAnnotation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;RegionBuilder&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a TSS annotation.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tss</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">tss</span><span class="o">.</span><span class="n">source</span> <span class="o">!=</span> <span class="n">TSSSource</span><span class="o">.</span><span class="n">CUSTOM</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tss_source</span> <span class="o">=</span> <span class="n">tss</span><span class="o">.</span><span class="n">source</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tss_source_version</span> <span class="o">=</span> <span class="n">tss</span><span class="o">.</span><span class="n">source_version</span>
        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="RegionBuilder.add_tss_manual">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder.add_tss_manual">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_tss_manual</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">chromosome</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">position</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="n">strand</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">transcript_id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;manual&quot;</span><span class="p">,</span>
        <span class="n">gene_id</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;RegionBuilder&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a TSS with manual coordinates.&quot;&quot;&quot;</span>
        <span class="n">tss</span> <span class="o">=</span> <span class="n">TSSAnnotation</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span> <span class="ow">or</span> <span class="n">gene_symbol</span><span class="p">,</span>
            <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="n">transcript_id</span><span class="p">,</span>
            <span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span>
            <span class="n">position</span><span class="o">=</span><span class="n">position</span><span class="p">,</span>
            <span class="n">strand</span><span class="o">=</span><span class="n">strand</span><span class="p">,</span>
            <span class="n">source</span><span class="o">=</span><span class="n">TSSSource</span><span class="o">.</span><span class="n">CUSTOM</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_tss</span><span class="p">(</span><span class="n">tss</span><span class="p">)</span></div>


<div class="viewcode-block" id="RegionBuilder.add_window">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder.add_window">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">add_window</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">window</span><span class="p">:</span> <span class="n">Window</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;RegionBuilder&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a window specification.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">window</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="RegionBuilder.use_default_window">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder.use_default_window">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">use_default_window</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;RegionBuilder&#39;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use the default window for the current modality.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">modality</span> <span class="ow">in</span> <span class="n">DEFAULT_WINDOWS</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">DEFAULT_WINDOWS</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">modality</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Fallback</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Window</span><span class="p">(</span><span class="n">upstream</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;default&quot;</span><span class="p">))</span>
        <span class="k">return</span> <span class="bp">self</span></div>


<div class="viewcode-block" id="RegionBuilder.build">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.RegionBuilder.build">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">build</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_symbol</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RegionSet</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build the region set.</span>

<span class="sd">        Generates one region per (TSS, window) combination.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_symbol: Override gene symbol (uses first TSS if not provided).</span>

<span class="sd">        Returns:</span>
<span class="sd">            RegionSet with all declared regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No TSS annotations provided. Use add_tss() first.&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No windows specified, using default for modality&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">use_default_window</span><span class="p">()</span>

        <span class="c1"># Determine gene info</span>
        <span class="n">first_tss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene_symbol</span> <span class="o">=</span> <span class="n">gene_symbol</span> <span class="ow">or</span> <span class="n">first_tss</span><span class="o">.</span><span class="n">gene_symbol</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="n">first_tss</span><span class="o">.</span><span class="n">gene_id</span>

        <span class="n">region_set</span> <span class="o">=</span> <span class="n">RegionSet</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="n">gene_id</span><span class="p">,</span>
            <span class="n">gene_symbol</span><span class="o">=</span><span class="n">gene_symbol</span><span class="p">,</span>
            <span class="n">genome_build</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_build</span><span class="p">,</span>
            <span class="n">modality</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">modality</span><span class="p">,</span>
            <span class="n">tss_annotations</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span>
            <span class="n">tss_source</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_source</span><span class="p">,</span>
            <span class="n">tss_source_version</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">tss_source_version</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Generate regions for each (TSS, window) combination</span>
        <span class="k">for</span> <span class="n">tss</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tss_list</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">window</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">windows</span><span class="p">:</span>
                <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">window</span><span class="o">.</span><span class="n">to_genomic</span><span class="p">(</span><span class="n">tss</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">tss</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>

                <span class="n">region</span> <span class="o">=</span> <span class="n">Region</span><span class="p">(</span>
                    <span class="n">chromosome</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                    <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
                    <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
                    <span class="n">strand</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                    <span class="n">gene_id</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">gene_id</span><span class="p">,</span>
                    <span class="n">gene_symbol</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">gene_symbol</span><span class="p">,</span>
                    <span class="n">tss_id</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">transcript_id</span><span class="p">,</span>
                    <span class="n">tss_position</span><span class="o">=</span><span class="n">tss</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                    <span class="n">window</span><span class="o">=</span><span class="n">window</span><span class="p">,</span>
                    <span class="n">modality</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">modality</span><span class="p">,</span>
                    <span class="n">genome_build</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">genome_build</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">region_set</span><span class="o">.</span><span class="n">add_region</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Built </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">region_set</span><span class="o">.</span><span class="n">regions</span><span class="p">)</span><span class="si">}</span><span class="s2"> regions for </span><span class="si">{</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">region_set</span></div>
</div>



<span class="c1"># =============================================================================</span>
<span class="c1"># TSS LOOKUP (from annotations)</span>
<span class="c1"># =============================================================================</span>

<span class="c1"># Common TSS data (built-in for key genes)</span>
<span class="c1"># This is a fallback when annotation files aren&#39;t available</span>
<span class="n">BUILTIN_TSS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="p">(</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span> <span class="s1">&#39;hg38&#39;</span><span class="p">):</span> <span class="p">[</span>
        <span class="n">TSSAnnotation</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;ENSG00000108557&#39;</span><span class="p">,</span>
            <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;ENST00000225568.10&#39;</span><span class="p">,</span>
            <span class="n">chromosome</span><span class="o">=</span><span class="s1">&#39;chr17&#39;</span><span class="p">,</span>
            <span class="n">position</span><span class="o">=</span><span class="mi">17584786</span><span class="p">,</span>  <span class="c1"># GENCODE v44</span>
            <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
            <span class="n">source</span><span class="o">=</span><span class="n">TSSSource</span><span class="o">.</span><span class="n">GENCODE</span><span class="p">,</span>
            <span class="n">source_version</span><span class="o">=</span><span class="s1">&#39;v44&#39;</span><span class="p">,</span>
            <span class="n">is_canonical</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">],</span>
    <span class="p">(</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span> <span class="s1">&#39;hg19&#39;</span><span class="p">):</span> <span class="p">[</span>
        <span class="n">TSSAnnotation</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;ENSG00000108557&#39;</span><span class="p">,</span>
            <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;RAI1&#39;</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;ENST00000225568&#39;</span><span class="p">,</span>
            <span class="n">chromosome</span><span class="o">=</span><span class="s1">&#39;chr17&#39;</span><span class="p">,</span>
            <span class="n">position</span><span class="o">=</span><span class="mi">17672279</span><span class="p">,</span>
            <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
            <span class="n">source</span><span class="o">=</span><span class="n">TSSSource</span><span class="o">.</span><span class="n">GENCODE</span><span class="p">,</span>
            <span class="n">source_version</span><span class="o">=</span><span class="s1">&#39;v19&#39;</span><span class="p">,</span>
            <span class="n">is_canonical</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">],</span>
    <span class="p">(</span><span class="s1">&#39;Rai1&#39;</span><span class="p">,</span> <span class="s1">&#39;mm10&#39;</span><span class="p">):</span> <span class="p">[</span>
        <span class="n">TSSAnnotation</span><span class="p">(</span>
            <span class="n">gene_id</span><span class="o">=</span><span class="s1">&#39;ENSMUSG00000062475&#39;</span><span class="p">,</span>
            <span class="n">gene_symbol</span><span class="o">=</span><span class="s1">&#39;Rai1&#39;</span><span class="p">,</span>
            <span class="n">transcript_id</span><span class="o">=</span><span class="s1">&#39;ENSMUST00000078364&#39;</span><span class="p">,</span>
            <span class="n">chromosome</span><span class="o">=</span><span class="s1">&#39;chr11&#39;</span><span class="p">,</span>
            <span class="n">position</span><span class="o">=</span><span class="mi">60104470</span><span class="p">,</span>  <span class="c1"># From Chang et al. paper context</span>
            <span class="n">strand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span>
            <span class="n">source</span><span class="o">=</span><span class="n">TSSSource</span><span class="o">.</span><span class="n">GENCODE</span><span class="p">,</span>
            <span class="n">source_version</span><span class="o">=</span><span class="s1">&#39;vM25&#39;</span><span class="p">,</span>
            <span class="n">is_canonical</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">],</span>
<span class="p">}</span>


<div class="viewcode-block" id="get_tss_for_gene">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.get_tss_for_gene">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_tss_for_gene</span><span class="p">(</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genome_build</span><span class="p">:</span> <span class="n">GenomeBuild</span><span class="p">,</span>
    <span class="n">source</span><span class="p">:</span> <span class="n">TSSSource</span> <span class="o">=</span> <span class="n">TSSSource</span><span class="o">.</span><span class="n">GENCODE</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">TSSAnnotation</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Look up TSS annotations for a gene.</span>

<span class="sd">    Currently uses built-in data. In production, would query</span>
<span class="sd">    GENCODE/RefSeq annotation files.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_symbol: Gene name (e.g., &quot;RAI1&quot;).</span>
<span class="sd">        genome_build: Genome build.</span>
<span class="sd">        source: Annotation source to use.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of TSSAnnotation objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="n">genome_build</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">BUILTIN_TSS</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">BUILTIN_TSS</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>

    <span class="c1"># Also try case variations</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="n">tss_list</span> <span class="ow">in</span> <span class="n">BUILTIN_TSS</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">g</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">gene_symbol</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">and</span> <span class="n">b</span> <span class="o">==</span> <span class="n">genome_build</span><span class="o">.</span><span class="n">value</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">tss_list</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No built-in TSS data for </span><span class="si">{</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">genome_build</span><span class="o">.</span><span class="n">value</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[]</span></div>



<div class="viewcode-block" id="build_regions_for_gene">
<a class="viewcode-back" href="../../../reference/index.html#phaselab.crispr.build_regions_for_gene">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">build_regions_for_gene</span><span class="p">(</span>
    <span class="n">gene_symbol</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genome_build</span><span class="p">:</span> <span class="n">GenomeBuild</span> <span class="o">=</span> <span class="n">GenomeBuild</span><span class="o">.</span><span class="n">HG38</span><span class="p">,</span>
    <span class="n">modality</span><span class="p">:</span> <span class="n">Modality</span> <span class="o">=</span> <span class="n">Modality</span><span class="o">.</span><span class="n">CRISPRA</span><span class="p">,</span>
    <span class="n">windows</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">Window</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">use_canonical_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">RegionSet</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convenience function to build regions for a gene using built-in TSS data.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_symbol: Gene name.</span>
<span class="sd">        genome_build: Genome build.</span>
<span class="sd">        modality: CRISPR modality.</span>
<span class="sd">        windows: Custom windows (uses default if None).</span>
<span class="sd">        use_canonical_only: Only use canonical/MANE transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        RegionSet for the gene.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; regions = build_regions_for_gene(&quot;RAI1&quot;, GenomeBuild.HG38, Modality.CRISPRA)</span>
<span class="sd">        &gt;&gt;&gt; print(regions.regions[0])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tss_list</span> <span class="o">=</span> <span class="n">get_tss_for_gene</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="n">genome_build</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">tss_list</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;No TSS data found for </span><span class="si">{</span><span class="n">gene_symbol</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">genome_build</span><span class="o">.</span><span class="n">value</span><span class="si">}</span><span class="s2">). &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Use RegionBuilder.add_tss_manual() to provide coordinates.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">use_canonical_only</span><span class="p">:</span>
        <span class="n">canonical</span> <span class="o">=</span> <span class="p">[</span><span class="n">t</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tss_list</span> <span class="k">if</span> <span class="n">t</span><span class="o">.</span><span class="n">is_canonical</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">canonical</span><span class="p">:</span>
            <span class="n">tss_list</span> <span class="o">=</span> <span class="n">canonical</span>

    <span class="n">builder</span> <span class="o">=</span> <span class="n">RegionBuilder</span><span class="p">(</span><span class="n">genome_build</span><span class="p">,</span> <span class="n">modality</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">tss</span> <span class="ow">in</span> <span class="n">tss_list</span><span class="p">:</span>
        <span class="n">builder</span><span class="o">.</span><span class="n">add_tss</span><span class="p">(</span><span class="n">tss</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">windows</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">w</span> <span class="ow">in</span> <span class="n">windows</span><span class="p">:</span>
            <span class="n">builder</span><span class="o">.</span><span class="n">add_window</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">builder</span><span class="o">.</span><span class="n">use_default_window</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">builder</span><span class="o">.</span><span class="n">build</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
       Copyright 2025, Dylan Vaca.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>